// @cloudflare/kumo/dist/components/sidebar.js
"use client";
import { A as r, B as i, K as S, S as s, w as b, y as d, x as n, d as o, e as u, f as t, h as l, g as M, c as p, r as A, i as I, l as R, m as B, k as C, v as _, j as g, n as T, p as m, o as D, a as E, t as G, u as K, b as U, q as c, s as v, z as L } from "../chunks/sidebar-o8y71x814ptc0xpf.js";
export {
  r as KUMO_SIDEBAR_DEFAULT_VARIANTS,
  i as KUMO_SIDEBAR_STYLING,
  S as KUMO_SIDEBAR_VARIANTS,
  s as Sidebar,
  b as SidebarCollapsible,
  d as SidebarCollapsibleContent,
  n as SidebarCollapsibleTrigger,
  o as SidebarContent,
  u as SidebarFooter,
  t as SidebarGroup,
  l as SidebarGroupContent,
  M as SidebarGroupLabel,
  p as SidebarHeader,
  A as SidebarInput,
  I as SidebarMenu,
  R as SidebarMenuAction,
  B as SidebarMenuBadge,
  C as SidebarMenuButton,
  _ as SidebarMenuChevron,
  g as SidebarMenuItem,
  T as SidebarMenuSub,
  m as SidebarMenuSubButton,
  D as SidebarMenuSubItem,
  E as SidebarProvider,
  G as SidebarRail,
  K as SidebarResizeHandle,
  U as SidebarRoot,
  c as SidebarSeparator,
  v as SidebarTrigger,
  L as useSidebar
};
