// @cloudflare/kumo/dist/index.js
"use client";
import { B as Y } from "./chunks/badge-kqox9toi0sygfbno.js";
import { B as O } from "./chunks/banner-eux4y8xaogjg64af.js";
import { a as Q } from "./chunks/banner-eux4y8xaogjg64af.js";
import { B as l } from "./chunks/button-n859eyw550yi2b9z.js";
import { L as aa, R as ea, b as oa } from "./chunks/button-n859eyw550yi2b9z.js";
import { D as sa } from "./chunks/date-range-picker-o8uyril1kogvhoei.js";
import { C as ia } from "./chunks/checkbox-dfl2fr8nchh43qfc.js";
import { C as la } from "./chunks/clipboard-text-ic9k5qjkljlr9z3b.js";
import { C as ma, a as pa } from "./chunks/code-bbnjm8vk9hxl129r.js";
import { C as Ta } from "./chunks/combobox-f5nyw45yiwx5f69l.js";
import { a as M, D as U, c as N, e as u } from "./chunks/dialog-n6uc2s99xwdn2pnb.js";
import { d as Aa, b as ba } from "./chunks/dialog-n6uc2s99xwdn2pnb.js";
import { D as Sa } from "./chunks/dropdown-mftv4iv9nzhprg81.js";
import { C as _a } from "./chunks/collapsible-jvebgqfqljzokj8h.js";
import { F as Ia, a as La, K as Ra, f as Ea } from "./chunks/field-mil8efu3x0s68eed.js";
import { b as Ma, K as Ua, L as Na, a as Ka, l as ha } from "./chunks/label-d14ibjmcbk1qmyrt.js";
import { I as K } from "./chunks/input-kmztt6h4mzy101ho.js";
import { i as Da } from "./chunks/input-kmztt6h4mzy101ho.js";
import { I as va, T as Pa } from "./chunks/input-area-hl1rdb9xcrqyt8xw.js";
import { I as Ba } from "./chunks/input-group-5luo0442jgsie018.js";
import { L as ya } from "./chunks/layer-card-eomdoafn3sfpih1d.js";
import { L as Ga } from "./chunks/loader-g8a6j76ue5nq0lr8.js";
import { S as ja } from "./chunks/skeleton-line-epxenksfesr2fkcv.js";
import { M as qa, u as Ha } from "./chunks/menubar-f6xelkurau8cl60f.js";
import { M as Xa } from "./chunks/meter-g1ja8cwtum0frcdj.js";
import { P as Ja } from "./chunks/pagination-kmtbb3twehv79tm8.js";
import { S as Za } from "./chunks/select-brzswxbhhf1ktx9t.js";
import { S as ee } from "./chunks/surface-ivrb4btwdoq91ytl.js";
import { S as re } from "./chunks/switch-c0avfhj6n85inmtb.js";
import { T as te } from "./chunks/tabs-fdkhdhv8kvnpp8nt.js";
import { T as ne } from "./chunks/table-htqdo9dgy8c6hwuq.js";
import { T as de } from "./chunks/text-be7ehenoyldhyjma.js";
import { a as pe, T as fe, u as Te } from "./chunks/toast-evn3ce20j18hpyqt.js";
import { T as Ae, a as be } from "./chunks/tooltip-fjxy4s4l75hjxp1x.js";
import { g as Se, K as xe, P as _e } from "./chunks/popover-f3t99000mahsnjzc.js";
import { a as Ie, K as Le, S as Re } from "./chunks/sensitive-input-gyf5hhgyolt07y1p.js";
import { b as Oe, K as Me, R as Ue, a as Ne, r as Ke } from "./chunks/radio-me5m5ei86beum5bo.js";
import { C as ge, a as De, K as Ve } from "./chunks/command-palette-ezbzp2fpbbo97ogf.js";
import { a as Pe, K as ke, L as Be, l as Fe } from "./chunks/link-ihastr6a2dmo1so5.js";
import { B as we } from "./chunks/breadcrumbs-davmangc0urzivbs.js";
import { E as ze } from "./chunks/empty-mmtirqntk6enx51o.js";
import { G as $e, a as qe, c as He, K as We, b as Xe, g as Ye } from "./chunks/grid-do93dv1rjggqxx7p.js";
import { C as Qe, a as Ze, K as ao, P as eo, g as oo } from "./chunks/cloudflare-logo-fya2cd2ljozkzjh8.js";
import { D as so } from "./chunks/date-picker-mjocqqbggb76n8rt.js";
import { Flow as io } from "./components/flow.js";
import { C as lo, a as mo, b as po, S as fo, T as To } from "./chunks/SankeyChart-i4vgcatj89zpgpl9.js";
import { A as Ao, b as bo, K as uo, a as So } from "./chunks/autocomplete-1hi2rgzo10bczkfa.js";
import { A as _o, B as Co, K as Io, S as Lo, w as Ro, y as Eo, x as Oo, d as Mo, e as Uo, f as No, h as Ko, g as ho, c as go, r as Do, i as Vo, l as vo, m as Po, k as ko, v as Bo, j as Fo, n as yo, p as wo, o as Go, a as zo, t as jo, u as $o, b as qo, q as Ho, s as Wo, z as Xo } from "./chunks/sidebar-o8y71x814ptc0xpf.js";
import { a as Jo, K as Qo, T as Zo } from "./chunks/table-of-contents-fzyv7uhnnyr13dqu.js";
import { c as h } from "./chunks/cn-ct4n7r74mh8y0f48.js";
import { s as er } from "./chunks/cn-ct4n7r74mh8y0f48.js";
import { L as rr, u as sr } from "./chunks/link-provider-mn2voeohon7cj9o4.js";
import { K as ir } from "./chunks/portal-provider-hwmkdmkpvct0cb76.js";
import { jsx as a, jsxs as e } from "react/jsx-runtime";
import { useState as S, useEffect as g, useCallback as d } from "react";
import { XIcon as D, WarningCircleIcon as V, CheckIcon as v, CopyIcon as P } from "@phosphor-icons/react";
import { i as lr } from "./chunks/vendor-base-ui-epfrwb4nfbd4btaz.js";
const H = {
  size: {
    sm: {
      classes: "",
      description: "Small dialog for simple delete confirmations"
    },
    base: {
      classes: "",
      description: "Default delete confirmation dialog size"
    }
  }
}, k = {
  size: "base"
};
function B({
  open: t,
  onOpenChange: x,
  resourceType: m,
  resourceName: o,
  onDelete: p,
  isDeleting: s = !1,
  caseSensitive: f = !0,
  deleteButtonText: _,
  size: C = k.size,
  errorMessage: T,
  className: I
}) {
  const [c, A] = S(""), [L, i] = S(!1);
  g(() => {
    t || (A(""), i(!1));
  }, [t]);
  const b = d(
    (r) => f ? r : r.toLowerCase(),
    [f]
  ), n = b(c) === b(o), R = d(async () => {
    !n || s || await p();
  }, [n, s, p]), E = d(async () => {
    await navigator.clipboard.writeText(o), i(!0), setTimeout(() => i(!1), 1500);
  }, [o]);
  return /* @__PURE__ */ a(M, { open: t, onOpenChange: x, children: /* @__PURE__ */ e(U, { size: C, className: h("p-0", I), children: [
    /* @__PURE__ */ e("div", { className: "flex items-center justify-between border-b border-kumo-line px-6 py-4", children: [
      /* @__PURE__ */ e(N, { className: "text-lg font-semibold", children: [
        "Delete ",
        o
      ] }),
      /* @__PURE__ */ a(
        u,
        {
          render: (r) => /* @__PURE__ */ a(
            l,
            {
              ...r,
              variant: "ghost",
              shape: "square",
              size: "sm",
              "aria-label": "Close",
              disabled: s,
              children: /* @__PURE__ */ a(D, { size: 18 })
            }
          )
        }
      )
    ] }),
    /* @__PURE__ */ e("div", { className: "flex flex-col p-6 gap-4", children: [
      /* @__PURE__ */ e("div", { className: "flex flex-col gap-2", children: [
        T && /* @__PURE__ */ a(O, { icon: /* @__PURE__ */ a(V, {}), variant: "error", children: T }),
        /* @__PURE__ */ e("p", { className: "text-base text-kumo-subtle max-w-prose text-pretty", children: [
          "This action cannot be undone. This will permanently delete the",
          " ",
          /* @__PURE__ */ a("span", { className: "font-medium text-kumo-default", children: o }),
          " ",
          m.toLowerCase(),
          "."
        ] })
      ] }),
      /* @__PURE__ */ e("div", { className: "flex flex-col gap-2", children: [
        /* @__PURE__ */ a("div", { className: "flex items-center gap-1.5 text-base", children: /* @__PURE__ */ e("span", { children: [
          "Type",
          " ",
          /* @__PURE__ */ e(
            "button",
            {
              className: "font-mono text-sm inline font-semibold bg-kumo-tint hover:bg-kumo-fill rounded-md px-2 py-1 group hover:cursor-pointer",
              onClick: E,
              "aria-label": `Copy ${o} to clipboard`,
              children: [
                o,
                L ? /* @__PURE__ */ a(
                  v,
                  {
                    size: 12,
                    weight: "bold",
                    className: "inline ml-1.5"
                  }
                ) : /* @__PURE__ */ a(
                  P,
                  {
                    size: 12,
                    weight: "bold",
                    className: "inline text-kumo-subtle group-hover:text-kumo-default ml-1.5"
                  }
                )
              ]
            }
          ),
          " ",
          "to confirm:"
        ] }) }),
        /* @__PURE__ */ a(
          K,
          {
            placeholder: o,
            value: c,
            onChange: (r) => A(r.target.value),
            disabled: s,
            autoComplete: "off",
            autoCorrect: "off",
            autoCapitalize: "off",
            spellCheck: !1,
            "aria-label": `Type ${o} to confirm deletion`,
            className: "w-full"
          }
        )
      ] })
    ] }),
    /* @__PURE__ */ e("div", { className: "flex justify-end gap-3 border-t border-kumo-line px-6 py-4", children: [
      /* @__PURE__ */ a(
        u,
        {
          render: (r) => /* @__PURE__ */ a(l, { ...r, variant: "secondary", disabled: s, children: "Cancel" })
        }
      ),
      /* @__PURE__ */ a(
        l,
        {
          variant: "destructive",
          onClick: R,
          disabled: !n || s,
          loading: s,
          children: _ || `Delete ${m}`
        }
      )
    ] })
  ] }) });
}
B.displayName = "DeleteResource";
export {
  Ao as Autocomplete,
  Y as Badge,
  O as Banner,
  Q as BannerVariant,
  we as Breadcrumbs,
  l as Button,
  lo as Chart,
  mo as ChartLegend,
  po as ChartPalette,
  ia as Checkbox,
  la as ClipboardText,
  Qe as CloudflareLogo,
  ma as Code,
  pa as CodeBlock,
  _a as Collapsible,
  Ta as Combobox,
  ge as CommandPalette,
  so as DatePicker,
  sa as DateRangePicker,
  B as DeleteResource,
  U as Dialog,
  u as DialogClose,
  Aa as DialogDescription,
  M as DialogRoot,
  N as DialogTitle,
  ba as DialogTrigger,
  Sa as DropdownMenu,
  ze as Empty,
  Ia as Field,
  io as Flow,
  $e as Grid,
  qe as GridItem,
  K as Input,
  va as InputArea,
  Ba as InputGroup,
  bo as KUMO_AUTOCOMPLETE_DEFAULT_VARIANTS,
  uo as KUMO_AUTOCOMPLETE_VARIANTS,
  Ze as KUMO_CLOUDFLARE_LOGO_DEFAULT_VARIANTS,
  ao as KUMO_CLOUDFLARE_LOGO_VARIANTS,
  De as KUMO_COMMAND_PALETTE_DEFAULT_VARIANTS,
  Ve as KUMO_COMMAND_PALETTE_VARIANTS,
  k as KUMO_DELETE_RESOURCE_DEFAULT_VARIANTS,
  H as KUMO_DELETE_RESOURCE_VARIANTS,
  La as KUMO_FIELD_DEFAULT_VARIANTS,
  Ra as KUMO_FIELD_VARIANTS,
  He as KUMO_GRID_DEFAULT_VARIANTS,
  We as KUMO_GRID_VARIANTS,
  Ma as KUMO_LABEL_DEFAULT_VARIANTS,
  Ua as KUMO_LABEL_VARIANTS,
  Pe as KUMO_LINK_DEFAULT_VARIANTS,
  ke as KUMO_LINK_VARIANTS,
  Se as KUMO_POPOVER_DEFAULT_VARIANTS,
  xe as KUMO_POPOVER_VARIANTS,
  Oe as KUMO_RADIO_DEFAULT_VARIANTS,
  Me as KUMO_RADIO_VARIANTS,
  Ie as KUMO_SENSITIVE_INPUT_DEFAULT_VARIANTS,
  Le as KUMO_SENSITIVE_INPUT_VARIANTS,
  _o as KUMO_SIDEBAR_DEFAULT_VARIANTS,
  Co as KUMO_SIDEBAR_STYLING,
  Io as KUMO_SIDEBAR_VARIANTS,
  Jo as KUMO_TABLE_OF_CONTENTS_DEFAULT_VARIANTS,
  Qo as KUMO_TABLE_OF_CONTENTS_VARIANTS,
  ir as KumoPortalProvider,
  Na as Label,
  ya as LayerCard,
  Be as Link,
  aa as LinkButton,
  rr as LinkProvider,
  Ga as Loader,
  qa as MenuBar,
  Xa as Meter,
  Ja as Pagination,
  _e as Popover,
  eo as PoweredByCloudflare,
  Ue as Radio,
  Ne as RadioGroup,
  ea as RefreshButton,
  fo as SankeyChart,
  Za as Select,
  Re as SensitiveInput,
  Lo as Sidebar,
  Ro as SidebarCollapsible,
  Eo as SidebarCollapsibleContent,
  Oo as SidebarCollapsibleTrigger,
  Mo as SidebarContent,
  Uo as SidebarFooter,
  No as SidebarGroup,
  Ko as SidebarGroupContent,
  ho as SidebarGroupLabel,
  go as SidebarHeader,
  Do as SidebarInput,
  Vo as SidebarMenu,
  vo as SidebarMenuAction,
  Po as SidebarMenuBadge,
  ko as SidebarMenuButton,
  Bo as SidebarMenuChevron,
  Fo as SidebarMenuItem,
  yo as SidebarMenuSub,
  wo as SidebarMenuSubButton,
  Go as SidebarMenuSubItem,
  zo as SidebarProvider,
  jo as SidebarRail,
  $o as SidebarResizeHandle,
  qo as SidebarRoot,
  Ho as SidebarSeparator,
  Wo as SidebarTrigger,
  ja as SkeletonLine,
  ee as Surface,
  re as Switch,
  ne as Table,
  Zo as TableOfContents,
  te as Tabs,
  de as Text,
  Pa as Textarea,
  To as TimeseriesChart,
  lr as Toast,
  pe as ToastProvider,
  fe as Toasty,
  Ae as Tooltip,
  be as TooltipProvider,
  So as autocompleteVariants,
  oa as buttonVariants,
  h as cn,
  Ea as fieldVariants,
  oo as generateCloudflareLogoSvg,
  Xe as gridItemVariants,
  Ye as gridVariants,
  Da as inputVariants,
  Ka as labelContentVariants,
  ha as labelVariants,
  Fe as linkVariants,
  Ke as radioVariants,
  er as safeRandomId,
  Te as useKumoToastManager,
  sr as useLinkComponent,
  Ha as useMenuNavigation,
  Xo as useSidebar
};
