// @cloudflare/kumo/dist/chunks/badge-kqox9toi0sygfbno.js
"use client";
import { jsx as n } from "react/jsx-runtime";
import { c as a } from "./cn-ct4n7r74mh8y0f48.js";
import { r as i } from "./resolve-variant-gw6eh7fa4st8ej7m.js";
const o = "inline-flex w-fit flex-none shrink-0 items-center justify-self-start rounded-full px-2 py-0.5 text-xs font-medium whitespace-nowrap", d = {
  variant: {
    /** Semantic token badges */
    primary: {
      classes: "bg-kumo-badge-inverted text-kumo-badge-inverted",
      description: "Primary badge"
    },
    secondary: {
      classes: "bg-kumo-fill text-kumo-badge-neutral-subtle",
      description: "Secondary badge"
    },
    error: {
      classes: "bg-kumo-danger-tint/60 text-kumo-danger",
      description: "Error badge"
    },
    warning: {
      classes: "bg-kumo-warning-tint/70 text-kumo-warning",
      description: "Warning badge"
    },
    success: {
      classes: "bg-kumo-success-tint/70 text-kumo-success",
      description: "Success badge"
    },
    destructive: {
      classes: "bg-kumo-badge-red text-white",
      description: "Deprecated. Use red instead."
    },
    info: {
      classes: "bg-kumo-info-tint/70 text-kumo-info",
      description: "Info badge"
    },
    beta: {
      classes: "border border-dashed border-kumo-brand bg-transparent text-kumo-link",
      description: "Indicates beta or experimental features"
    },
    outline: {
      classes: "border border-kumo-fill bg-transparent text-kumo-default",
      description: "Bordered badge with transparent background"
    },
    /** Other color token variants */
    red: {
      classes: "bg-kumo-badge-red text-white",
      description: "Red badge"
    },
    green: {
      classes: "bg-kumo-badge-green text-white",
      description: "Green badge"
    },
    neutral: {
      classes: "bg-kumo-badge-neutral text-white",
      description: "Neutral badge"
    },
    orange: {
      classes: "bg-kumo-badge-orange text-black",
      description: "Orange badge"
    },
    purple: {
      classes: "bg-kumo-badge-purple text-white",
      description: "Purple badge"
    },
    teal: {
      classes: "bg-kumo-badge-teal text-white",
      description: "Teal badge"
    },
    "teal-subtle": {
      classes: "bg-kumo-badge-teal-subtle text-kumo-badge-teal-subtle",
      description: "Subtle teal badge"
    },
    blue: {
      classes: "bg-kumo-badge-blue text-white",
      description: "Blue badge"
    }
  }
}, t = {
  variant: "primary"
};
function b({
  variant: e = t.variant
} = {}) {
  return a(
    // Base styles (exported as KUMO_BADGE_BASE_STYLES for Figma plugin)
    o,
    // Apply variant styles from KUMO_BADGE_VARIANTS (fallback to primary if variant not found)
    i(d.variant, e, t.variant).classes
  );
}
function l({
  variant: e = t.variant,
  className: s,
  children: r
}) {
  return /* @__PURE__ */ n("span", { className: a(b({ variant: e }), s), children: r });
}
export {
  l as B,
  o as K
};
