// @cloudflare/kumo/dist/chunks/banner-eux4y8xaogjg64af.js
"use client";
import { jsxs as i, jsx as n } from "react/jsx-runtime";
import { forwardRef as h, isValidElement as g } from "react";
import { c as r } from "./cn-ct4n7r74mh8y0f48.js";
import { r as k } from "./resolve-variant-gw6eh7fa4st8ej7m.js";
const E = "flex w-full items-start gap-3 rounded-lg border px-4 py-3 text-base", N = {
  variant: {
    default: {
      classes: "bg-kumo-info-tint/30 border-kumo-info/50 text-kumo-info selection:bg-kumo-info",
      iconClasses: "text-kumo-info",
      description: "Informational banner for general messages"
    },
    alert: {
      classes: "bg-kumo-warning-tint/15 border-kumo-warning/50 text-kumo-warning selection:bg-kumo-warning",
      iconClasses: "text-kumo-warning",
      description: "Warning banner for cautionary messages"
    },
    error: {
      classes: "bg-kumo-danger-tint/15 border-kumo-danger/50 text-kumo-danger selection:bg-kumo-danger",
      iconClasses: "text-kumo-danger",
      description: "Error banner for critical issues"
    }
  }
}, t = {
  variant: "default"
};
function x({
  variant: e = t.variant
} = {}) {
  return r(
    // Base styles (exported as KUMO_BANNER_BASE_STYLES for Figma plugin)
    E,
    // Apply variant styles from KUMO_BANNER_VARIANTS
    k(N.variant, e, t.variant).classes
  );
}
var R = /* @__PURE__ */ ((e) => (e[e.DEFAULT = 0] = "DEFAULT", e[e.ALERT = 1] = "ALERT", e[e.ERROR = 2] = "ERROR", e))(R || {});
const A = h(function({
  icon: a,
  title: o,
  description: s,
  action: m,
  children: b,
  text: p,
  variant: l = t.variant,
  className: d
}, u) {
  const f = k(N.variant, l, t.variant);
  if (o || s)
    return /* @__PURE__ */ i("div", { ref: u, className: r(x({ variant: l }), d), children: [
      a && /* @__PURE__ */ n(
        "span",
        {
          className: r(
            "shrink-0 flex items-center h-[1.375em]",
            f.iconClasses
          ),
          children: a
        }
      ),
      /* @__PURE__ */ i("div", { className: "flex min-w-0 flex-1 items-center justify-between gap-3", children: [
        /* @__PURE__ */ i("div", { className: "flex flex-col gap-0.5", children: [
          o && /* @__PURE__ */ n("p", { className: "font-medium leading-snug", children: o }),
          s && /* @__PURE__ */ n("div", { className: "text-sm leading-snug", children: g(s) ? s : /* @__PURE__ */ n("p", { children: s }) })
        ] }),
        m && /* @__PURE__ */ n("div", { className: "flex shrink-0 items-center gap-2", children: m })
      ] })
    ] });
  const c = b ?? p, v = g(c) ? c : /* @__PURE__ */ n("p", { children: c });
  return /* @__PURE__ */ i("div", { ref: u, className: r(x({ variant: l }), d), children: [
    a && /* @__PURE__ */ n("span", { className: r("shrink-0", f.iconClasses), children: a }),
    v
  ] });
});
A.displayName = "Banner";
export {
  A as B,
  R as a
};
