// @cloudflare/kumo/dist/chunks/collapsible-jvebgqfqljzokj8h.js
"use client";
import { jsx as o, jsxs as g } from "react/jsx-runtime";
import { CaretDownIcon as u } from "@phosphor-icons/react";
import { forwardRef as r } from "react";
import { c as t } from "./cn-ct4n7r74mh8y0f48.js";
import { aO as C, aP as i, aQ as n } from "./vendor-base-ui-epfrwb4nfbd4btaz.js";
function s({ className: e, ...l }) {
  return /* @__PURE__ */ o(C, { className: e, ...l });
}
s.displayName = "Collapsible.Root";
const p = r(
  ({ className: e, ...l }, a) => /* @__PURE__ */ o(
    n,
    {
      ref: a,
      className: t("cursor-pointer", e),
      ...l
    }
  )
);
p.displayName = "Collapsible.Trigger";
const m = r(
  ({ className: e, ...l }, a) => /* @__PURE__ */ o(
    i,
    {
      ref: a,
      className: e,
      ...l
    }
  )
);
m.displayName = "Collapsible.Panel";
const c = r(({ children: e, className: l }, a) => /* @__PURE__ */ g(
  n,
  {
    ref: a,
    className: t(
      // Defensive resets to prevent global button styles from polluting the trigger
      "bg-transparent border-none shadow-none p-0 m-0",
      // Base styles for the trigger
      "flex cursor-pointer items-center gap-1 text-sm text-kumo-link select-none",
      l
    ),
    children: [
      e,
      " ",
      /* @__PURE__ */ o(u, { className: "h-4 w-4 transition-transform [[data-panel-open]_&]:rotate-180" })
    ]
  }
));
c.displayName = "Collapsible.DefaultTrigger";
const b = r(({ children: e, className: l, ...a }, f) => /* @__PURE__ */ o(
  i,
  {
    ref: f,
    className: t("my-2 space-y-4 border-l-2 border-kumo-fill pl-4", l),
    ...a,
    children: e
  }
));
b.displayName = "Collapsible.DefaultPanel";
const T = Object.assign(s, {
  Root: s,
  Trigger: p,
  Panel: m,
  DefaultTrigger: c,
  DefaultPanel: b
});
export {
  T as C
};
