// @cloudflare/kumo/dist/chunks/dialog-n6uc2s99xwdn2pnb.js
"use client";
import { jsx as t, jsxs as R } from "react/jsx-runtime";
import { createContext as b, useContext as N } from "react";
import { L as P } from "./layer-card-eomdoafn3sfpih1d.js";
import { c as D } from "./cn-ct4n7r74mh8y0f48.js";
import { r as A } from "./resolve-variant-gw6eh7fa4st8ej7m.js";
import { u as k } from "./portal-provider-hwmkdmkpvct0cb76.js";
import { ap as h, aq as z, ar as n, as as c, at as g, au as d, av as p, aw as m, ax as u } from "./vendor-base-ui-epfrwb4nfbd4btaz.js";
const L = {
  size: {
    base: {
      classes: "sm:min-w-96",
      description: "Default dialog width"
    },
    sm: {
      classes: "min-w-72",
      description: "Small dialog for simple confirmations"
    },
    lg: {
      classes: "min-w-[32rem]",
      description: "Large dialog for complex content"
    },
    xl: {
      classes: "min-w-[48rem]",
      description: "Extra large dialog for detailed views"
    }
  }
}, r = {
  size: "base",
  role: "dialog"
}, f = b("dialog");
function s() {
  return N(f);
}
function O({
  size: a = r.size
} = {}) {
  return D(
    // Base styles
    "shadow-m ring ring-kumo-line fixed top-1/2 left-1/2 w-full sm:w-auto max-w-[calc(100vw-2rem)] sm:max-w-[calc(100vw-3rem)] -translate-x-1/2 -translate-y-1/2 overflow-hidden rounded-xl bg-kumo-base text-kumo-default duration-150 data-ending-style:scale-90 data-ending-style:opacity-0 data-starting-style:scale-90 data-starting-style:opacity-0",
    // Apply size from KUMO_DIALOG_VARIANTS
    A(L.size, a, r.size).classes
  );
}
function _({
  className: a,
  children: o,
  style: i,
  size: e = r.size,
  container: C
}) {
  const l = s(), v = k();
  return /* @__PURE__ */ R(l === "alertdialog" ? p : p, { container: C ?? v ?? void 0, children: [
    /* @__PURE__ */ t(l === "alertdialog" ? m : m, { className: "fixed inset-0 bg-kumo-recessed opacity-80 transition-all duration-150 data-ending-style:opacity-0 data-starting-style:opacity-0" }),
    /* @__PURE__ */ t(
      P,
      {
        render: /* @__PURE__ */ t(l === "alertdialog" ? u : u, {}),
        className: D(O({ size: e }), a),
        style: {
          transitionProperty: "scale, opacity",
          transitionTimingFunction: "var(--default-transition-timing-function)",
          "--tw-shadow": "0 20px 25px -5px rgb(0 0 0 / 0.03), 0 8px 10px -6px rgb(0 0 0 / 0.03)",
          ...i
        },
        children: o
      }
    )
  ] });
}
function x({
  children: a,
  role: o = r.role,
  ...i
}) {
  const e = o === "alertdialog" ? h : z;
  return /* @__PURE__ */ t(f.Provider, { value: o, children: /* @__PURE__ */ t(e, { ...i, children: a }) });
}
x.displayName = "Dialog.Root";
function y({ children: a, ...o }) {
  const e = s() === "alertdialog" ? n : n;
  return /* @__PURE__ */ t(e, { ...o, children: a });
}
y.displayName = "Dialog.Trigger";
function w({ className: a, ...o }) {
  const e = s() === "alertdialog" ? c : c;
  return /* @__PURE__ */ t(e, { className: a, ...o });
}
w.displayName = "Dialog.Title";
function B({ className: a, ...o }) {
  const e = s() === "alertdialog" ? g : g;
  return /* @__PURE__ */ t(e, { className: a, ...o });
}
B.displayName = "Dialog.Description";
function T({ children: a, ...o }) {
  const e = s() === "alertdialog" ? d : d;
  return /* @__PURE__ */ t(e, { ...o, children: a });
}
T.displayName = "Dialog.Close";
const q = Object.assign(_, {
  Root: x,
  Trigger: y,
  Title: w,
  Description: B,
  Close: T
});
export {
  q as D,
  x as a,
  y as b,
  w as c,
  B as d,
  T as e
};
