// @cloudflare/kumo/dist/chunks/empty-mmtirqntk6enx51o.js
"use client";
import { jsxs as i, jsx as e } from "react/jsx-runtime";
import { CheckIcon as d, CopyIcon as x } from "@phosphor-icons/react";
import { useState as f } from "react";
import { B as b } from "./button-n859eyw550yi2b9z.js";
import { c as a } from "./cn-ct4n7r74mh8y0f48.js";
import { r as h } from "./resolve-variant-gw6eh7fa4st8ej7m.js";
const g = {
  size: {
    sm: {
      classes: "px-6 py-8 gap-4",
      description: "Compact empty state for smaller containers"
    },
    base: {
      classes: "px-10 py-16 gap-6",
      description: "Default empty state size"
    },
    lg: {
      classes: "px-12 py-20 gap-8",
      description: "Large empty state for prominent placement"
    }
  }
}, l = {
  size: "base"
};
function k({
  size: t = l.size
} = {}) {
  return a(
    "flex w-full flex-col items-center rounded-xl border border-kumo-fill bg-kumo-control text-kumo-default",
    h(g.size, t, l.size).classes
  );
}
function T({
  icon: t,
  title: n,
  description: o,
  commandLine: s,
  contents: m,
  size: c = "base",
  className: p
}) {
  const [u, r] = f(!1);
  return /* @__PURE__ */ i("div", { className: a(k({ size: c }), p), children: [
    t,
    /* @__PURE__ */ e("h2", { className: "text-2xl font-semibold", children: n }),
    o && /* @__PURE__ */ e("p", { className: "max-w-140 text-center text-kumo-subtle", children: o }),
    s && /* @__PURE__ */ i(
      "div",
      {
        className: a(
          "group/cmd relative inline-flex h-10 max-w-8/10 transform-gpu items-center gap-2 rounded-lg font-mono shadow-sm",
          "bg-kumo-overlay pr-2 pl-3",
          "transition-all duration-300 hover:border-kumo-interact/80 hover:shadow-md",
          "border border-kumo-fill/60"
        ),
        children: [
          /* @__PURE__ */ e("span", { className: "text-xs text-kumo-inactive select-none", children: "$" }),
          /* @__PURE__ */ e("span", { className: "no-scrollbar overflow-scroll text-base whitespace-nowrap text-kumo-brand", children: s }),
          /* @__PURE__ */ e(
            b,
            {
              className: "group",
              size: "sm",
              variant: "ghost",
              shape: "square",
              "aria-label": "Copy command",
              onClick: async () => {
                r(!0), setTimeout(() => {
                  r(!1);
                }, 1e3), await navigator.clipboard.writeText(s);
              },
              children: u ? /* @__PURE__ */ e(d, { size: 16, className: "animate-bounce-in text-kumo-success" }) : /* @__PURE__ */ e(
                x,
                {
                  size: 16,
                  className: "text-kumo-inactive group-hover:text-kumo-brand"
                }
              )
            }
          )
        ]
      }
    ),
    m
  ] });
}
export {
  T as E,
  g as K,
  l as a,
  k as e
};
