// @cloudflare/kumo/dist/chunks/label-d14ibjmcbk1qmyrt.js
"use client";
import { jsx as e, jsxs as c, Fragment as f } from "react/jsx-runtime";
import { Info as u } from "@phosphor-icons/react";
import { c as a } from "./cn-ct4n7r74mh8y0f48.js";
import { B as p } from "./button-n859eyw550yi2b9z.js";
import { T as b } from "./tooltip-fjxy4s4l75hjxp1x.js";
const T = {
  // Label currently has no variant options but structure is ready for future additions
}, B = {};
function d(n = {}) {
  return a(
    // Base styles - when used standalone, apply text styling
    // When used inside Field, the parent FieldBase.Label provides these styles
    "m-0 text-base font-medium text-kumo-default"
  );
}
function o() {
  return a(
    // Content wrapper styles - always applied
    "inline-flex items-center gap-1"
  );
}
function x({
  children: n,
  showOptional: i = !1,
  tooltip: t,
  className: s,
  htmlFor: l,
  asContent: m = !1
}) {
  const r = /* @__PURE__ */ c(f, { children: [
    n,
    i && /* @__PURE__ */ e("span", { className: "font-normal text-kumo-subtle", children: "(optional)" }),
    t && /* @__PURE__ */ e(
      b,
      {
        content: t,
        render: /* @__PURE__ */ e(
          p,
          {
            variant: "ghost",
            size: "xs",
            shape: "square",
            "aria-label": "More information",
            children: /* @__PURE__ */ e(u, { className: "size-4" })
          }
        )
      }
    )
  ] });
  return m ? /* @__PURE__ */ e("span", { className: a(o(), s), children: r }) : /* @__PURE__ */ e(
    "label",
    {
      htmlFor: l,
      className: a(d(), o(), s),
      children: r
    }
  );
}
x.displayName = "Label";
export {
  T as K,
  x as L,
  o as a,
  B as b,
  d as l
};
