// @cloudflare/kumo/dist/chunks/layer-card-eomdoafn3sfpih1d.js
"use client";
import { jsx as s } from "react/jsx-runtime";
import { forwardRef as y, Children as p, isValidElement as C, Fragment as g } from "react";
import { c as n } from "./cn-ct4n7r74mh8y0f48.js";
import { aU as L, aV as S } from "./vendor-base-ui-epfrwb4nfbd4btaz.js";
const A = "overflow-hidden rounded-lg bg-kumo-base shadow-xs ring ring-kumo-line", R = "flex w-full flex-col overflow-hidden rounded-lg bg-kumo-elevated text-base ring ring-kumo-hairline", _ = "-my-2 flex items-center gap-2 bg-kumo-elevated p-4 text-base font-medium text-kumo-subtle", E = "relative flex flex-col gap-2 overflow-hidden rounded-lg bg-kumo-base p-4 pr-3 text-inherit no-underline ring ring-kumo-fill";
function x(r = {}) {
  return n(A);
}
function i(r) {
  return p.toArray(r).some((e) => C(e) ? e.type === o || e.type === t ? !0 : e.type === g ? i(e.props.children) : !1 : !1);
}
const d = y(function({ children: e, className: a, render: l, ...m }, u) {
  const f = i(e), c = {
    className: n(
      f ? R : x(),
      a
    )
  };
  return L({
    defaultTagName: "div",
    render: l,
    ref: u,
    props: S(c, m, { children: e })
  });
});
function t({
  children: r,
  className: e,
  ...a
}) {
  return /* @__PURE__ */ s("div", { className: n(_, e), ...a, children: r });
}
function o({ children: r, className: e, ...a }) {
  return /* @__PURE__ */ s("div", { className: n(E, e), ...a, children: r });
}
d.displayName = "LayerCard";
t.displayName = "LayerCard.Secondary";
o.displayName = "LayerCard.Primary";
const h = Object.assign(d, {
  Primary: o,
  Secondary: t
});
export {
  h as L
};
