// @cloudflare/kumo/dist/chunks/link-ihastr6a2dmo1so5.js
"use client";
import { jsxs as d, jsx as i } from "react/jsx-runtime";
import { forwardRef as k, useEffect as f } from "react";
import { c as r } from "./cn-ct4n7r74mh8y0f48.js";
import { r as L } from "./resolve-variant-gw6eh7fa4st8ej7m.js";
import { u as h } from "./link-provider-mn2voeohon7cj9o4.js";
import { aU as x, aV as v } from "./vendor-base-ui-epfrwb4nfbd4btaz.js";
const o = (n) => /* @__PURE__ */ d(
  "svg",
  {
    width: "1em",
    height: "1em",
    viewBox: "0 0 24 24",
    fill: "none",
    stroke: "currentColor",
    strokeLinecap: "round",
    strokeLinejoin: "round",
    "aria-hidden": "true",
    className: "link-external-icon",
    ...n,
    children: [
      /* @__PURE__ */ i("path", { d: "M9 4H8.8C7.11984 4 6.27976 4 5.63803 4.32698C5.07354 4.6146 4.6146 5.07354 4.32698 5.63803C4 6.27976 4 7.11984 4 8.8V15.2C4 16.8802 4 17.7202 4.32698 18.362C4.6146 18.9265 5.07354 19.3854 5.63803 19.673C6.27976 20 7.11984 20 8.8 20H15.2C16.8802 20 17.7202 20 18.362 19.673C18.9265 19.3854 19.3854 18.9265 19.673 18.362C20 17.7202 20 16.8802 20 15.2V15" }),
      /* @__PURE__ */ i("path", { d: "M14 4H20M20 4V10M20 4L11 13" })
    ]
  }
);
o.displayName = "Link.ExternalIcon";
const C = {
  variant: {
    inline: {
      classes: (
        // text-kumo-link provides defensive color that won't be overridden by global `a` styles
        "text-kumo-link underline underline-offset-[0.15em] decoration-[0.0625em] link-current transition-colors"
      ),
      description: "Inline text link that flows with content"
    },
    current: {
      classes: "text-current underline underline-offset-[0.15em] decoration-[0.0625em] link-current transition-colors",
      description: "Link that inherits color from parent text"
    },
    plain: {
      classes: (
        // text-kumo-link provides defensive color that won't be overridden by global `a` styles
        "text-kumo-link hover:text-kumo-link/70 transition-colors"
      ),
      description: "Link without underline decoration"
    }
  }
}, t = {
  variant: "inline"
};
function g({
  variant: n = t.variant
} = {}) {
  return r(L(C.variant, n, t.variant).classes);
}
const s = k(function({ className: a, variant: l = "inline", render: c, ...e }, u) {
  const m = h();
  process.env.NODE_ENV !== "production" && f(() => {
    "to" in e && e.to !== void 0 && console.warn(
      `[kumo] Link: The \`to\` prop is deprecated. Use \`href\` instead.

If your app uses a client-side router, configure a LinkProvider that
maps \`href\` to your router's navigation prop. See:
https://kumo.cfops.it/utilities/link-provider

Migration example:
  Before: <Link to="/page">…</Link>
  After:  <Link href="/page">…</Link>`
    );
  }, []);
  const p = {
    className: r(
      g({ variant: l }),
      "group/link inline-flex items-center gap-[0.1875em]"
    )
  };
  return x({
    render: c ?? /* @__PURE__ */ i(m, {}),
    ref: u,
    props: v(p, e, { className: a })
  });
});
s.displayName = "Link";
const _ = Object.assign(s, {
  ExternalIcon: o
});
export {
  C as K,
  _ as L,
  t as a,
  g as l
};
