// @cloudflare/kumo/dist/chunks/SankeyChart-i4vgcatj89zpgpl9.js
"use client";
import { jsx as w, jsxs as C } from "react/jsx-runtime";
import { forwardRef as ne, useRef as M, useEffect as D, useMemo as q, useCallback as re } from "react";
import { c as L } from "./cn-ct4n7r74mh8y0f48.js";
var Z = /* @__PURE__ */ ((e) => (e.Attention = "#FC574A", e.Warning = "#F8A054", e.Success = "#00A63E", e.Neutral = "#B9D6FF", e.Disabled = "#CBCBCB", e.Skeleton = "#DDDDDD", e))(Z || {}), ee = /* @__PURE__ */ ((e) => (e.Attention = "#FC574A", e.Warning = "#F8A054", e.Success = "#00A63E", e.Neutral = "#8EC5FF", e.Disabled = "#878787", e.Skeleton = "#5C5C5C", e))(ee || {});
const se = {
  blues: ["#E1EAF4", "#8EBCF6", "#4290F0", "#0E58B4", "#03254F"]
}, oe = {
  blues: ["#03254F", "#0E58B4", "#4290F0", "#A6BFDD", "#E1EAF4"]
}, U = [
  "#4290F0",
  "#F5B647",
  "#E8649D",
  "#8D58EE",
  "#50C3B6",
  "#D37536"
  /* Orange */
], X = [
  "#4290F0",
  "#EEB720",
  "#E8649D",
  "#8D58EE",
  "#50C3B6",
  "#D37536"
  /* Orange */
];
var j;
((e) => {
  function t(s, d = !1) {
    return d ? ee[s] : Z[s];
  }
  e.semantic = t;
  function c(s, d = !1) {
    return d ? X[s % X.length] : U[s % U.length];
  }
  e.categorical = c;
  function i(s, d = !1) {
    return d ? [...oe[s]] : [...se[s]];
  }
  e.sequential = i;
  function r(s, d = !1) {
    const h = {
      light: { primary: "#6B7280", secondary: "#9CA3AF" },
      dark: { primary: "#9CA3AF", secondary: "#6B7280" }
    };
    return d ? h.dark[s] : h.light[s];
  }
  e.text = r;
})(j || (j = {}));
const Y = (e) => {
  const { dangerousHtmlFormatter: t, ...c } = e;
  return {
    ...c,
    formatter: t
  };
}, ae = (e) => e.tooltip ? {
  ...e,
  tooltip: Array.isArray(e.tooltip) ? e.tooltip.map(Y) : Y(e.tooltip)
} : e, K = ne(function({
  echarts: t,
  options: c,
  optionUpdateBehavior: i,
  className: r,
  isDarkMode: s,
  height: d = 350,
  onEvents: h
}, p) {
  const f = M(null), N = M(null), S = M({}), E = M({}), F = M(/* @__PURE__ */ new Set());
  return D(() => {
    if (!f.current) return;
    const u = t.init(
      f.current,
      s ? "dark" : {
        color: s ? X : U
      }
    );
    return N.current = u, typeof p == "function" ? p(u) : p && (p.current = u), () => {
      for (const b of F.current) {
        const l = E.current[b];
        l && u.off(b, l);
      }
      F.current.clear(), typeof p == "function" ? p(null) : p && (p.current = null), N.current = null, u.dispose();
    };
  }, [f, s]), D(() => {
    const u = N.current;
    u && u.setOption(ae(c), {
      notMerge: !1,
      lazyUpdate: !0,
      ...i
    });
  }, [s, i, c]), D(() => {
    S.current = h ?? {};
  }, [h]), D(() => {
    const u = N.current;
    if (!u) return;
    const b = /* @__PURE__ */ new Set();
    for (const [l, A] of Object.entries(h ?? {}))
      typeof A == "function" && (b.add(l), E.current[l] || (E.current[l] = (B) => {
        S.current[l]?.(B);
      }), F.current.has(l) || u.on(l, E.current[l]));
    for (const l of F.current) {
      if (b.has(l)) continue;
      const A = E.current[l];
      A && u.off(l, A);
    }
    F.current = b;
  }, [t, s, h]), D(() => {
    const u = N.current, b = f.current;
    if (!u || !b) return;
    let l = !0;
    const A = new ResizeObserver(() => {
      if (l) {
        l = !1;
        return;
      }
      u.resize();
    });
    return A.observe(b), () => A.disconnect();
  }, []), /* @__PURE__ */ w(
    "div",
    {
      ref: f,
      className: L("w-full", r),
      style: { height: d },
      tabIndex: c.aria?.enabled ? 0 : void 0,
      role: c.aria?.enabled ? "img" : void 0
    }
  );
});
K.displayName = "Chart";
function be({
  echarts: e,
  type: t = "line",
  data: c,
  xAxisName: i,
  xAxisTickCount: r,
  xAxisTickFormat: s,
  yAxisTickFormat: d,
  yAxisTickLabelFormat: h,
  yAxisName: p,
  yAxisTickCount: f,
  tooltipValueFormat: N,
  onTimeRangeChange: S,
  height: E = 350,
  incomplete: F,
  isDarkMode: u,
  gradient: b,
  loading: l,
  ariaDescription: A
}) {
  const B = M(null), R = F?.before, H = F?.after, P = q(() => {
    const g = [], I = t === "bar" ? { type: "bar", stack: "total" } : { type: "line", showSymbol: !1 };
    for (const o of c) {
      const n = R && t === "line" ? o.data.filter((O) => O[0] <= R) : [], a = H && t === "line" ? o.data.filter((O) => O[0] >= H) : [], m = n.length > 0 || a.length > 0 ? o.data.slice(
        Math.max(0, n.length - 1),
        Math.max(0, o.data.length - a.length + 1)
      ) : o.data, x = b && t === "line" ? {
        color: new e.graphic.LinearGradient(0, 0, 0, 1, [
          { offset: 0, color: J(o.color, 0.4) },
          { offset: 1, color: J(o.color, 0) }
        ])
      } : void 0;
      g.push({
        data: m,
        color: o.color,
        name: o.name,
        emphasis: { focus: "series" },
        ...x ? { areaStyle: x } : {},
        ...I
      });
      const $ = {
        color: o.color,
        name: o.name,
        type: "line",
        lineStyle: { type: "dashed" },
        showSymbol: !1,
        emphasis: { focus: "series" }
      };
      n.length > 0 && g.push({
        ...$,
        data: n
      }), a.length > 0 && g.push({
        ...$,
        data: a
      });
    }
    return {
      aria: {
        enabled: !0,
        ...A && { label: { description: A } }
      },
      brush: {
        xAxisIndex: "all",
        brushType: "lineX",
        brushMode: "single",
        outOfBrush: {
          colorAlpha: 0.3
        },
        brushStyle: {
          borderWidth: 1,
          color: "rgba(120,140,180,0.3)",
          borderColor: "rgba(120,140,180,0.8)"
        }
      },
      tooltip: {
        trigger: "axis",
        appendTo: "body",
        axisPointer: { type: "shadow" },
        dangerousHtmlFormatter: (o) => {
          const n = Array.isArray(o) ? o : [o], a = /* @__PURE__ */ new Set(), m = n.filter((T) => a.has(T.seriesName) ? !1 : (a.add(T.seriesName), !0)), x = m[0], $ = x?.value?.[0] ?? x?.axisValue, O = $ != null ? `<div style="font-weight:600;margin-bottom:4px;">${e.format.encodeHTML(le($))}</div>` : "", _ = m.map((T) => {
            const k = T?.value?.[1], G = N ?? h, te = G ? e.format.encodeHTML(String(G(k))) : e.format.encodeHTML(String(k));
            return `${T.marker} ${e.format.encodeHTML(T.seriesName)}: <strong>${te}</strong>`;
          }).join("<br/>");
          return `${O}${_}`;
        }
      },
      backgroundColor: "transparent",
      toolbox: { show: !1 },
      xAxis: {
        name: i,
        nameLocation: "middle",
        nameGap: 30,
        type: "time",
        splitLine: {
          show: !1
        },
        axisLine: { show: !1 },
        splitNumber: r ?? 5,
        ...s && {
          axisLabel: {
            formatter: (o) => s(o)
          }
        }
      },
      yAxis: {
        name: p,
        nameLocation: "middle",
        nameGap: 40,
        type: "value",
        axisTick: { show: !0 },
        axisLabel: {
          margin: 15,
          ...d && {
            formatter: (o) => d(o)
          }
        },
        splitLine: {
          show: !0,
          lineStyle: { type: "dashed", width: 1 }
        },
        splitNumber: f
      },
      grid: {
        left: p ? 30 : 24,
        right: 24,
        top: 24,
        bottom: i ? 30 : 24
      },
      series: g
    };
  }, [
    c,
    i,
    r,
    s,
    d,
    h,
    p,
    f,
    N,
    R,
    H,
    t,
    b,
    e,
    A
  ]), y = q(() => S ? {
    brushend: (g) => {
      const I = g.areas[0].coordRange;
      S(I[0], I[1]), B.current?.dispatchAction({ type: "brush", areas: [] });
    }
  } : {}, [S]), v = !!S;
  return D(() => {
    const g = B.current;
    if (g && v)
      return g.dispatchAction({
        type: "takeGlobalCursor",
        key: "brush",
        brushOption: {
          brushType: "lineX",
          brushMode: "single"
        }
      }), () => {
        g.dispatchAction({
          type: "takeGlobalCursor",
          key: "brush",
          brushOption: {
            brushType: !1
          }
        });
      };
  }, [B, v, l]), /* @__PURE__ */ C("div", { className: "relative w-full", style: { height: E }, children: [
    l && /* @__PURE__ */ w(ce, { height: E, isDarkMode: u }),
    !l && /* @__PURE__ */ w(
      K,
      {
        echarts: e,
        ref: B,
        options: P,
        height: E,
        isDarkMode: u,
        onEvents: y
      }
    )
  ] });
}
function ce({
  height: e,
  isDarkMode: t
}) {
  const c = e / 2, i = Math.min(e * 0.12, 28), r = 400, s = 120, d = [];
  for (let f = 0; f <= s; f++) {
    const N = -r + f / s * r * 3, S = c + Math.sin(f / s * 2 * Math.PI * 3) * i;
    d.push(`${f === 0 ? "M" : "L"}${N.toFixed(2)},${S.toFixed(2)}`);
  }
  const h = d.join(" "), p = t ? "rgba(255,255,255,0.5)" : "rgba(0,0,0,0.2)";
  return /* @__PURE__ */ w(
    "div",
    {
      "aria-hidden": "true",
      className: "absolute inset-0 overflow-hidden",
      style: { height: e },
      children: /* @__PURE__ */ w(
        "svg",
        {
          width: "100%",
          height: e,
          viewBox: `0 0 ${r} ${e}`,
          preserveAspectRatio: "none",
          className: "w-full animate-pulse",
          children: /* @__PURE__ */ w(
            "path",
            {
              d: h,
              fill: "none",
              stroke: p,
              strokeWidth: "2",
              style: {
                animation: "kumo-chart-wave 2.4s linear infinite",
                transformOrigin: "0 0"
              }
            }
          )
        }
      )
    }
  );
}
function J(e, t) {
  const c = Math.max(0, Math.min(1, t)), i = e.match(
    /rgba?\(\s*([\d.]+)\s*,\s*([\d.]+)\s*,\s*([\d.]+)/i
  );
  if (i)
    return `rgba(${i[1]}, ${i[2]}, ${i[3]}, ${c})`;
  let r = e.replace(/^#/, "");
  r.length === 3 && (r = r[0] + r[0] + r[1] + r[1] + r[2] + r[2]), r.length === 8 && (r = r.slice(0, 6));
  const s = parseInt(r.slice(0, 2), 16), d = parseInt(r.slice(2, 4), 16), h = parseInt(r.slice(4, 6), 16);
  return `rgba(${s}, ${d}, ${h}, ${c})`;
}
function z(e) {
  return e.toString().padStart(2, "0");
}
function le(e) {
  const t = new Date(e);
  return `${t.getFullYear()}-${z(t.getMonth() + 1)}-${z(t.getDate())} ${z(t.getHours())}:${z(t.getMinutes())}:${z(t.getSeconds())}`;
}
function ie({ color: e, value: t, name: c, unit: i, inactive: r }) {
  return /* @__PURE__ */ C("div", { className: "inline-flex flex-col gap-2 min-w-42 py-2", children: [
    /* @__PURE__ */ C("div", { className: "flex items-center gap-2", children: [
      /* @__PURE__ */ w(
        "span",
        {
          className: L("size-2 rounded-full inline-block", {
            "opacity-50": r
          }),
          style: { backgroundColor: e }
        }
      ),
      /* @__PURE__ */ w("span", { className: L("text-xs", { "opacity-50": r }), children: c })
    ] }),
    /* @__PURE__ */ C("div", { className: "flex items-baseline gap-0.5", children: [
      /* @__PURE__ */ w(
        "span",
        {
          className: L("text-lg font-medium leading-none", {
            "opacity-50": r
          }),
          children: t
        }
      ),
      i && /* @__PURE__ */ w(
        "span",
        {
          className: L("text-xs text-kumo-subtle leading-none", {
            "opacity-50": r
          }),
          children: i
        }
      )
    ] })
  ] });
}
function ue({ color: e, value: t, name: c, inactive: i }) {
  return /* @__PURE__ */ C("div", { className: "inline-flex items-center gap-2", children: [
    /* @__PURE__ */ w(
      "span",
      {
        className: L("size-2 rounded-full inline-block", {
          "opacity-50": i
        }),
        style: { backgroundColor: e }
      }
    ),
    /* @__PURE__ */ w("span", { className: L("text-xs", { "opacity-50": i }), children: c }),
    /* @__PURE__ */ w("span", { className: L("text-xs font-medium", { "opacity-50": i }), children: t })
  ] });
}
const ye = {
  SmallItem: ue,
  LargeItem: ie
}, de = (e) => e.toLocaleString();
function fe(e) {
  return typeof e == "object" && e !== null;
}
const W = (e) => e.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#39;"), Q = (e) => e.replace(/[{}|]/g, (t) => `\\${t}`), V = (e) => {
  const t = "#666";
  return !e || typeof e != "string" ? t : /^#(?:[0-9a-f]{3}|[0-9a-f]{6}|[0-9a-f]{8})$/i.test(e) || /^rgba?\(\s*\d{1,3}\s*,\s*\d{1,3}\s*,\s*\d{1,3}\s*(?:,\s*[\d.]+\s*)?\)$/i.test(
    e
  ) || /^hsla?\(\s*\d{1,3}\s*,\s*\d{1,3}%\s*,\s*\d{1,3}%\s*(?:,\s*[\d.]+\s*)?\)$/i.test(
    e
  ) || /^[a-z]{3,20}$/i.test(e) ? e : t;
};
function pe({
  echarts: e,
  nodes: t,
  links: c,
  height: i = 400,
  nodeWidth: r = 8,
  nodePadding: s = 10,
  showTooltip: d = !0,
  showNodeValues: h,
  formatValue: p = de,
  tooltipFormatter: f,
  defaultNodeColor: N,
  linkColor: S = "gradient",
  linkOpacity: E = 0.5,
  className: F,
  isDarkMode: u,
  onNodeClick: b,
  onLinkClick: l
}) {
  const A = t.some((y) => y.value !== void 0), B = h ?? A, R = q(() => {
    const y = j.text("primary", u), v = j.text("secondary", u), g = new Map(t.map((n) => [n.name, n])), I = t.map((n, a) => ({
      name: n.name,
      value: n.value,
      itemStyle: {
        color: n.color ?? N ?? j.categorical(a, u)
      }
    })), o = c.map((n) => ({
      source: t[n.source]?.name ?? "",
      target: t[n.target]?.name ?? "",
      value: n.value
    }));
    return {
      animation: !0,
      animationDuration: 500,
      animationDurationUpdate: 300,
      animationEasingUpdate: "cubicInOut",
      tooltip: d ? {
        trigger: "item",
        triggerOn: "mousemove",
        dangerousHtmlFormatter: (n) => {
          if (!fe(n)) return "";
          if (n.dataType === "node" && n.name) {
            const a = g.get(n.name), m = V(
              a?.color ?? n.color ?? "#666"
            );
            if (f)
              return f({
                type: "node",
                name: n.name,
                node: a,
                color: m
              });
            const x = W(n.name);
            return `<div style="display:flex;align-items:center;gap:6px;"><span style="display:inline-block;width:10px;height:10px;border-radius:50%;background:${m}"></span><strong>${x}</strong></div>`;
          }
          if (n.dataType === "edge" && n.data) {
            const { source: a, target: m, value: x } = n.data;
            if (f)
              return f({
                type: "link",
                name: `${a} → ${m}`,
                link: {
                  source: a ?? "",
                  target: m ?? "",
                  value: x ?? 0
                }
              });
            const $ = g.get(a ?? ""), O = g.get(m ?? ""), _ = V($?.color ?? "#666"), T = V(O?.color ?? "#666"), k = W(a ?? ""), G = W(m ?? "");
            return `<div style="display:flex;align-items:center;gap:6px;margin-bottom:4px;">
                  <span style="display:inline-block;width:10px;height:10px;border-radius:50%;background:${_}"></span>
                  <strong>${k}</strong>
                  <span style="color:${v}">→</span>
                  <span style="display:inline-block;width:10px;height:10px;border-radius:50%;background:${T}"></span>
                  <strong>${G}</strong>
                </div>
                <strong>${x !== void 0 ? W(p(x)) : ""}</strong>`;
          }
          return "";
        }
      } : void 0,
      series: [
        {
          type: "sankey",
          data: I,
          links: o,
          draggable: !1,
          emphasis: {
            focus: "adjacency"
          },
          nodeWidth: r,
          nodeGap: s,
          lineStyle: {
            color: S === "gradient" ? "source" : "#d1d5db",
            opacity: S === "gradient" ? E : 0.4,
            curveness: 0.5
          },
          label: {
            show: !0,
            color: y,
            fontSize: 12,
            formatter: B ? (n) => {
              const a = n.name ?? "", m = g.get(a), x = Q(a);
              return m?.value !== void 0 ? `{value|${Q(p(m.value))}}
{name|${x}}` : x;
            } : void 0,
            rich: B ? {
              value: {
                fontSize: 11,
                color: y,
                lineHeight: 16
              },
              name: {
                fontSize: 12,
                color: y,
                fontWeight: 700
              }
            } : void 0
          }
        }
      ]
    };
  }, [
    t,
    c,
    d,
    r,
    s,
    N,
    u,
    S,
    E,
    B,
    p,
    f
  ]), H = re(
    (y) => {
      if (y.dataType === "node" && b && y.name) {
        const v = t.findIndex((o) => o.name === y.name), I = {
          ...v >= 0 ? t[v] : null,
          name: y.name
        };
        b(I);
      } else if (y.dataType === "edge" && l && y.data) {
        const v = y.data, g = typeof v == "object" && v !== null && "source" in v ? String(v.source) : "", I = typeof v == "object" && v !== null && "target" in v ? String(v.target) : "", o = t.findIndex(($) => $.name === g), n = t.findIndex(($) => $.name === I);
        if (o === -1 || n === -1) return;
        const a = y.value, m = typeof a == "number" ? a : Array.isArray(a) && typeof a[0] == "number" ? a[0] : 0, x = c.find(
          ($) => $.source === o && $.target === n
        );
        l({
          ...x,
          source: o,
          target: n,
          value: m
        });
      }
    },
    [t, c, b, l]
  ), P = q(
    () => ({
      click: H
    }),
    [H]
  );
  return /* @__PURE__ */ w(
    K,
    {
      echarts: e,
      options: R,
      className: F,
      isDarkMode: u,
      height: i,
      onEvents: P
    }
  );
}
pe.displayName = "SankeyChart";
export {
  K as C,
  pe as S,
  be as T,
  ye as a,
  j as b
};
