// @cloudflare/kumo/dist/chunks/table-of-contents-fzyv7uhnnyr13dqu.js
"use client";
import { jsx as e, jsxs as f } from "react/jsx-runtime";
import { forwardRef as o, isValidElement as k, cloneElement as h } from "react";
import { c as n } from "./cn-ct4n7r74mh8y0f48.js";
const m = {
  state: {
    default: {
      classes: "text-kumo-subtle hover:border-kumo-line hover:text-kumo-default hover:font-medium",
      description: "Inactive section link"
    },
    active: {
      classes: "border-kumo-brand font-medium text-kumo-default",
      description: "Currently visible / active section"
    }
  }
}, A = {
  state: "default"
}, p = "block w-full truncate border-l-2 border-transparent py-0.5 pl-4 text-sm text-left no-underline", T = o(
  ({ className: t, "aria-label": s = "Table of contents", ...l }, a) => /* @__PURE__ */ e("nav", { ref: a, "aria-label": s, className: t, ...l })
), N = o(({ className: t, ...s }, l) => /* @__PURE__ */ e(
  "p",
  {
    ref: l,
    className: n(
      "mb-3 text-xs font-semibold tracking-wide text-kumo-subtle uppercase",
      t
    ),
    ...s
  }
)), x = o(({ className: t, ...s }, l) => /* @__PURE__ */ e(
  "ul",
  {
    ref: l,
    className: n(
      "flex flex-col gap-2 border-l-2 border-kumo-hairline",
      t
    ),
    ...s
  }
)), C = o(({ active: t = !1, className: s, children: l, render: a, ...i }, c) => {
  const r = t ? m.state.active.classes : m.state.default.classes, d = n(p, r, s), u = {
    ref: c,
    "aria-current": t ? "true" : void 0,
    className: d,
    children: /* @__PURE__ */ e("span", { className: "block min-w-0 leading-5", children: l }),
    ...i
  };
  return a && k(a) ? /* @__PURE__ */ e("li", { className: "-ml-0.5", children: h(a, u) }) : /* @__PURE__ */ e("li", { className: "-ml-0.5", children: /* @__PURE__ */ e("a", { ...u }) });
}), b = "flex flex-col gap-2 border-l-2 border-kumo-hairline [&>li>a]:pl-7 [&>li>button]:pl-7", O = o(({ label: t, href: s, active: l = !1, className: a, children: i, ...c }, r) => {
  if (s) {
    const d = l ? m.state.active.classes : m.state.default.classes;
    return /* @__PURE__ */ f(
      "li",
      {
        ref: r,
        className: n("-ml-0.5 flex flex-col gap-2", a),
        ...c,
        children: [
          /* @__PURE__ */ e(
            "a",
            {
              href: s,
              "aria-current": l ? "true" : void 0,
              className: n(p, d),
              children: /* @__PURE__ */ e("span", { className: "block min-w-0 leading-5", children: t })
            }
          ),
          /* @__PURE__ */ e("ul", { className: n(b), children: i })
        ]
      }
    );
  }
  return /* @__PURE__ */ f(
    "li",
    {
      ref: r,
      className: n("-ml-0.5 flex flex-col gap-2", a),
      ...c,
      children: [
        /* @__PURE__ */ e("p", { className: "py-0.5 pl-4 text-sm leading-5 font-medium text-kumo-subtle", children: t }),
        /* @__PURE__ */ e("ul", { className: n(b), children: i })
      ]
    }
  );
});
T.displayName = "TableOfContents";
N.displayName = "TableOfContents.Title";
x.displayName = "TableOfContents.List";
C.displayName = "TableOfContents.Item";
O.displayName = "TableOfContents.Group";
const S = Object.assign(T, {
  Title: N,
  List: x,
  Item: C,
  Group: O
});
export {
  m as K,
  S as T,
  A as a
};
