// @cloudflare/kumo/dist/chunks/vendor-utils-ixnhxmehwihk5tr7.js
"use client";
import A from "react";
var $ = (e, n, t) => {
  if (n.length === 1 && n[0] === t) {
    let c = !1;
    try {
      const a = {};
      e(a) === a && (c = !0);
    } catch {
    }
    if (c) {
      let a;
      try {
        throw new Error();
      } catch (d) {
        ({ stack: a } = d);
      }
      console.warn(
        `The result function returned its own inputs without modification. e.g
\`createSelector([state => state.todos], todos => todos)\`
This could lead to inefficient memoization and unnecessary re-renders.
Ensure transformation logic is in the result function, and extraction logic is in the input selectors.`,
        { stack: a }
      );
    }
  }
}, P = (e, n, t) => {
  const { memoize: c, memoizeOptions: a } = n, { inputSelectorResults: d, inputSelectorResultsCopy: y } = e, s = c(() => ({}), ...a);
  if (!(s.apply(null, d) === s.apply(null, y))) {
    let o;
    try {
      throw new Error();
    } catch (i) {
      ({ stack: o } = i);
    }
    console.warn(
      `An input selector returned a different result when passed same arguments.
This means your output selector will likely run more frequently than intended.
Avoid returning a new reference inside your input selector, e.g.
\`createSelector([state => state.todos.map(todo => todo.id)], todoIds => todoIds.length)\``,
      {
        arguments: t,
        firstInputs: d,
        secondInputs: y,
        stack: o
      }
    );
  }
}, Y = {
  inputStabilityCheck: "once",
  identityFunctionCheck: "once"
}, L = /* @__PURE__ */ Symbol("NOT_FOUND");
function J(e, n = `expected a function, instead received ${typeof e}`) {
  if (typeof e != "function")
    throw new TypeError(n);
}
function Q(e, n = `expected an object, instead received ${typeof e}`) {
  if (typeof e != "object")
    throw new TypeError(n);
}
function X(e, n = "expected all items to be functions, instead received the following types: ") {
  if (!e.every((t) => typeof t == "function")) {
    const t = e.map(
      (c) => typeof c == "function" ? `function ${c.name || "unnamed"}()` : typeof c
    ).join(", ");
    throw new TypeError(`${n}[${t}]`);
  }
}
var N = (e) => Array.isArray(e) ? e : [e];
function Z(e) {
  const n = Array.isArray(e[0]) ? e[0] : e;
  return X(
    n,
    "createSelector expects all input-selectors to be functions, but received the following types: "
  ), n;
}
function z(e, n) {
  const t = [], { length: c } = e;
  for (let a = 0; a < c; a++)
    t.push(e[a].apply(null, n));
  return t;
}
var ee = (e, n) => {
  const { identityFunctionCheck: t, inputStabilityCheck: c } = {
    ...Y,
    ...n
  };
  return {
    identityFunctionCheck: {
      shouldRun: t === "always" || t === "once" && e,
      run: $
    },
    inputStabilityCheck: {
      shouldRun: c === "always" || c === "once" && e,
      run: P
    }
  };
};
function te(e) {
  let n;
  return {
    get(t) {
      return n && e(n.key, t) ? n.value : L;
    },
    put(t, c) {
      n = { key: t, value: c };
    },
    getEntries() {
      return n ? [n] : [];
    },
    clear() {
      n = void 0;
    }
  };
}
function ne(e, n) {
  let t = [];
  function c(s) {
    const v = t.findIndex((o) => n(s, o.key));
    if (v > -1) {
      const o = t[v];
      return v > 0 && (t.splice(v, 1), t.unshift(o)), o.value;
    }
    return L;
  }
  function a(s, v) {
    c(s) === L && (t.unshift({ key: s, value: v }), t.length > e && t.pop());
  }
  function d() {
    return t;
  }
  function y() {
    t = [];
  }
  return { get: c, put: a, getEntries: d, clear: y };
}
var re = (e, n) => e === n;
function oe(e) {
  return function(t, c) {
    if (t === null || c === null || t.length !== c.length)
      return !1;
    const { length: a } = t;
    for (let d = 0; d < a; d++)
      if (!e(t[d], c[d]))
        return !1;
    return !0;
  };
}
function he(e, n) {
  const t = typeof n == "object" ? n : { equalityCheck: n }, {
    equalityCheck: c = re,
    maxSize: a = 1,
    resultEqualityCheck: d
  } = t, y = oe(c);
  let s = 0;
  const v = a <= 1 ? te(y) : ne(a, y);
  function o() {
    let i = v.get(arguments);
    if (i === L) {
      if (i = e.apply(null, arguments), s++, d) {
        const p = v.getEntries().find(
          (r) => d(r.value, i)
        );
        p && (i = p.value, s !== 0 && s--);
      }
      v.put(arguments, i);
    }
    return i;
  }
  return o.clearCache = () => {
    v.clear(), o.resetResultsCount();
  }, o.resultsCount = () => s, o.resetResultsCount = () => {
    s = 0;
  }, o;
}
var ue = class {
  constructor(e) {
    this.value = e;
  }
  deref() {
    return this.value;
  }
}, ie = typeof WeakRef < "u" ? WeakRef : ue, se = 0, G = 1;
function w() {
  return {
    s: se,
    v: void 0,
    o: null,
    p: null
  };
}
function F(e, n = {}) {
  let t = w();
  const { resultEqualityCheck: c } = n;
  let a, d = 0;
  function y() {
    let s = t;
    const { length: v } = arguments;
    for (let l = 0, p = v; l < p; l++) {
      const r = arguments[l];
      if (typeof r == "function" || typeof r == "object" && r !== null) {
        let u = s.o;
        u === null && (s.o = u = /* @__PURE__ */ new WeakMap());
        const f = u.get(r);
        f === void 0 ? (s = w(), u.set(r, s)) : s = f;
      } else {
        let u = s.p;
        u === null && (s.p = u = /* @__PURE__ */ new Map());
        const f = u.get(r);
        f === void 0 ? (s = w(), u.set(r, s)) : s = f;
      }
    }
    const o = s;
    let i;
    if (s.s === G)
      i = s.v;
    else if (i = e.apply(null, arguments), d++, c) {
      const l = a?.deref?.() ?? a;
      l != null && c(l, i) && (i = l, d !== 0 && d--), a = typeof i == "object" && i !== null || typeof i == "function" ? new ie(i) : i;
    }
    return o.s = G, o.v = i, i;
  }
  return y.clearCache = () => {
    t = w(), y.resetResultsCount();
  }, y.resultsCount = () => d, y.resetResultsCount = () => {
    d = 0;
  }, y;
}
function ce(e, ...n) {
  const t = typeof e == "function" ? {
    memoize: e,
    memoizeOptions: n
  } : e, c = (...a) => {
    let d = 0, y = 0, s, v = {}, o = a.pop();
    typeof o == "object" && (v = o, o = a.pop()), J(
      o,
      `createSelector expects an output function after the inputs, but received: [${typeof o}]`
    );
    const i = {
      ...t,
      ...v
    }, {
      memoize: l,
      memoizeOptions: p = [],
      argsMemoize: r = F,
      argsMemoizeOptions: u = [],
      devModeChecks: f = {}
    } = i, m = N(p), S = N(u), E = Z(a), R = l(function() {
      return d++, o.apply(
        null,
        arguments
      );
    }, ...m);
    let O = !0;
    const C = r(function() {
      y++;
      const _ = z(
        E,
        arguments
      );
      if (s = R.apply(null, _), process.env.NODE_ENV !== "production") {
        const { identityFunctionCheck: g, inputStabilityCheck: j } = ee(O, f);
        if (g.shouldRun && g.run(
          o,
          _,
          s
        ), j.shouldRun) {
          const U = z(
            E,
            arguments
          );
          j.run(
            { inputSelectorResults: _, inputSelectorResultsCopy: U },
            { memoize: l, memoizeOptions: m },
            arguments
          );
        }
        O && (O = !1);
      }
      return s;
    }, ...S);
    return Object.assign(C, {
      resultFunc: o,
      memoizedResultFunc: R,
      dependencies: E,
      dependencyRecomputations: () => y,
      resetDependencyRecomputations: () => {
        y = 0;
      },
      lastResult: () => s,
      recomputations: () => d,
      resetRecomputations: () => {
        d = 0;
      },
      memoize: l,
      argsMemoize: r
    });
  };
  return Object.assign(c, {
    withTypes: () => c
  }), c;
}
var ae = /* @__PURE__ */ ce(F), le = Object.assign(
  (e, n = ae) => {
    Q(
      e,
      `createStructuredSelector expects first argument to be an object where each property is a selector, instead received a ${typeof e}`
    );
    const t = Object.keys(e), c = t.map(
      (d) => e[d]
    );
    return n(
      c,
      (...d) => d.reduce((y, s, v) => (y[t[v]] = s, y), {})
    );
  },
  { withTypes: () => le }
), b = { exports: {} }, D = {};
var x;
function fe() {
  if (x) return D;
  x = 1;
  var e = A;
  function n(l, p) {
    return l === p && (l !== 0 || 1 / l === 1 / p) || l !== l && p !== p;
  }
  var t = typeof Object.is == "function" ? Object.is : n, c = e.useState, a = e.useEffect, d = e.useLayoutEffect, y = e.useDebugValue;
  function s(l, p) {
    var r = p(), u = c({ inst: { value: r, getSnapshot: p } }), f = u[0].inst, m = u[1];
    return d(
      function() {
        f.value = r, f.getSnapshot = p, v(f) && m({ inst: f });
      },
      [l, r, p]
    ), a(
      function() {
        return v(f) && m({ inst: f }), l(function() {
          v(f) && m({ inst: f });
        });
      },
      [l]
    ), y(r), r;
  }
  function v(l) {
    var p = l.getSnapshot;
    l = l.value;
    try {
      var r = p();
      return !t(l, r);
    } catch {
      return !0;
    }
  }
  function o(l, p) {
    return p();
  }
  var i = typeof window > "u" || typeof window.document > "u" || typeof window.document.createElement > "u" ? o : s;
  return D.useSyncExternalStore = e.useSyncExternalStore !== void 0 ? e.useSyncExternalStore : i, D;
}
var I = {};
var W;
function de() {
  return W || (W = 1, process.env.NODE_ENV !== "production" && (function() {
    function e(r, u) {
      return r === u && (r !== 0 || 1 / r === 1 / u) || r !== r && u !== u;
    }
    function n(r, u) {
      i || a.startTransition === void 0 || (i = !0, console.error(
        "You are using an outdated, pre-release alpha of React 18 that does not support useSyncExternalStore. The use-sync-external-store shim will not work correctly. Upgrade to a newer pre-release."
      ));
      var f = u();
      if (!l) {
        var m = u();
        d(f, m) || (console.error(
          "The result of getSnapshot should be cached to avoid an infinite loop"
        ), l = !0);
      }
      m = y({
        inst: { value: f, getSnapshot: u }
      });
      var S = m[0].inst, E = m[1];
      return v(
        function() {
          S.value = f, S.getSnapshot = u, t(S) && E({ inst: S });
        },
        [r, f, u]
      ), s(
        function() {
          return t(S) && E({ inst: S }), r(function() {
            t(S) && E({ inst: S });
          });
        },
        [r]
      ), o(f), f;
    }
    function t(r) {
      var u = r.getSnapshot;
      r = r.value;
      try {
        var f = u();
        return !d(r, f);
      } catch {
        return !0;
      }
    }
    function c(r, u) {
      return u();
    }
    typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ < "u" && typeof __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStart == "function" && __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStart(Error());
    var a = A, d = typeof Object.is == "function" ? Object.is : e, y = a.useState, s = a.useEffect, v = a.useLayoutEffect, o = a.useDebugValue, i = !1, l = !1, p = typeof window > "u" || typeof window.document > "u" || typeof window.document.createElement > "u" ? c : n;
    I.useSyncExternalStore = a.useSyncExternalStore !== void 0 ? a.useSyncExternalStore : p, typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ < "u" && typeof __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStop == "function" && __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStop(Error());
  })()), I;
}
var K;
function M() {
  return K || (K = 1, process.env.NODE_ENV === "production" ? b.exports = fe() : b.exports = de()), b.exports;
}
var _e = M(), T = { exports: {} }, k = {};
var B;
function pe() {
  if (B) return k;
  B = 1;
  var e = A, n = M();
  function t(o, i) {
    return o === i && (o !== 0 || 1 / o === 1 / i) || o !== o && i !== i;
  }
  var c = typeof Object.is == "function" ? Object.is : t, a = n.useSyncExternalStore, d = e.useRef, y = e.useEffect, s = e.useMemo, v = e.useDebugValue;
  return k.useSyncExternalStoreWithSelector = function(o, i, l, p, r) {
    var u = d(null);
    if (u.current === null) {
      var f = { hasValue: !1, value: null };
      u.current = f;
    } else f = u.current;
    u = s(
      function() {
        function S(h) {
          if (!E) {
            if (E = !0, R = h, h = p(h), r !== void 0 && f.hasValue) {
              var _ = f.value;
              if (r(_, h))
                return O = _;
            }
            return O = h;
          }
          if (_ = O, c(R, h)) return _;
          var g = p(h);
          return r !== void 0 && r(_, g) ? (R = h, _) : (R = h, O = g);
        }
        var E = !1, R, O, C = l === void 0 ? null : l;
        return [
          function() {
            return S(i());
          },
          C === null ? void 0 : function() {
            return S(C());
          }
        ];
      },
      [i, l, p, r]
    );
    var m = a(o, u[0], u[1]);
    return y(
      function() {
        f.hasValue = !0, f.value = m;
      },
      [m]
    ), v(m), m;
  }, k;
}
var V = {};
var H;
function ve() {
  return H || (H = 1, process.env.NODE_ENV !== "production" && (function() {
    function e(o, i) {
      return o === i && (o !== 0 || 1 / o === 1 / i) || o !== o && i !== i;
    }
    typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ < "u" && typeof __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStart == "function" && __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStart(Error());
    var n = A, t = M(), c = typeof Object.is == "function" ? Object.is : e, a = t.useSyncExternalStore, d = n.useRef, y = n.useEffect, s = n.useMemo, v = n.useDebugValue;
    V.useSyncExternalStoreWithSelector = function(o, i, l, p, r) {
      var u = d(null);
      if (u.current === null) {
        var f = { hasValue: !1, value: null };
        u.current = f;
      } else f = u.current;
      u = s(
        function() {
          function S(h) {
            if (!E) {
              if (E = !0, R = h, h = p(h), r !== void 0 && f.hasValue) {
                var _ = f.value;
                if (r(_, h))
                  return O = _;
              }
              return O = h;
            }
            if (_ = O, c(R, h))
              return _;
            var g = p(h);
            return r !== void 0 && r(_, g) ? (R = h, _) : (R = h, O = g);
          }
          var E = !1, R, O, C = l === void 0 ? null : l;
          return [
            function() {
              return S(i());
            },
            C === null ? void 0 : function() {
              return S(C());
            }
          ];
        },
        [i, l, p, r]
      );
      var m = a(o, u[0], u[1]);
      return y(
        function() {
          f.hasValue = !0, f.value = m;
        },
        [m]
      ), v(m), m;
    }, typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ < "u" && typeof __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStop == "function" && __REACT_DEVTOOLS_GLOBAL_HOOK__.registerInternalModuleStop(Error());
  })()), V;
}
var q;
function ye() {
  return q || (q = 1, process.env.NODE_ENV === "production" ? T.exports = pe() : T.exports = ve()), T.exports;
}
var Se = ye();
export {
  ce as c,
  he as l,
  _e as s,
  Se as w
};
