// @cloudflare/kumo/dist/ai/schemas.js
"use client";
function d(e, o, t) {
  function n(l, u) {
    if (l._zod || Object.defineProperty(l, "_zod", {
      value: {
        def: u,
        constr: i,
        traits: /* @__PURE__ */ new Set()
      },
      enumerable: !1
    }), l._zod.traits.has(e))
      return;
    l._zod.traits.add(e), o(l, u);
    const p = i.prototype, h = Object.keys(p);
    for (let _ = 0; _ < h.length; _++) {
      const w = h[_];
      w in l || (l[w] = p[w].bind(l));
    }
  }
  const r = t?.Parent ?? Object;
  class a extends r {
  }
  Object.defineProperty(a, "name", { value: e });
  function i(l) {
    var u;
    const p = t?.Parent ? new a() : this;
    n(p, l), (u = p._zod).deferred ?? (u.deferred = []);
    for (const h of p._zod.deferred)
      h();
    return p;
  }
  return Object.defineProperty(i, "init", { value: n }), Object.defineProperty(i, Symbol.hasInstance, {
    value: (l) => t?.Parent && l instanceof t.Parent ? !0 : l?._zod?.traits?.has(e)
  }), Object.defineProperty(i, "name", { value: e }), i;
}
class q extends Error {
  constructor() {
    super("Encountered Promise during synchronous parse. Use .parseAsync() instead.");
  }
}
class Ge extends Error {
  constructor(o) {
    super(`Encountered unidirectional transform during encode: ${o}`), this.name = "ZodEncodeError";
  }
}
const qe = {};
function F(e) {
  return qe;
}
function Ke(e) {
  const o = Object.values(e).filter((n) => typeof n == "number");
  return Object.entries(e).filter(([n, r]) => o.indexOf(+n) === -1).map(([n, r]) => r);
}
function ge(e, o) {
  return typeof o == "bigint" ? o.toString() : o;
}
function be(e) {
  return {
    get value() {
      {
        const o = e();
        return Object.defineProperty(this, "value", { value: o }), o;
      }
    }
  };
}
function ye(e) {
  return e == null;
}
function ze(e) {
  const o = e.startsWith("^") ? 1 : 0, t = e.endsWith("$") ? e.length - 1 : e.length;
  return e.slice(o, t);
}
function To(e, o) {
  const t = (e.toString().split(".")[1] || "").length, n = o.toString();
  let r = (n.split(".")[1] || "").length;
  if (r === 0 && /\d?e-\d?/.test(n)) {
    const u = n.match(/\d?e-(\d?)/);
    u?.[1] && (r = Number.parseInt(u[1]));
  }
  const a = t > r ? t : r, i = Number.parseInt(e.toFixed(a).replace(".", "")), l = Number.parseInt(o.toFixed(a).replace(".", ""));
  return i % l / 10 ** a;
}
const Ze = Symbol("evaluating");
function $(e, o, t) {
  let n;
  Object.defineProperty(e, o, {
    get() {
      if (n !== Ze)
        return n === void 0 && (n = Ze, n = t()), n;
    },
    set(r) {
      Object.defineProperty(e, o, {
        value: r
        // configurable: true,
      });
    },
    configurable: !0
  });
}
function B(e, o, t) {
  Object.defineProperty(e, o, {
    value: t,
    writable: !0,
    enumerable: !0,
    configurable: !0
  });
}
function M(...e) {
  const o = {};
  for (const t of e) {
    const n = Object.getOwnPropertyDescriptors(t);
    Object.assign(o, n);
  }
  return Object.defineProperties({}, o);
}
function Pe(e) {
  return JSON.stringify(e);
}
function Oo(e) {
  return e.toLowerCase().trim().replace(/[^\w\s-]/g, "").replace(/[\s_-]+/g, "-").replace(/^-+|-+$/g, "");
}
const Xe = "captureStackTrace" in Error ? Error.captureStackTrace : (...e) => {
};
function ae(e) {
  return typeof e == "object" && e !== null && !Array.isArray(e);
}
const Io = be(() => {
  if (typeof navigator < "u" && navigator?.userAgent?.includes("Cloudflare"))
    return !1;
  try {
    const e = Function;
    return new e(""), !0;
  } catch {
    return !1;
  }
});
function K(e) {
  if (ae(e) === !1)
    return !1;
  const o = e.constructor;
  if (o === void 0 || typeof o != "function")
    return !0;
  const t = o.prototype;
  return !(ae(t) === !1 || Object.prototype.hasOwnProperty.call(t, "isPrototypeOf") === !1);
}
function Ye(e) {
  return K(e) ? { ...e } : Array.isArray(e) ? [...e] : e;
}
const No = /* @__PURE__ */ new Set(["string", "number", "symbol"]);
function ue(e) {
  return e.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function J(e, o, t) {
  const n = new e._zod.constr(o ?? e._zod.def);
  return (!o || t?.parent) && (n._zod.parent = e), n;
}
function k(e) {
  const o = e;
  if (!o)
    return {};
  if (typeof o == "string")
    return { error: () => o };
  if (o?.message !== void 0) {
    if (o?.error !== void 0)
      throw new Error("Cannot specify both `message` and `error` params");
    o.error = o.message;
  }
  return delete o.message, typeof o.error == "string" ? { ...o, error: () => o.error } : o;
}
function Eo(e) {
  return Object.keys(e).filter((o) => e[o]._zod.optin === "optional" && e[o]._zod.optout === "optional");
}
const Co = {
  safeint: [Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER],
  int32: [-2147483648, 2147483647],
  uint32: [0, 4294967295],
  float32: [-34028234663852886e22, 34028234663852886e22],
  float64: [-Number.MAX_VALUE, Number.MAX_VALUE]
};
function Ao(e, o) {
  const t = e._zod.def, n = t.checks;
  if (n && n.length > 0)
    throw new Error(".pick() cannot be used on object schemas containing refinements");
  const a = M(e._zod.def, {
    get shape() {
      const i = {};
      for (const l in o) {
        if (!(l in t.shape))
          throw new Error(`Unrecognized key: "${l}"`);
        o[l] && (i[l] = t.shape[l]);
      }
      return B(this, "shape", i), i;
    },
    checks: []
  });
  return J(e, a);
}
function xo(e, o) {
  const t = e._zod.def, n = t.checks;
  if (n && n.length > 0)
    throw new Error(".omit() cannot be used on object schemas containing refinements");
  const a = M(e._zod.def, {
    get shape() {
      const i = { ...e._zod.def.shape };
      for (const l in o) {
        if (!(l in t.shape))
          throw new Error(`Unrecognized key: "${l}"`);
        o[l] && delete i[l];
      }
      return B(this, "shape", i), i;
    },
    checks: []
  });
  return J(e, a);
}
function jo(e, o) {
  if (!K(o))
    throw new Error("Invalid input to extend: expected a plain object");
  const t = e._zod.def.checks;
  if (t && t.length > 0) {
    const a = e._zod.def.shape;
    for (const i in o)
      if (Object.getOwnPropertyDescriptor(a, i) !== void 0)
        throw new Error("Cannot overwrite keys on object schemas containing refinements. Use `.safeExtend()` instead.");
  }
  const r = M(e._zod.def, {
    get shape() {
      const a = { ...e._zod.def.shape, ...o };
      return B(this, "shape", a), a;
    }
  });
  return J(e, r);
}
function Ro(e, o) {
  if (!K(o))
    throw new Error("Invalid input to safeExtend: expected a plain object");
  const t = M(e._zod.def, {
    get shape() {
      const n = { ...e._zod.def.shape, ...o };
      return B(this, "shape", n), n;
    }
  });
  return J(e, t);
}
function Do(e, o) {
  const t = M(e._zod.def, {
    get shape() {
      const n = { ...e._zod.def.shape, ...o._zod.def.shape };
      return B(this, "shape", n), n;
    },
    get catchall() {
      return o._zod.def.catchall;
    },
    checks: []
    // delete existing checks
  });
  return J(e, t);
}
function Uo(e, o, t) {
  const r = o._zod.def.checks;
  if (r && r.length > 0)
    throw new Error(".partial() cannot be used on object schemas containing refinements");
  const i = M(o._zod.def, {
    get shape() {
      const l = o._zod.def.shape, u = { ...l };
      if (t)
        for (const p in t) {
          if (!(p in l))
            throw new Error(`Unrecognized key: "${p}"`);
          t[p] && (u[p] = e ? new e({
            type: "optional",
            innerType: l[p]
          }) : l[p]);
        }
      else
        for (const p in l)
          u[p] = e ? new e({
            type: "optional",
            innerType: l[p]
          }) : l[p];
      return B(this, "shape", u), u;
    },
    checks: []
  });
  return J(o, i);
}
function Fo(e, o, t) {
  const n = M(o._zod.def, {
    get shape() {
      const r = o._zod.def.shape, a = { ...r };
      if (t)
        for (const i in t) {
          if (!(i in a))
            throw new Error(`Unrecognized key: "${i}"`);
          t[i] && (a[i] = new e({
            type: "nonoptional",
            innerType: r[i]
          }));
        }
      else
        for (const i in r)
          a[i] = new e({
            type: "nonoptional",
            innerType: r[i]
          });
      return B(this, "shape", a), a;
    }
  });
  return J(o, n);
}
function G(e, o = 0) {
  if (e.aborted === !0)
    return !0;
  for (let t = o; t < e.issues.length; t++)
    if (e.issues[t]?.continue !== !0)
      return !0;
  return !1;
}
function V(e, o) {
  return o.map((t) => {
    var n;
    return (n = t).path ?? (n.path = []), t.path.unshift(e), t;
  });
}
function oe(e) {
  return typeof e == "string" ? e : e?.message;
}
function L(e, o, t) {
  const n = { ...e, path: e.path ?? [] };
  if (!e.message) {
    const r = oe(e.inst?._zod.def?.error?.(e)) ?? oe(o?.error?.(e)) ?? oe(t.customError?.(e)) ?? oe(t.localeError?.(e)) ?? "Invalid input";
    n.message = r;
  }
  return delete n.inst, delete n.continue, o?.reportInput || delete n.input, n;
}
function ke(e) {
  return Array.isArray(e) ? "array" : typeof e == "string" ? "string" : "unknown";
}
function Q(...e) {
  const [o, t, n] = e;
  return typeof o == "string" ? {
    message: o,
    code: "custom",
    input: t,
    inst: n
  } : { ...o };
}
const He = (e, o) => {
  e.name = "$ZodError", Object.defineProperty(e, "_zod", {
    value: e._zod,
    enumerable: !1
  }), Object.defineProperty(e, "issues", {
    value: o,
    enumerable: !1
  }), e.message = JSON.stringify(o, ge, 2), Object.defineProperty(e, "toString", {
    value: () => e.message,
    enumerable: !1
  });
}, Qe = d("$ZodError", He), eo = d("$ZodError", He, { Parent: Error });
function Lo(e, o = (t) => t.message) {
  const t = {}, n = [];
  for (const r of e.issues)
    r.path.length > 0 ? (t[r.path[0]] = t[r.path[0]] || [], t[r.path[0]].push(o(r))) : n.push(o(r));
  return { formErrors: n, fieldErrors: t };
}
function Mo(e, o = (t) => t.message) {
  const t = { _errors: [] }, n = (r) => {
    for (const a of r.issues)
      if (a.code === "invalid_union" && a.errors.length)
        a.errors.map((i) => n({ issues: i }));
      else if (a.code === "invalid_key")
        n({ issues: a.issues });
      else if (a.code === "invalid_element")
        n({ issues: a.issues });
      else if (a.path.length === 0)
        t._errors.push(o(a));
      else {
        let i = t, l = 0;
        for (; l < a.path.length; ) {
          const u = a.path[l];
          l === a.path.length - 1 ? (i[u] = i[u] || { _errors: [] }, i[u]._errors.push(o(a))) : i[u] = i[u] || { _errors: [] }, i = i[u], l++;
        }
      }
  };
  return n(e), t;
}
const we = (e) => (o, t, n, r) => {
  const a = n ? Object.assign(n, { async: !1 }) : { async: !1 }, i = o._zod.run({ value: t, issues: [] }, a);
  if (i instanceof Promise)
    throw new q();
  if (i.issues.length) {
    const l = new (r?.Err ?? e)(i.issues.map((u) => L(u, a, F())));
    throw Xe(l, r?.callee), l;
  }
  return i.value;
}, Se = (e) => async (o, t, n, r) => {
  const a = n ? Object.assign(n, { async: !0 }) : { async: !0 };
  let i = o._zod.run({ value: t, issues: [] }, a);
  if (i instanceof Promise && (i = await i), i.issues.length) {
    const l = new (r?.Err ?? e)(i.issues.map((u) => L(u, a, F())));
    throw Xe(l, r?.callee), l;
  }
  return i.value;
}, pe = (e) => (o, t, n) => {
  const r = n ? { ...n, async: !1 } : { async: !1 }, a = o._zod.run({ value: t, issues: [] }, r);
  if (a instanceof Promise)
    throw new q();
  return a.issues.length ? {
    success: !1,
    error: new (e ?? Qe)(a.issues.map((i) => L(i, r, F())))
  } : { success: !0, data: a.value };
}, Jo = /* @__PURE__ */ pe(eo), de = (e) => async (o, t, n) => {
  const r = n ? Object.assign(n, { async: !0 }) : { async: !0 };
  let a = o._zod.run({ value: t, issues: [] }, r);
  return a instanceof Promise && (a = await a), a.issues.length ? {
    success: !1,
    error: new e(a.issues.map((i) => L(i, r, F())))
  } : { success: !0, data: a.value };
}, Vo = /* @__PURE__ */ de(eo), Bo = (e) => (o, t, n) => {
  const r = n ? Object.assign(n, { direction: "backward" }) : { direction: "backward" };
  return we(e)(o, t, r);
}, Wo = (e) => (o, t, n) => we(e)(o, t, n), Go = (e) => async (o, t, n) => {
  const r = n ? Object.assign(n, { direction: "backward" }) : { direction: "backward" };
  return Se(e)(o, t, r);
}, qo = (e) => async (o, t, n) => Se(e)(o, t, n), Ko = (e) => (o, t, n) => {
  const r = n ? Object.assign(n, { direction: "backward" }) : { direction: "backward" };
  return pe(e)(o, t, r);
}, Xo = (e) => (o, t, n) => pe(e)(o, t, n), Yo = (e) => async (o, t, n) => {
  const r = n ? Object.assign(n, { direction: "backward" }) : { direction: "backward" };
  return de(e)(o, t, r);
}, Ho = (e) => async (o, t, n) => de(e)(o, t, n), Qo = /^[cC][^\s-]{8,}$/, et = /^[0-9a-z]+$/, ot = /^[0-9A-HJKMNP-TV-Za-hjkmnp-tv-z]{26}$/, tt = /^[0-9a-vA-V]{20}$/, nt = /^[A-Za-z0-9]{27}$/, rt = /^[a-zA-Z0-9_-]{21}$/, it = /^P(?:(\d+W)|(?!.*W)(?=\d|T\d)(\d+Y)?(\d+M)?(\d+D)?(T(?=\d)(\d+H)?(\d+M)?(\d+([.,]\d+)?S)?)?)$/, at = /^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})$/, Te = (e) => e ? new RegExp(`^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-${e}[0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12})$`) : /^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-8][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}|00000000-0000-0000-0000-000000000000|ffffffff-ffff-ffff-ffff-ffffffffffff)$/, st = /^(?!\.)(?!.*\.\.)([A-Za-z0-9_'+\-\.]*)[A-Za-z0-9_+-]@([A-Za-z0-9][A-Za-z0-9\-]*\.)+[A-Za-z]{2,}$/, ct = "^(\\p{Extended_Pictographic}|\\p{Emoji_Component})+$";
function lt() {
  return new RegExp(ct, "u");
}
const ut = /^(?:(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\.){3}(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])$/, pt = /^(([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:))$/, dt = /^((25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\.){3}(25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\/([0-9]|[1-2][0-9]|3[0-2])$/, ft = /^(([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}|::|([0-9a-fA-F]{1,4})?::([0-9a-fA-F]{1,4}:?){0,6})\/(12[0-8]|1[01][0-9]|[1-9]?[0-9])$/, ht = /^$|^(?:[0-9a-zA-Z+/]{4})*(?:(?:[0-9a-zA-Z+/]{2}==)|(?:[0-9a-zA-Z+/]{3}=))?$/, oo = /^[A-Za-z0-9_-]*$/, mt = /^\+[1-9]\d{6,14}$/, to = "(?:(?:\\d\\d[2468][048]|\\d\\d[13579][26]|\\d\\d0[48]|[02468][048]00|[13579][26]00)-02-29|\\d{4}-(?:(?:0[13578]|1[02])-(?:0[1-9]|[12]\\d|3[01])|(?:0[469]|11)-(?:0[1-9]|[12]\\d|30)|(?:02)-(?:0[1-9]|1\\d|2[0-8])))", gt = /* @__PURE__ */ new RegExp(`^${to}$`);
function no(e) {
  const o = "(?:[01]\\d|2[0-3]):[0-5]\\d";
  return typeof e.precision == "number" ? e.precision === -1 ? `${o}` : e.precision === 0 ? `${o}:[0-5]\\d` : `${o}:[0-5]\\d\\.\\d{${e.precision}}` : `${o}(?::[0-5]\\d(?:\\.\\d+)?)?`;
}
function _t(e) {
  return new RegExp(`^${no(e)}$`);
}
function vt(e) {
  const o = no({ precision: e.precision }), t = ["Z"];
  e.local && t.push(""), e.offset && t.push("([+-](?:[01]\\d|2[0-3]):[0-5]\\d)");
  const n = `${o}(?:${t.join("|")})`;
  return new RegExp(`^${to}T(?:${n})$`);
}
const bt = (e) => {
  const o = e ? `[\\s\\S]{${e?.minimum ?? 0},${e?.maximum ?? ""}}` : "[\\s\\S]*";
  return new RegExp(`^${o}$`);
}, yt = /^-?\d+$/, ro = /^-?\d+(?:\.\d+)?$/, zt = /^(?:true|false)$/i, kt = /^null$/i, wt = /^[^A-Z]*$/, St = /^[^a-z]*$/, x = /* @__PURE__ */ d("$ZodCheck", (e, o) => {
  var t;
  e._zod ?? (e._zod = {}), e._zod.def = o, (t = e._zod).onattach ?? (t.onattach = []);
}), io = {
  number: "number",
  bigint: "bigint",
  object: "date"
}, ao = /* @__PURE__ */ d("$ZodCheckLessThan", (e, o) => {
  x.init(e, o);
  const t = io[typeof o.value];
  e._zod.onattach.push((n) => {
    const r = n._zod.bag, a = (o.inclusive ? r.maximum : r.exclusiveMaximum) ?? Number.POSITIVE_INFINITY;
    o.value < a && (o.inclusive ? r.maximum = o.value : r.exclusiveMaximum = o.value);
  }), e._zod.check = (n) => {
    (o.inclusive ? n.value <= o.value : n.value < o.value) || n.issues.push({
      origin: t,
      code: "too_big",
      maximum: typeof o.value == "object" ? o.value.getTime() : o.value,
      input: n.value,
      inclusive: o.inclusive,
      inst: e,
      continue: !o.abort
    });
  };
}), so = /* @__PURE__ */ d("$ZodCheckGreaterThan", (e, o) => {
  x.init(e, o);
  const t = io[typeof o.value];
  e._zod.onattach.push((n) => {
    const r = n._zod.bag, a = (o.inclusive ? r.minimum : r.exclusiveMinimum) ?? Number.NEGATIVE_INFINITY;
    o.value > a && (o.inclusive ? r.minimum = o.value : r.exclusiveMinimum = o.value);
  }), e._zod.check = (n) => {
    (o.inclusive ? n.value >= o.value : n.value > o.value) || n.issues.push({
      origin: t,
      code: "too_small",
      minimum: typeof o.value == "object" ? o.value.getTime() : o.value,
      input: n.value,
      inclusive: o.inclusive,
      inst: e,
      continue: !o.abort
    });
  };
}), $t = /* @__PURE__ */ d("$ZodCheckMultipleOf", (e, o) => {
  x.init(e, o), e._zod.onattach.push((t) => {
    var n;
    (n = t._zod.bag).multipleOf ?? (n.multipleOf = o.value);
  }), e._zod.check = (t) => {
    if (typeof t.value != typeof o.value)
      throw new Error("Cannot mix number and bigint in multiple_of check.");
    (typeof t.value == "bigint" ? t.value % o.value === BigInt(0) : To(t.value, o.value) === 0) || t.issues.push({
      origin: typeof t.value,
      code: "not_multiple_of",
      divisor: o.value,
      input: t.value,
      inst: e,
      continue: !o.abort
    });
  };
}), Zt = /* @__PURE__ */ d("$ZodCheckNumberFormat", (e, o) => {
  x.init(e, o), o.format = o.format || "float64";
  const t = o.format?.includes("int"), n = t ? "int" : "number", [r, a] = Co[o.format];
  e._zod.onattach.push((i) => {
    const l = i._zod.bag;
    l.format = o.format, l.minimum = r, l.maximum = a, t && (l.pattern = yt);
  }), e._zod.check = (i) => {
    const l = i.value;
    if (t) {
      if (!Number.isInteger(l)) {
        i.issues.push({
          expected: n,
          format: o.format,
          code: "invalid_type",
          continue: !1,
          input: l,
          inst: e
        });
        return;
      }
      if (!Number.isSafeInteger(l)) {
        l > 0 ? i.issues.push({
          input: l,
          code: "too_big",
          maximum: Number.MAX_SAFE_INTEGER,
          note: "Integers must be within the safe integer range.",
          inst: e,
          origin: n,
          inclusive: !0,
          continue: !o.abort
        }) : i.issues.push({
          input: l,
          code: "too_small",
          minimum: Number.MIN_SAFE_INTEGER,
          note: "Integers must be within the safe integer range.",
          inst: e,
          origin: n,
          inclusive: !0,
          continue: !o.abort
        });
        return;
      }
    }
    l < r && i.issues.push({
      origin: "number",
      input: l,
      code: "too_small",
      minimum: r,
      inclusive: !0,
      inst: e,
      continue: !o.abort
    }), l > a && i.issues.push({
      origin: "number",
      input: l,
      code: "too_big",
      maximum: a,
      inclusive: !0,
      inst: e,
      continue: !o.abort
    });
  };
}), Pt = /* @__PURE__ */ d("$ZodCheckMaxLength", (e, o) => {
  var t;
  x.init(e, o), (t = e._zod.def).when ?? (t.when = (n) => {
    const r = n.value;
    return !ye(r) && r.length !== void 0;
  }), e._zod.onattach.push((n) => {
    const r = n._zod.bag.maximum ?? Number.POSITIVE_INFINITY;
    o.maximum < r && (n._zod.bag.maximum = o.maximum);
  }), e._zod.check = (n) => {
    const r = n.value;
    if (r.length <= o.maximum)
      return;
    const i = ke(r);
    n.issues.push({
      origin: i,
      code: "too_big",
      maximum: o.maximum,
      inclusive: !0,
      input: r,
      inst: e,
      continue: !o.abort
    });
  };
}), Tt = /* @__PURE__ */ d("$ZodCheckMinLength", (e, o) => {
  var t;
  x.init(e, o), (t = e._zod.def).when ?? (t.when = (n) => {
    const r = n.value;
    return !ye(r) && r.length !== void 0;
  }), e._zod.onattach.push((n) => {
    const r = n._zod.bag.minimum ?? Number.NEGATIVE_INFINITY;
    o.minimum > r && (n._zod.bag.minimum = o.minimum);
  }), e._zod.check = (n) => {
    const r = n.value;
    if (r.length >= o.minimum)
      return;
    const i = ke(r);
    n.issues.push({
      origin: i,
      code: "too_small",
      minimum: o.minimum,
      inclusive: !0,
      input: r,
      inst: e,
      continue: !o.abort
    });
  };
}), Ot = /* @__PURE__ */ d("$ZodCheckLengthEquals", (e, o) => {
  var t;
  x.init(e, o), (t = e._zod.def).when ?? (t.when = (n) => {
    const r = n.value;
    return !ye(r) && r.length !== void 0;
  }), e._zod.onattach.push((n) => {
    const r = n._zod.bag;
    r.minimum = o.length, r.maximum = o.length, r.length = o.length;
  }), e._zod.check = (n) => {
    const r = n.value, a = r.length;
    if (a === o.length)
      return;
    const i = ke(r), l = a > o.length;
    n.issues.push({
      origin: i,
      ...l ? { code: "too_big", maximum: o.length } : { code: "too_small", minimum: o.length },
      inclusive: !0,
      exact: !0,
      input: n.value,
      inst: e,
      continue: !o.abort
    });
  };
}), fe = /* @__PURE__ */ d("$ZodCheckStringFormat", (e, o) => {
  var t, n;
  x.init(e, o), e._zod.onattach.push((r) => {
    const a = r._zod.bag;
    a.format = o.format, o.pattern && (a.patterns ?? (a.patterns = /* @__PURE__ */ new Set()), a.patterns.add(o.pattern));
  }), o.pattern ? (t = e._zod).check ?? (t.check = (r) => {
    o.pattern.lastIndex = 0, !o.pattern.test(r.value) && r.issues.push({
      origin: "string",
      code: "invalid_format",
      format: o.format,
      input: r.value,
      ...o.pattern ? { pattern: o.pattern.toString() } : {},
      inst: e,
      continue: !o.abort
    });
  }) : (n = e._zod).check ?? (n.check = () => {
  });
}), It = /* @__PURE__ */ d("$ZodCheckRegex", (e, o) => {
  fe.init(e, o), e._zod.check = (t) => {
    o.pattern.lastIndex = 0, !o.pattern.test(t.value) && t.issues.push({
      origin: "string",
      code: "invalid_format",
      format: "regex",
      input: t.value,
      pattern: o.pattern.toString(),
      inst: e,
      continue: !o.abort
    });
  };
}), Nt = /* @__PURE__ */ d("$ZodCheckLowerCase", (e, o) => {
  o.pattern ?? (o.pattern = wt), fe.init(e, o);
}), Et = /* @__PURE__ */ d("$ZodCheckUpperCase", (e, o) => {
  o.pattern ?? (o.pattern = St), fe.init(e, o);
}), Ct = /* @__PURE__ */ d("$ZodCheckIncludes", (e, o) => {
  x.init(e, o);
  const t = ue(o.includes), n = new RegExp(typeof o.position == "number" ? `^.{${o.position}}${t}` : t);
  o.pattern = n, e._zod.onattach.push((r) => {
    const a = r._zod.bag;
    a.patterns ?? (a.patterns = /* @__PURE__ */ new Set()), a.patterns.add(n);
  }), e._zod.check = (r) => {
    r.value.includes(o.includes, o.position) || r.issues.push({
      origin: "string",
      code: "invalid_format",
      format: "includes",
      includes: o.includes,
      input: r.value,
      inst: e,
      continue: !o.abort
    });
  };
}), At = /* @__PURE__ */ d("$ZodCheckStartsWith", (e, o) => {
  x.init(e, o);
  const t = new RegExp(`^${ue(o.prefix)}.*`);
  o.pattern ?? (o.pattern = t), e._zod.onattach.push((n) => {
    const r = n._zod.bag;
    r.patterns ?? (r.patterns = /* @__PURE__ */ new Set()), r.patterns.add(t);
  }), e._zod.check = (n) => {
    n.value.startsWith(o.prefix) || n.issues.push({
      origin: "string",
      code: "invalid_format",
      format: "starts_with",
      prefix: o.prefix,
      input: n.value,
      inst: e,
      continue: !o.abort
    });
  };
}), xt = /* @__PURE__ */ d("$ZodCheckEndsWith", (e, o) => {
  x.init(e, o);
  const t = new RegExp(`.*${ue(o.suffix)}$`);
  o.pattern ?? (o.pattern = t), e._zod.onattach.push((n) => {
    const r = n._zod.bag;
    r.patterns ?? (r.patterns = /* @__PURE__ */ new Set()), r.patterns.add(t);
  }), e._zod.check = (n) => {
    n.value.endsWith(o.suffix) || n.issues.push({
      origin: "string",
      code: "invalid_format",
      format: "ends_with",
      suffix: o.suffix,
      input: n.value,
      inst: e,
      continue: !o.abort
    });
  };
}), jt = /* @__PURE__ */ d("$ZodCheckOverwrite", (e, o) => {
  x.init(e, o), e._zod.check = (t) => {
    t.value = o.tx(t.value);
  };
});
class Rt {
  constructor(o = []) {
    this.content = [], this.indent = 0, this && (this.args = o);
  }
  indented(o) {
    this.indent += 1, o(this), this.indent -= 1;
  }
  write(o) {
    if (typeof o == "function") {
      o(this, { execution: "sync" }), o(this, { execution: "async" });
      return;
    }
    const n = o.split(`
`).filter((i) => i), r = Math.min(...n.map((i) => i.length - i.trimStart().length)), a = n.map((i) => i.slice(r)).map((i) => " ".repeat(this.indent * 2) + i);
    for (const i of a)
      this.content.push(i);
  }
  compile() {
    const o = Function, t = this?.args, r = [...(this?.content ?? [""]).map((a) => `  ${a}`)];
    return new o(...t, r.join(`
`));
  }
}
const Dt = {
  major: 4,
  minor: 3,
  patch: 6
}, Z = /* @__PURE__ */ d("$ZodType", (e, o) => {
  var t;
  e ?? (e = {}), e._zod.def = o, e._zod.bag = e._zod.bag || {}, e._zod.version = Dt;
  const n = [...e._zod.def.checks ?? []];
  e._zod.traits.has("$ZodCheck") && n.unshift(e);
  for (const r of n)
    for (const a of r._zod.onattach)
      a(e);
  if (n.length === 0)
    (t = e._zod).deferred ?? (t.deferred = []), e._zod.deferred?.push(() => {
      e._zod.run = e._zod.parse;
    });
  else {
    const r = (i, l, u) => {
      let p = G(i), h;
      for (const _ of l) {
        if (_._zod.def.when) {
          if (!_._zod.def.when(i))
            continue;
        } else if (p)
          continue;
        const w = i.issues.length, S = _._zod.check(i);
        if (S instanceof Promise && u?.async === !1)
          throw new q();
        if (h || S instanceof Promise)
          h = (h ?? Promise.resolve()).then(async () => {
            await S, i.issues.length !== w && (p || (p = G(i, w)));
          });
        else {
          if (i.issues.length === w)
            continue;
          p || (p = G(i, w));
        }
      }
      return h ? h.then(() => i) : i;
    }, a = (i, l, u) => {
      if (G(i))
        return i.aborted = !0, i;
      const p = r(l, n, u);
      if (p instanceof Promise) {
        if (u.async === !1)
          throw new q();
        return p.then((h) => e._zod.parse(h, u));
      }
      return e._zod.parse(p, u);
    };
    e._zod.run = (i, l) => {
      if (l.skipChecks)
        return e._zod.parse(i, l);
      if (l.direction === "backward") {
        const p = e._zod.parse({ value: i.value, issues: [] }, { ...l, skipChecks: !0 });
        return p instanceof Promise ? p.then((h) => a(h, i, l)) : a(p, i, l);
      }
      const u = e._zod.parse(i, l);
      if (u instanceof Promise) {
        if (l.async === !1)
          throw new q();
        return u.then((p) => r(p, n, l));
      }
      return r(u, n, l);
    };
  }
  $(e, "~standard", () => ({
    validate: (r) => {
      try {
        const a = Jo(e, r);
        return a.success ? { value: a.data } : { issues: a.error?.issues };
      } catch {
        return Vo(e, r).then((i) => i.success ? { value: i.data } : { issues: i.error?.issues });
      }
    },
    vendor: "zod",
    version: 1
  }));
}), $e = /* @__PURE__ */ d("$ZodString", (e, o) => {
  Z.init(e, o), e._zod.pattern = [...e?._zod.bag?.patterns ?? []].pop() ?? bt(e._zod.bag), e._zod.parse = (t, n) => {
    if (o.coerce)
      try {
        t.value = String(t.value);
      } catch {
      }
    return typeof t.value == "string" || t.issues.push({
      expected: "string",
      code: "invalid_type",
      input: t.value,
      inst: e
    }), t;
  };
}), T = /* @__PURE__ */ d("$ZodStringFormat", (e, o) => {
  fe.init(e, o), $e.init(e, o);
}), Ut = /* @__PURE__ */ d("$ZodGUID", (e, o) => {
  o.pattern ?? (o.pattern = at), T.init(e, o);
}), Ft = /* @__PURE__ */ d("$ZodUUID", (e, o) => {
  if (o.version) {
    const n = {
      v1: 1,
      v2: 2,
      v3: 3,
      v4: 4,
      v5: 5,
      v6: 6,
      v7: 7,
      v8: 8
    }[o.version];
    if (n === void 0)
      throw new Error(`Invalid UUID version: "${o.version}"`);
    o.pattern ?? (o.pattern = Te(n));
  } else
    o.pattern ?? (o.pattern = Te());
  T.init(e, o);
}), Lt = /* @__PURE__ */ d("$ZodEmail", (e, o) => {
  o.pattern ?? (o.pattern = st), T.init(e, o);
}), Mt = /* @__PURE__ */ d("$ZodURL", (e, o) => {
  T.init(e, o), e._zod.check = (t) => {
    try {
      const n = t.value.trim(), r = new URL(n);
      o.hostname && (o.hostname.lastIndex = 0, o.hostname.test(r.hostname) || t.issues.push({
        code: "invalid_format",
        format: "url",
        note: "Invalid hostname",
        pattern: o.hostname.source,
        input: t.value,
        inst: e,
        continue: !o.abort
      })), o.protocol && (o.protocol.lastIndex = 0, o.protocol.test(r.protocol.endsWith(":") ? r.protocol.slice(0, -1) : r.protocol) || t.issues.push({
        code: "invalid_format",
        format: "url",
        note: "Invalid protocol",
        pattern: o.protocol.source,
        input: t.value,
        inst: e,
        continue: !o.abort
      })), o.normalize ? t.value = r.href : t.value = n;
      return;
    } catch {
      t.issues.push({
        code: "invalid_format",
        format: "url",
        input: t.value,
        inst: e,
        continue: !o.abort
      });
    }
  };
}), Jt = /* @__PURE__ */ d("$ZodEmoji", (e, o) => {
  o.pattern ?? (o.pattern = lt()), T.init(e, o);
}), Vt = /* @__PURE__ */ d("$ZodNanoID", (e, o) => {
  o.pattern ?? (o.pattern = rt), T.init(e, o);
}), Bt = /* @__PURE__ */ d("$ZodCUID", (e, o) => {
  o.pattern ?? (o.pattern = Qo), T.init(e, o);
}), Wt = /* @__PURE__ */ d("$ZodCUID2", (e, o) => {
  o.pattern ?? (o.pattern = et), T.init(e, o);
}), Gt = /* @__PURE__ */ d("$ZodULID", (e, o) => {
  o.pattern ?? (o.pattern = ot), T.init(e, o);
}), qt = /* @__PURE__ */ d("$ZodXID", (e, o) => {
  o.pattern ?? (o.pattern = tt), T.init(e, o);
}), Kt = /* @__PURE__ */ d("$ZodKSUID", (e, o) => {
  o.pattern ?? (o.pattern = nt), T.init(e, o);
}), Xt = /* @__PURE__ */ d("$ZodISODateTime", (e, o) => {
  o.pattern ?? (o.pattern = vt(o)), T.init(e, o);
}), Yt = /* @__PURE__ */ d("$ZodISODate", (e, o) => {
  o.pattern ?? (o.pattern = gt), T.init(e, o);
}), Ht = /* @__PURE__ */ d("$ZodISOTime", (e, o) => {
  o.pattern ?? (o.pattern = _t(o)), T.init(e, o);
}), Qt = /* @__PURE__ */ d("$ZodISODuration", (e, o) => {
  o.pattern ?? (o.pattern = it), T.init(e, o);
}), en = /* @__PURE__ */ d("$ZodIPv4", (e, o) => {
  o.pattern ?? (o.pattern = ut), T.init(e, o), e._zod.bag.format = "ipv4";
}), on = /* @__PURE__ */ d("$ZodIPv6", (e, o) => {
  o.pattern ?? (o.pattern = pt), T.init(e, o), e._zod.bag.format = "ipv6", e._zod.check = (t) => {
    try {
      new URL(`http://[${t.value}]`);
    } catch {
      t.issues.push({
        code: "invalid_format",
        format: "ipv6",
        input: t.value,
        inst: e,
        continue: !o.abort
      });
    }
  };
}), tn = /* @__PURE__ */ d("$ZodCIDRv4", (e, o) => {
  o.pattern ?? (o.pattern = dt), T.init(e, o);
}), nn = /* @__PURE__ */ d("$ZodCIDRv6", (e, o) => {
  o.pattern ?? (o.pattern = ft), T.init(e, o), e._zod.check = (t) => {
    const n = t.value.split("/");
    try {
      if (n.length !== 2)
        throw new Error();
      const [r, a] = n;
      if (!a)
        throw new Error();
      const i = Number(a);
      if (`${i}` !== a)
        throw new Error();
      if (i < 0 || i > 128)
        throw new Error();
      new URL(`http://[${r}]`);
    } catch {
      t.issues.push({
        code: "invalid_format",
        format: "cidrv6",
        input: t.value,
        inst: e,
        continue: !o.abort
      });
    }
  };
});
function co(e) {
  if (e === "")
    return !0;
  if (e.length % 4 !== 0)
    return !1;
  try {
    return atob(e), !0;
  } catch {
    return !1;
  }
}
const rn = /* @__PURE__ */ d("$ZodBase64", (e, o) => {
  o.pattern ?? (o.pattern = ht), T.init(e, o), e._zod.bag.contentEncoding = "base64", e._zod.check = (t) => {
    co(t.value) || t.issues.push({
      code: "invalid_format",
      format: "base64",
      input: t.value,
      inst: e,
      continue: !o.abort
    });
  };
});
function an(e) {
  if (!oo.test(e))
    return !1;
  const o = e.replace(/[-_]/g, (n) => n === "-" ? "+" : "/"), t = o.padEnd(Math.ceil(o.length / 4) * 4, "=");
  return co(t);
}
const sn = /* @__PURE__ */ d("$ZodBase64URL", (e, o) => {
  o.pattern ?? (o.pattern = oo), T.init(e, o), e._zod.bag.contentEncoding = "base64url", e._zod.check = (t) => {
    an(t.value) || t.issues.push({
      code: "invalid_format",
      format: "base64url",
      input: t.value,
      inst: e,
      continue: !o.abort
    });
  };
}), cn = /* @__PURE__ */ d("$ZodE164", (e, o) => {
  o.pattern ?? (o.pattern = mt), T.init(e, o);
});
function ln(e, o = null) {
  try {
    const t = e.split(".");
    if (t.length !== 3)
      return !1;
    const [n] = t;
    if (!n)
      return !1;
    const r = JSON.parse(atob(n));
    return !("typ" in r && r?.typ !== "JWT" || !r.alg || o && (!("alg" in r) || r.alg !== o));
  } catch {
    return !1;
  }
}
const un = /* @__PURE__ */ d("$ZodJWT", (e, o) => {
  T.init(e, o), e._zod.check = (t) => {
    ln(t.value, o.alg) || t.issues.push({
      code: "invalid_format",
      format: "jwt",
      input: t.value,
      inst: e,
      continue: !o.abort
    });
  };
}), lo = /* @__PURE__ */ d("$ZodNumber", (e, o) => {
  Z.init(e, o), e._zod.pattern = e._zod.bag.pattern ?? ro, e._zod.parse = (t, n) => {
    if (o.coerce)
      try {
        t.value = Number(t.value);
      } catch {
      }
    const r = t.value;
    if (typeof r == "number" && !Number.isNaN(r) && Number.isFinite(r))
      return t;
    const a = typeof r == "number" ? Number.isNaN(r) ? "NaN" : Number.isFinite(r) ? void 0 : "Infinity" : void 0;
    return t.issues.push({
      expected: "number",
      code: "invalid_type",
      input: r,
      inst: e,
      ...a ? { received: a } : {}
    }), t;
  };
}), pn = /* @__PURE__ */ d("$ZodNumberFormat", (e, o) => {
  Zt.init(e, o), lo.init(e, o);
}), dn = /* @__PURE__ */ d("$ZodBoolean", (e, o) => {
  Z.init(e, o), e._zod.pattern = zt, e._zod.parse = (t, n) => {
    if (o.coerce)
      try {
        t.value = !!t.value;
      } catch {
      }
    const r = t.value;
    return typeof r == "boolean" || t.issues.push({
      expected: "boolean",
      code: "invalid_type",
      input: r,
      inst: e
    }), t;
  };
}), fn = /* @__PURE__ */ d("$ZodNull", (e, o) => {
  Z.init(e, o), e._zod.pattern = kt, e._zod.values = /* @__PURE__ */ new Set([null]), e._zod.parse = (t, n) => {
    const r = t.value;
    return r === null || t.issues.push({
      expected: "null",
      code: "invalid_type",
      input: r,
      inst: e
    }), t;
  };
}), hn = /* @__PURE__ */ d("$ZodUnknown", (e, o) => {
  Z.init(e, o), e._zod.parse = (t) => t;
}), mn = /* @__PURE__ */ d("$ZodNever", (e, o) => {
  Z.init(e, o), e._zod.parse = (t, n) => (t.issues.push({
    expected: "never",
    code: "invalid_type",
    input: t.value,
    inst: e
  }), t);
});
function Oe(e, o, t) {
  e.issues.length && o.issues.push(...V(t, e.issues)), o.value[t] = e.value;
}
const gn = /* @__PURE__ */ d("$ZodArray", (e, o) => {
  Z.init(e, o), e._zod.parse = (t, n) => {
    const r = t.value;
    if (!Array.isArray(r))
      return t.issues.push({
        expected: "array",
        code: "invalid_type",
        input: r,
        inst: e
      }), t;
    t.value = Array(r.length);
    const a = [];
    for (let i = 0; i < r.length; i++) {
      const l = r[i], u = o.element._zod.run({
        value: l,
        issues: []
      }, n);
      u instanceof Promise ? a.push(u.then((p) => Oe(p, t, i))) : Oe(u, t, i);
    }
    return a.length ? Promise.all(a).then(() => t) : t;
  };
});
function se(e, o, t, n, r) {
  if (e.issues.length) {
    if (r && !(t in n))
      return;
    o.issues.push(...V(t, e.issues));
  }
  e.value === void 0 ? t in n && (o.value[t] = void 0) : o.value[t] = e.value;
}
function uo(e) {
  const o = Object.keys(e.shape);
  for (const n of o)
    if (!e.shape?.[n]?._zod?.traits?.has("$ZodType"))
      throw new Error(`Invalid element at key "${n}": expected a Zod schema`);
  const t = Eo(e.shape);
  return {
    ...e,
    keys: o,
    keySet: new Set(o),
    numKeys: o.length,
    optionalKeys: new Set(t)
  };
}
function po(e, o, t, n, r, a) {
  const i = [], l = r.keySet, u = r.catchall._zod, p = u.def.type, h = u.optout === "optional";
  for (const _ in o) {
    if (l.has(_))
      continue;
    if (p === "never") {
      i.push(_);
      continue;
    }
    const w = u.run({ value: o[_], issues: [] }, n);
    w instanceof Promise ? e.push(w.then((S) => se(S, t, _, o, h))) : se(w, t, _, o, h);
  }
  return i.length && t.issues.push({
    code: "unrecognized_keys",
    keys: i,
    input: o,
    inst: a
  }), e.length ? Promise.all(e).then(() => t) : t;
}
const _n = /* @__PURE__ */ d("$ZodObject", (e, o) => {
  if (Z.init(e, o), !Object.getOwnPropertyDescriptor(o, "shape")?.get) {
    const l = o.shape;
    Object.defineProperty(o, "shape", {
      get: () => {
        const u = { ...l };
        return Object.defineProperty(o, "shape", {
          value: u
        }), u;
      }
    });
  }
  const n = be(() => uo(o));
  $(e._zod, "propValues", () => {
    const l = o.shape, u = {};
    for (const p in l) {
      const h = l[p]._zod;
      if (h.values) {
        u[p] ?? (u[p] = /* @__PURE__ */ new Set());
        for (const _ of h.values)
          u[p].add(_);
      }
    }
    return u;
  });
  const r = ae, a = o.catchall;
  let i;
  e._zod.parse = (l, u) => {
    i ?? (i = n.value);
    const p = l.value;
    if (!r(p))
      return l.issues.push({
        expected: "object",
        code: "invalid_type",
        input: p,
        inst: e
      }), l;
    l.value = {};
    const h = [], _ = i.shape;
    for (const w of i.keys) {
      const S = _[w], E = S._zod.optout === "optional", N = S._zod.run({ value: p[w], issues: [] }, u);
      N instanceof Promise ? h.push(N.then((ee) => se(ee, l, w, p, E))) : se(N, l, w, p, E);
    }
    return a ? po(h, p, l, u, n.value, e) : h.length ? Promise.all(h).then(() => l) : l;
  };
}), vn = /* @__PURE__ */ d("$ZodObjectJIT", (e, o) => {
  _n.init(e, o);
  const t = e._zod.parse, n = be(() => uo(o)), r = (w) => {
    const S = new Rt(["shape", "payload", "ctx"]), E = n.value, N = (D) => {
      const A = Pe(D);
      return `shape[${A}]._zod.run({ value: input[${A}], issues: [] }, ctx)`;
    };
    S.write("const input = payload.value;");
    const ee = /* @__PURE__ */ Object.create(null);
    let $o = 0;
    for (const D of E.keys)
      ee[D] = `key_${$o++}`;
    S.write("const newResult = {};");
    for (const D of E.keys) {
      const A = ee[D], R = Pe(D), Po = w[D]?._zod?.optout === "optional";
      S.write(`const ${A} = ${N(D)};`), Po ? S.write(`
        if (${A}.issues.length) {
          if (${R} in input) {
            payload.issues = payload.issues.concat(${A}.issues.map(iss => ({
              ...iss,
              path: iss.path ? [${R}, ...iss.path] : [${R}]
            })));
          }
        }
        
        if (${A}.value === undefined) {
          if (${R} in input) {
            newResult[${R}] = undefined;
          }
        } else {
          newResult[${R}] = ${A}.value;
        }
        
      `) : S.write(`
        if (${A}.issues.length) {
          payload.issues = payload.issues.concat(${A}.issues.map(iss => ({
            ...iss,
            path: iss.path ? [${R}, ...iss.path] : [${R}]
          })));
        }
        
        if (${A}.value === undefined) {
          if (${R} in input) {
            newResult[${R}] = undefined;
          }
        } else {
          newResult[${R}] = ${A}.value;
        }
        
      `);
    }
    S.write("payload.value = newResult;"), S.write("return payload;");
    const Zo = S.compile();
    return (D, A) => Zo(w, D, A);
  };
  let a;
  const i = ae, l = !qe.jitless, p = l && Io.value, h = o.catchall;
  let _;
  e._zod.parse = (w, S) => {
    _ ?? (_ = n.value);
    const E = w.value;
    return i(E) ? l && p && S?.async === !1 && S.jitless !== !0 ? (a || (a = r(o.shape)), w = a(w, S), h ? po([], E, w, S, _, e) : w) : t(w, S) : (w.issues.push({
      expected: "object",
      code: "invalid_type",
      input: E,
      inst: e
    }), w);
  };
});
function Ie(e, o, t, n) {
  for (const a of e)
    if (a.issues.length === 0)
      return o.value = a.value, o;
  const r = e.filter((a) => !G(a));
  return r.length === 1 ? (o.value = r[0].value, r[0]) : (o.issues.push({
    code: "invalid_union",
    input: o.value,
    inst: t,
    errors: e.map((a) => a.issues.map((i) => L(i, n, F())))
  }), o);
}
const bn = /* @__PURE__ */ d("$ZodUnion", (e, o) => {
  Z.init(e, o), $(e._zod, "optin", () => o.options.some((r) => r._zod.optin === "optional") ? "optional" : void 0), $(e._zod, "optout", () => o.options.some((r) => r._zod.optout === "optional") ? "optional" : void 0), $(e._zod, "values", () => {
    if (o.options.every((r) => r._zod.values))
      return new Set(o.options.flatMap((r) => Array.from(r._zod.values)));
  }), $(e._zod, "pattern", () => {
    if (o.options.every((r) => r._zod.pattern)) {
      const r = o.options.map((a) => a._zod.pattern);
      return new RegExp(`^(${r.map((a) => ze(a.source)).join("|")})$`);
    }
  });
  const t = o.options.length === 1, n = o.options[0]._zod.run;
  e._zod.parse = (r, a) => {
    if (t)
      return n(r, a);
    let i = !1;
    const l = [];
    for (const u of o.options) {
      const p = u._zod.run({
        value: r.value,
        issues: []
      }, a);
      if (p instanceof Promise)
        l.push(p), i = !0;
      else {
        if (p.issues.length === 0)
          return p;
        l.push(p);
      }
    }
    return i ? Promise.all(l).then((u) => Ie(u, r, e, a)) : Ie(l, r, e, a);
  };
}), yn = /* @__PURE__ */ d("$ZodIntersection", (e, o) => {
  Z.init(e, o), e._zod.parse = (t, n) => {
    const r = t.value, a = o.left._zod.run({ value: r, issues: [] }, n), i = o.right._zod.run({ value: r, issues: [] }, n);
    return a instanceof Promise || i instanceof Promise ? Promise.all([a, i]).then(([u, p]) => Ne(t, u, p)) : Ne(t, a, i);
  };
});
function _e(e, o) {
  if (e === o)
    return { valid: !0, data: e };
  if (e instanceof Date && o instanceof Date && +e == +o)
    return { valid: !0, data: e };
  if (K(e) && K(o)) {
    const t = Object.keys(o), n = Object.keys(e).filter((a) => t.indexOf(a) !== -1), r = { ...e, ...o };
    for (const a of n) {
      const i = _e(e[a], o[a]);
      if (!i.valid)
        return {
          valid: !1,
          mergeErrorPath: [a, ...i.mergeErrorPath]
        };
      r[a] = i.data;
    }
    return { valid: !0, data: r };
  }
  if (Array.isArray(e) && Array.isArray(o)) {
    if (e.length !== o.length)
      return { valid: !1, mergeErrorPath: [] };
    const t = [];
    for (let n = 0; n < e.length; n++) {
      const r = e[n], a = o[n], i = _e(r, a);
      if (!i.valid)
        return {
          valid: !1,
          mergeErrorPath: [n, ...i.mergeErrorPath]
        };
      t.push(i.data);
    }
    return { valid: !0, data: t };
  }
  return { valid: !1, mergeErrorPath: [] };
}
function Ne(e, o, t) {
  const n = /* @__PURE__ */ new Map();
  let r;
  for (const l of o.issues)
    if (l.code === "unrecognized_keys") {
      r ?? (r = l);
      for (const u of l.keys)
        n.has(u) || n.set(u, {}), n.get(u).l = !0;
    } else
      e.issues.push(l);
  for (const l of t.issues)
    if (l.code === "unrecognized_keys")
      for (const u of l.keys)
        n.has(u) || n.set(u, {}), n.get(u).r = !0;
    else
      e.issues.push(l);
  const a = [...n].filter(([, l]) => l.l && l.r).map(([l]) => l);
  if (a.length && r && e.issues.push({ ...r, keys: a }), G(e))
    return e;
  const i = _e(o.value, t.value);
  if (!i.valid)
    throw new Error(`Unmergable intersection. Error path: ${JSON.stringify(i.mergeErrorPath)}`);
  return e.value = i.data, e;
}
const zn = /* @__PURE__ */ d("$ZodTuple", (e, o) => {
  Z.init(e, o);
  const t = o.items;
  e._zod.parse = (n, r) => {
    const a = n.value;
    if (!Array.isArray(a))
      return n.issues.push({
        input: a,
        inst: e,
        expected: "tuple",
        code: "invalid_type"
      }), n;
    n.value = [];
    const i = [], l = [...t].reverse().findIndex((h) => h._zod.optin !== "optional"), u = l === -1 ? 0 : t.length - l;
    if (!o.rest) {
      const h = a.length > t.length, _ = a.length < u - 1;
      if (h || _)
        return n.issues.push({
          ...h ? { code: "too_big", maximum: t.length, inclusive: !0 } : { code: "too_small", minimum: t.length },
          input: a,
          inst: e,
          origin: "array"
        }), n;
    }
    let p = -1;
    for (const h of t) {
      if (p++, p >= a.length && p >= u)
        continue;
      const _ = h._zod.run({
        value: a[p],
        issues: []
      }, r);
      _ instanceof Promise ? i.push(_.then((w) => te(w, n, p))) : te(_, n, p);
    }
    if (o.rest) {
      const h = a.slice(t.length);
      for (const _ of h) {
        p++;
        const w = o.rest._zod.run({
          value: _,
          issues: []
        }, r);
        w instanceof Promise ? i.push(w.then((S) => te(S, n, p))) : te(w, n, p);
      }
    }
    return i.length ? Promise.all(i).then(() => n) : n;
  };
});
function te(e, o, t) {
  e.issues.length && o.issues.push(...V(t, e.issues)), o.value[t] = e.value;
}
const kn = /* @__PURE__ */ d("$ZodRecord", (e, o) => {
  Z.init(e, o), e._zod.parse = (t, n) => {
    const r = t.value;
    if (!K(r))
      return t.issues.push({
        expected: "record",
        code: "invalid_type",
        input: r,
        inst: e
      }), t;
    const a = [], i = o.keyType._zod.values;
    if (i) {
      t.value = {};
      const l = /* @__PURE__ */ new Set();
      for (const p of i)
        if (typeof p == "string" || typeof p == "number" || typeof p == "symbol") {
          l.add(typeof p == "number" ? p.toString() : p);
          const h = o.valueType._zod.run({ value: r[p], issues: [] }, n);
          h instanceof Promise ? a.push(h.then((_) => {
            _.issues.length && t.issues.push(...V(p, _.issues)), t.value[p] = _.value;
          })) : (h.issues.length && t.issues.push(...V(p, h.issues)), t.value[p] = h.value);
        }
      let u;
      for (const p in r)
        l.has(p) || (u = u ?? [], u.push(p));
      u && u.length > 0 && t.issues.push({
        code: "unrecognized_keys",
        input: r,
        inst: e,
        keys: u
      });
    } else {
      t.value = {};
      for (const l of Reflect.ownKeys(r)) {
        if (l === "__proto__")
          continue;
        let u = o.keyType._zod.run({ value: l, issues: [] }, n);
        if (u instanceof Promise)
          throw new Error("Async schemas not supported in object keys currently");
        if (typeof l == "string" && ro.test(l) && u.issues.length) {
          const _ = o.keyType._zod.run({ value: Number(l), issues: [] }, n);
          if (_ instanceof Promise)
            throw new Error("Async schemas not supported in object keys currently");
          _.issues.length === 0 && (u = _);
        }
        if (u.issues.length) {
          o.mode === "loose" ? t.value[l] = r[l] : t.issues.push({
            code: "invalid_key",
            origin: "record",
            issues: u.issues.map((_) => L(_, n, F())),
            input: l,
            path: [l],
            inst: e
          });
          continue;
        }
        const h = o.valueType._zod.run({ value: r[l], issues: [] }, n);
        h instanceof Promise ? a.push(h.then((_) => {
          _.issues.length && t.issues.push(...V(l, _.issues)), t.value[u.value] = _.value;
        })) : (h.issues.length && t.issues.push(...V(l, h.issues)), t.value[u.value] = h.value);
      }
    }
    return a.length ? Promise.all(a).then(() => t) : t;
  };
}), wn = /* @__PURE__ */ d("$ZodEnum", (e, o) => {
  Z.init(e, o);
  const t = Ke(o.entries), n = new Set(t);
  e._zod.values = n, e._zod.pattern = new RegExp(`^(${t.filter((r) => No.has(typeof r)).map((r) => typeof r == "string" ? ue(r) : r.toString()).join("|")})$`), e._zod.parse = (r, a) => {
    const i = r.value;
    return n.has(i) || r.issues.push({
      code: "invalid_value",
      values: t,
      input: i,
      inst: e
    }), r;
  };
}), Sn = /* @__PURE__ */ d("$ZodTransform", (e, o) => {
  Z.init(e, o), e._zod.parse = (t, n) => {
    if (n.direction === "backward")
      throw new Ge(e.constructor.name);
    const r = o.transform(t.value, t);
    if (n.async)
      return (r instanceof Promise ? r : Promise.resolve(r)).then((i) => (t.value = i, t));
    if (r instanceof Promise)
      throw new q();
    return t.value = r, t;
  };
});
function Ee(e, o) {
  return e.issues.length && o === void 0 ? { issues: [], value: void 0 } : e;
}
const fo = /* @__PURE__ */ d("$ZodOptional", (e, o) => {
  Z.init(e, o), e._zod.optin = "optional", e._zod.optout = "optional", $(e._zod, "values", () => o.innerType._zod.values ? /* @__PURE__ */ new Set([...o.innerType._zod.values, void 0]) : void 0), $(e._zod, "pattern", () => {
    const t = o.innerType._zod.pattern;
    return t ? new RegExp(`^(${ze(t.source)})?$`) : void 0;
  }), e._zod.parse = (t, n) => {
    if (o.innerType._zod.optin === "optional") {
      const r = o.innerType._zod.run(t, n);
      return r instanceof Promise ? r.then((a) => Ee(a, t.value)) : Ee(r, t.value);
    }
    return t.value === void 0 ? t : o.innerType._zod.run(t, n);
  };
}), $n = /* @__PURE__ */ d("$ZodExactOptional", (e, o) => {
  fo.init(e, o), $(e._zod, "values", () => o.innerType._zod.values), $(e._zod, "pattern", () => o.innerType._zod.pattern), e._zod.parse = (t, n) => o.innerType._zod.run(t, n);
}), Zn = /* @__PURE__ */ d("$ZodNullable", (e, o) => {
  Z.init(e, o), $(e._zod, "optin", () => o.innerType._zod.optin), $(e._zod, "optout", () => o.innerType._zod.optout), $(e._zod, "pattern", () => {
    const t = o.innerType._zod.pattern;
    return t ? new RegExp(`^(${ze(t.source)}|null)$`) : void 0;
  }), $(e._zod, "values", () => o.innerType._zod.values ? /* @__PURE__ */ new Set([...o.innerType._zod.values, null]) : void 0), e._zod.parse = (t, n) => t.value === null ? t : o.innerType._zod.run(t, n);
}), Pn = /* @__PURE__ */ d("$ZodDefault", (e, o) => {
  Z.init(e, o), e._zod.optin = "optional", $(e._zod, "values", () => o.innerType._zod.values), e._zod.parse = (t, n) => {
    if (n.direction === "backward")
      return o.innerType._zod.run(t, n);
    if (t.value === void 0)
      return t.value = o.defaultValue, t;
    const r = o.innerType._zod.run(t, n);
    return r instanceof Promise ? r.then((a) => Ce(a, o)) : Ce(r, o);
  };
});
function Ce(e, o) {
  return e.value === void 0 && (e.value = o.defaultValue), e;
}
const Tn = /* @__PURE__ */ d("$ZodPrefault", (e, o) => {
  Z.init(e, o), e._zod.optin = "optional", $(e._zod, "values", () => o.innerType._zod.values), e._zod.parse = (t, n) => (n.direction === "backward" || t.value === void 0 && (t.value = o.defaultValue), o.innerType._zod.run(t, n));
}), On = /* @__PURE__ */ d("$ZodNonOptional", (e, o) => {
  Z.init(e, o), $(e._zod, "values", () => {
    const t = o.innerType._zod.values;
    return t ? new Set([...t].filter((n) => n !== void 0)) : void 0;
  }), e._zod.parse = (t, n) => {
    const r = o.innerType._zod.run(t, n);
    return r instanceof Promise ? r.then((a) => Ae(a, e)) : Ae(r, e);
  };
});
function Ae(e, o) {
  return !e.issues.length && e.value === void 0 && e.issues.push({
    code: "invalid_type",
    expected: "nonoptional",
    input: e.value,
    inst: o
  }), e;
}
const In = /* @__PURE__ */ d("$ZodCatch", (e, o) => {
  Z.init(e, o), $(e._zod, "optin", () => o.innerType._zod.optin), $(e._zod, "optout", () => o.innerType._zod.optout), $(e._zod, "values", () => o.innerType._zod.values), e._zod.parse = (t, n) => {
    if (n.direction === "backward")
      return o.innerType._zod.run(t, n);
    const r = o.innerType._zod.run(t, n);
    return r instanceof Promise ? r.then((a) => (t.value = a.value, a.issues.length && (t.value = o.catchValue({
      ...t,
      error: {
        issues: a.issues.map((i) => L(i, n, F()))
      },
      input: t.value
    }), t.issues = []), t)) : (t.value = r.value, r.issues.length && (t.value = o.catchValue({
      ...t,
      error: {
        issues: r.issues.map((a) => L(a, n, F()))
      },
      input: t.value
    }), t.issues = []), t);
  };
}), Nn = /* @__PURE__ */ d("$ZodPipe", (e, o) => {
  Z.init(e, o), $(e._zod, "values", () => o.in._zod.values), $(e._zod, "optin", () => o.in._zod.optin), $(e._zod, "optout", () => o.out._zod.optout), $(e._zod, "propValues", () => o.in._zod.propValues), e._zod.parse = (t, n) => {
    if (n.direction === "backward") {
      const a = o.out._zod.run(t, n);
      return a instanceof Promise ? a.then((i) => ne(i, o.in, n)) : ne(a, o.in, n);
    }
    const r = o.in._zod.run(t, n);
    return r instanceof Promise ? r.then((a) => ne(a, o.out, n)) : ne(r, o.out, n);
  };
});
function ne(e, o, t) {
  return e.issues.length ? (e.aborted = !0, e) : o._zod.run({ value: e.value, issues: e.issues }, t);
}
const En = /* @__PURE__ */ d("$ZodReadonly", (e, o) => {
  Z.init(e, o), $(e._zod, "propValues", () => o.innerType._zod.propValues), $(e._zod, "values", () => o.innerType._zod.values), $(e._zod, "optin", () => o.innerType?._zod?.optin), $(e._zod, "optout", () => o.innerType?._zod?.optout), e._zod.parse = (t, n) => {
    if (n.direction === "backward")
      return o.innerType._zod.run(t, n);
    const r = o.innerType._zod.run(t, n);
    return r instanceof Promise ? r.then(xe) : xe(r);
  };
});
function xe(e) {
  return e.value = Object.freeze(e.value), e;
}
const Cn = /* @__PURE__ */ d("$ZodLazy", (e, o) => {
  Z.init(e, o), $(e._zod, "innerType", () => o.getter()), $(e._zod, "pattern", () => e._zod.innerType?._zod?.pattern), $(e._zod, "propValues", () => e._zod.innerType?._zod?.propValues), $(e._zod, "optin", () => e._zod.innerType?._zod?.optin ?? void 0), $(e._zod, "optout", () => e._zod.innerType?._zod?.optout ?? void 0), e._zod.parse = (t, n) => e._zod.innerType._zod.run(t, n);
}), An = /* @__PURE__ */ d("$ZodCustom", (e, o) => {
  x.init(e, o), Z.init(e, o), e._zod.parse = (t, n) => t, e._zod.check = (t) => {
    const n = t.value, r = o.fn(n);
    if (r instanceof Promise)
      return r.then((a) => je(a, t, n, e));
    je(r, t, n, e);
  };
});
function je(e, o, t, n) {
  if (!e) {
    const r = {
      code: "custom",
      input: t,
      inst: n,
      // incorporates params.error into issue reporting
      path: [...n._zod.def.path ?? []],
      // incorporates params.error into issue reporting
      continue: !n._zod.def.abort
      // params: inst._zod.def.params,
    };
    n._zod.def.params && (r.params = n._zod.def.params), o.issues.push(Q(r));
  }
}
var Re;
class xn {
  constructor() {
    this._map = /* @__PURE__ */ new WeakMap(), this._idmap = /* @__PURE__ */ new Map();
  }
  add(o, ...t) {
    const n = t[0];
    return this._map.set(o, n), n && typeof n == "object" && "id" in n && this._idmap.set(n.id, o), this;
  }
  clear() {
    return this._map = /* @__PURE__ */ new WeakMap(), this._idmap = /* @__PURE__ */ new Map(), this;
  }
  remove(o) {
    const t = this._map.get(o);
    return t && typeof t == "object" && "id" in t && this._idmap.delete(t.id), this._map.delete(o), this;
  }
  get(o) {
    const t = o._zod.parent;
    if (t) {
      const n = { ...this.get(t) ?? {} };
      delete n.id;
      const r = { ...n, ...this._map.get(o) };
      return Object.keys(r).length ? r : void 0;
    }
    return this._map.get(o);
  }
  has(o) {
    return this._map.has(o);
  }
}
function jn() {
  return new xn();
}
(Re = globalThis).__zod_globalRegistry ?? (Re.__zod_globalRegistry = jn());
const Y = globalThis.__zod_globalRegistry;
// @__NO_SIDE_EFFECTS__
function Rn(e, o) {
  return new e({
    type: "string",
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Dn(e, o) {
  return new e({
    type: "string",
    format: "email",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function De(e, o) {
  return new e({
    type: "string",
    format: "guid",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Un(e, o) {
  return new e({
    type: "string",
    format: "uuid",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Fn(e, o) {
  return new e({
    type: "string",
    format: "uuid",
    check: "string_format",
    abort: !1,
    version: "v4",
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Ln(e, o) {
  return new e({
    type: "string",
    format: "uuid",
    check: "string_format",
    abort: !1,
    version: "v6",
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Mn(e, o) {
  return new e({
    type: "string",
    format: "uuid",
    check: "string_format",
    abort: !1,
    version: "v7",
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Jn(e, o) {
  return new e({
    type: "string",
    format: "url",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Vn(e, o) {
  return new e({
    type: "string",
    format: "emoji",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Bn(e, o) {
  return new e({
    type: "string",
    format: "nanoid",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Wn(e, o) {
  return new e({
    type: "string",
    format: "cuid",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Gn(e, o) {
  return new e({
    type: "string",
    format: "cuid2",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function qn(e, o) {
  return new e({
    type: "string",
    format: "ulid",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Kn(e, o) {
  return new e({
    type: "string",
    format: "xid",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Xn(e, o) {
  return new e({
    type: "string",
    format: "ksuid",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Yn(e, o) {
  return new e({
    type: "string",
    format: "ipv4",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Hn(e, o) {
  return new e({
    type: "string",
    format: "ipv6",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Qn(e, o) {
  return new e({
    type: "string",
    format: "cidrv4",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function er(e, o) {
  return new e({
    type: "string",
    format: "cidrv6",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function or(e, o) {
  return new e({
    type: "string",
    format: "base64",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function tr(e, o) {
  return new e({
    type: "string",
    format: "base64url",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function nr(e, o) {
  return new e({
    type: "string",
    format: "e164",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function rr(e, o) {
  return new e({
    type: "string",
    format: "jwt",
    check: "string_format",
    abort: !1,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function ir(e, o) {
  return new e({
    type: "string",
    format: "datetime",
    check: "string_format",
    offset: !1,
    local: !1,
    precision: null,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function ar(e, o) {
  return new e({
    type: "string",
    format: "date",
    check: "string_format",
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function sr(e, o) {
  return new e({
    type: "string",
    format: "time",
    check: "string_format",
    precision: null,
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function cr(e, o) {
  return new e({
    type: "string",
    format: "duration",
    check: "string_format",
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function lr(e, o) {
  return new e({
    type: "number",
    checks: [],
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function ur(e, o) {
  return new e({
    type: "number",
    check: "number_format",
    abort: !1,
    format: "safeint",
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function pr(e, o) {
  return new e({
    type: "boolean",
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function dr(e, o) {
  return new e({
    type: "null",
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function fr(e) {
  return new e({
    type: "unknown"
  });
}
// @__NO_SIDE_EFFECTS__
function hr(e, o) {
  return new e({
    type: "never",
    ...k(o)
  });
}
// @__NO_SIDE_EFFECTS__
function Ue(e, o) {
  return new ao({
    check: "less_than",
    ...k(o),
    value: e,
    inclusive: !1
  });
}
// @__NO_SIDE_EFFECTS__
function he(e, o) {
  return new ao({
    check: "less_than",
    ...k(o),
    value: e,
    inclusive: !0
  });
}
// @__NO_SIDE_EFFECTS__
function Fe(e, o) {
  return new so({
    check: "greater_than",
    ...k(o),
    value: e,
    inclusive: !1
  });
}
// @__NO_SIDE_EFFECTS__
function me(e, o) {
  return new so({
    check: "greater_than",
    ...k(o),
    value: e,
    inclusive: !0
  });
}
// @__NO_SIDE_EFFECTS__
function Le(e, o) {
  return new $t({
    check: "multiple_of",
    ...k(o),
    value: e
  });
}
// @__NO_SIDE_EFFECTS__
function ho(e, o) {
  return new Pt({
    check: "max_length",
    ...k(o),
    maximum: e
  });
}
// @__NO_SIDE_EFFECTS__
function ce(e, o) {
  return new Tt({
    check: "min_length",
    ...k(o),
    minimum: e
  });
}
// @__NO_SIDE_EFFECTS__
function mo(e, o) {
  return new Ot({
    check: "length_equals",
    ...k(o),
    length: e
  });
}
// @__NO_SIDE_EFFECTS__
function mr(e, o) {
  return new It({
    check: "string_format",
    format: "regex",
    ...k(o),
    pattern: e
  });
}
// @__NO_SIDE_EFFECTS__
function gr(e) {
  return new Nt({
    check: "string_format",
    format: "lowercase",
    ...k(e)
  });
}
// @__NO_SIDE_EFFECTS__
function _r(e) {
  return new Et({
    check: "string_format",
    format: "uppercase",
    ...k(e)
  });
}
// @__NO_SIDE_EFFECTS__
function vr(e, o) {
  return new Ct({
    check: "string_format",
    format: "includes",
    ...k(o),
    includes: e
  });
}
// @__NO_SIDE_EFFECTS__
function br(e, o) {
  return new At({
    check: "string_format",
    format: "starts_with",
    ...k(o),
    prefix: e
  });
}
// @__NO_SIDE_EFFECTS__
function yr(e, o) {
  return new xt({
    check: "string_format",
    format: "ends_with",
    ...k(o),
    suffix: e
  });
}
// @__NO_SIDE_EFFECTS__
function X(e) {
  return new jt({
    check: "overwrite",
    tx: e
  });
}
// @__NO_SIDE_EFFECTS__
function zr(e) {
  return /* @__PURE__ */ X((o) => o.normalize(e));
}
// @__NO_SIDE_EFFECTS__
function kr() {
  return /* @__PURE__ */ X((e) => e.trim());
}
// @__NO_SIDE_EFFECTS__
function wr() {
  return /* @__PURE__ */ X((e) => e.toLowerCase());
}
// @__NO_SIDE_EFFECTS__
function Sr() {
  return /* @__PURE__ */ X((e) => e.toUpperCase());
}
// @__NO_SIDE_EFFECTS__
function $r() {
  return /* @__PURE__ */ X((e) => Oo(e));
}
// @__NO_SIDE_EFFECTS__
function Zr(e, o, t) {
  return new e({
    type: "array",
    element: o,
    // get element() {
    //   return element;
    // },
    ...k(t)
  });
}
// @__NO_SIDE_EFFECTS__
function Pr(e, o, t) {
  return new e({
    type: "custom",
    check: "custom",
    fn: o,
    ...k(t)
  });
}
// @__NO_SIDE_EFFECTS__
function Tr(e) {
  const o = /* @__PURE__ */ Or((t) => (t.addIssue = (n) => {
    if (typeof n == "string")
      t.issues.push(Q(n, t.value, o._zod.def));
    else {
      const r = n;
      r.fatal && (r.continue = !1), r.code ?? (r.code = "custom"), r.input ?? (r.input = t.value), r.inst ?? (r.inst = o), r.continue ?? (r.continue = !o._zod.def.abort), t.issues.push(Q(r));
    }
  }, e(t.value, t)));
  return o;
}
// @__NO_SIDE_EFFECTS__
function Or(e, o) {
  const t = new x({
    check: "custom",
    ...k(o)
  });
  return t._zod.check = e, t;
}
function go(e) {
  let o = e?.target ?? "draft-2020-12";
  return o === "draft-4" && (o = "draft-04"), o === "draft-7" && (o = "draft-07"), {
    processors: e.processors ?? {},
    metadataRegistry: e?.metadata ?? Y,
    target: o,
    unrepresentable: e?.unrepresentable ?? "throw",
    override: e?.override ?? (() => {
    }),
    io: e?.io ?? "output",
    counter: 0,
    seen: /* @__PURE__ */ new Map(),
    cycles: e?.cycles ?? "ref",
    reused: e?.reused ?? "inline",
    external: e?.external ?? void 0
  };
}
function O(e, o, t = { path: [], schemaPath: [] }) {
  var n;
  const r = e._zod.def, a = o.seen.get(e);
  if (a)
    return a.count++, t.schemaPath.includes(e) && (a.cycle = t.path), a.schema;
  const i = { schema: {}, count: 1, cycle: void 0, path: t.path };
  o.seen.set(e, i);
  const l = e._zod.toJSONSchema?.();
  if (l)
    i.schema = l;
  else {
    const h = {
      ...t,
      schemaPath: [...t.schemaPath, e],
      path: t.path
    };
    if (e._zod.processJSONSchema)
      e._zod.processJSONSchema(o, i.schema, h);
    else {
      const w = i.schema, S = o.processors[r.type];
      if (!S)
        throw new Error(`[toJSONSchema]: Non-representable type encountered: ${r.type}`);
      S(e, o, w, h);
    }
    const _ = e._zod.parent;
    _ && (i.ref || (i.ref = _), O(_, o, h), o.seen.get(_).isParent = !0);
  }
  const u = o.metadataRegistry.get(e);
  return u && Object.assign(i.schema, u), o.io === "input" && C(e) && (delete i.schema.examples, delete i.schema.default), o.io === "input" && i.schema._prefault && ((n = i.schema).default ?? (n.default = i.schema._prefault)), delete i.schema._prefault, o.seen.get(e).schema;
}
function _o(e, o) {
  const t = e.seen.get(o);
  if (!t)
    throw new Error("Unprocessed schema. This is a bug in Zod.");
  const n = /* @__PURE__ */ new Map();
  for (const i of e.seen.entries()) {
    const l = e.metadataRegistry.get(i[0])?.id;
    if (l) {
      const u = n.get(l);
      if (u && u !== i[0])
        throw new Error(`Duplicate schema id "${l}" detected during JSON Schema conversion. Two different schemas cannot share the same id when converted together.`);
      n.set(l, i[0]);
    }
  }
  const r = (i) => {
    const l = e.target === "draft-2020-12" ? "$defs" : "definitions";
    if (e.external) {
      const _ = e.external.registry.get(i[0])?.id, w = e.external.uri ?? ((E) => E);
      if (_)
        return { ref: w(_) };
      const S = i[1].defId ?? i[1].schema.id ?? `schema${e.counter++}`;
      return i[1].defId = S, { defId: S, ref: `${w("__shared")}#/${l}/${S}` };
    }
    if (i[1] === t)
      return { ref: "#" };
    const p = `#/${l}/`, h = i[1].schema.id ?? `__schema${e.counter++}`;
    return { defId: h, ref: p + h };
  }, a = (i) => {
    if (i[1].schema.$ref)
      return;
    const l = i[1], { ref: u, defId: p } = r(i);
    l.def = { ...l.schema }, p && (l.defId = p);
    const h = l.schema;
    for (const _ in h)
      delete h[_];
    h.$ref = u;
  };
  if (e.cycles === "throw")
    for (const i of e.seen.entries()) {
      const l = i[1];
      if (l.cycle)
        throw new Error(`Cycle detected: #/${l.cycle?.join("/")}/<root>

Set the \`cycles\` parameter to \`"ref"\` to resolve cyclical schemas with defs.`);
    }
  for (const i of e.seen.entries()) {
    const l = i[1];
    if (o === i[0]) {
      a(i);
      continue;
    }
    if (e.external) {
      const p = e.external.registry.get(i[0])?.id;
      if (o !== i[0] && p) {
        a(i);
        continue;
      }
    }
    if (e.metadataRegistry.get(i[0])?.id) {
      a(i);
      continue;
    }
    if (l.cycle) {
      a(i);
      continue;
    }
    if (l.count > 1 && e.reused === "ref") {
      a(i);
      continue;
    }
  }
}
function vo(e, o) {
  const t = e.seen.get(o);
  if (!t)
    throw new Error("Unprocessed schema. This is a bug in Zod.");
  const n = (i) => {
    const l = e.seen.get(i);
    if (l.ref === null)
      return;
    const u = l.def ?? l.schema, p = { ...u }, h = l.ref;
    if (l.ref = null, h) {
      n(h);
      const w = e.seen.get(h), S = w.schema;
      if (S.$ref && (e.target === "draft-07" || e.target === "draft-04" || e.target === "openapi-3.0") ? (u.allOf = u.allOf ?? [], u.allOf.push(S)) : Object.assign(u, S), Object.assign(u, p), i._zod.parent === h)
        for (const N in u)
          N === "$ref" || N === "allOf" || N in p || delete u[N];
      if (S.$ref && w.def)
        for (const N in u)
          N === "$ref" || N === "allOf" || N in w.def && JSON.stringify(u[N]) === JSON.stringify(w.def[N]) && delete u[N];
    }
    const _ = i._zod.parent;
    if (_ && _ !== h) {
      n(_);
      const w = e.seen.get(_);
      if (w?.schema.$ref && (u.$ref = w.schema.$ref, w.def))
        for (const S in u)
          S === "$ref" || S === "allOf" || S in w.def && JSON.stringify(u[S]) === JSON.stringify(w.def[S]) && delete u[S];
    }
    e.override({
      zodSchema: i,
      jsonSchema: u,
      path: l.path ?? []
    });
  };
  for (const i of [...e.seen.entries()].reverse())
    n(i[0]);
  const r = {};
  if (e.target === "draft-2020-12" ? r.$schema = "https://json-schema.org/draft/2020-12/schema" : e.target === "draft-07" ? r.$schema = "http://json-schema.org/draft-07/schema#" : e.target === "draft-04" ? r.$schema = "http://json-schema.org/draft-04/schema#" : e.target, e.external?.uri) {
    const i = e.external.registry.get(o)?.id;
    if (!i)
      throw new Error("Schema is missing an `id` property");
    r.$id = e.external.uri(i);
  }
  Object.assign(r, t.def ?? t.schema);
  const a = e.external?.defs ?? {};
  for (const i of e.seen.entries()) {
    const l = i[1];
    l.def && l.defId && (a[l.defId] = l.def);
  }
  e.external || Object.keys(a).length > 0 && (e.target === "draft-2020-12" ? r.$defs = a : r.definitions = a);
  try {
    const i = JSON.parse(JSON.stringify(r));
    return Object.defineProperty(i, "~standard", {
      value: {
        ...o["~standard"],
        jsonSchema: {
          input: le(o, "input", e.processors),
          output: le(o, "output", e.processors)
        }
      },
      enumerable: !1,
      writable: !1
    }), i;
  } catch {
    throw new Error("Error converting schema to JSON.");
  }
}
function C(e, o) {
  const t = o ?? { seen: /* @__PURE__ */ new Set() };
  if (t.seen.has(e))
    return !1;
  t.seen.add(e);
  const n = e._zod.def;
  if (n.type === "transform")
    return !0;
  if (n.type === "array")
    return C(n.element, t);
  if (n.type === "set")
    return C(n.valueType, t);
  if (n.type === "lazy")
    return C(n.getter(), t);
  if (n.type === "promise" || n.type === "optional" || n.type === "nonoptional" || n.type === "nullable" || n.type === "readonly" || n.type === "default" || n.type === "prefault")
    return C(n.innerType, t);
  if (n.type === "intersection")
    return C(n.left, t) || C(n.right, t);
  if (n.type === "record" || n.type === "map")
    return C(n.keyType, t) || C(n.valueType, t);
  if (n.type === "pipe")
    return C(n.in, t) || C(n.out, t);
  if (n.type === "object") {
    for (const r in n.shape)
      if (C(n.shape[r], t))
        return !0;
    return !1;
  }
  if (n.type === "union") {
    for (const r of n.options)
      if (C(r, t))
        return !0;
    return !1;
  }
  if (n.type === "tuple") {
    for (const r of n.items)
      if (C(r, t))
        return !0;
    return !!(n.rest && C(n.rest, t));
  }
  return !1;
}
const Ir = (e, o = {}) => (t) => {
  const n = go({ ...t, processors: o });
  return O(e, n), _o(n, e), vo(n, e);
}, le = (e, o, t = {}) => (n) => {
  const { libraryOptions: r, target: a } = n ?? {}, i = go({ ...r ?? {}, target: a, io: o, processors: t });
  return O(e, i), _o(i, e), vo(i, e);
}, Nr = {
  guid: "uuid",
  url: "uri",
  datetime: "date-time",
  json_string: "json-string",
  regex: ""
  // do not set
}, Er = (e, o, t, n) => {
  const r = t;
  r.type = "string";
  const { minimum: a, maximum: i, format: l, patterns: u, contentEncoding: p } = e._zod.bag;
  if (typeof a == "number" && (r.minLength = a), typeof i == "number" && (r.maxLength = i), l && (r.format = Nr[l] ?? l, r.format === "" && delete r.format, l === "time" && delete r.format), p && (r.contentEncoding = p), u && u.size > 0) {
    const h = [...u];
    h.length === 1 ? r.pattern = h[0].source : h.length > 1 && (r.allOf = [
      ...h.map((_) => ({
        ...o.target === "draft-07" || o.target === "draft-04" || o.target === "openapi-3.0" ? { type: "string" } : {},
        pattern: _.source
      }))
    ]);
  }
}, Cr = (e, o, t, n) => {
  const r = t, { minimum: a, maximum: i, format: l, multipleOf: u, exclusiveMaximum: p, exclusiveMinimum: h } = e._zod.bag;
  typeof l == "string" && l.includes("int") ? r.type = "integer" : r.type = "number", typeof h == "number" && (o.target === "draft-04" || o.target === "openapi-3.0" ? (r.minimum = h, r.exclusiveMinimum = !0) : r.exclusiveMinimum = h), typeof a == "number" && (r.minimum = a, typeof h == "number" && o.target !== "draft-04" && (h >= a ? delete r.minimum : delete r.exclusiveMinimum)), typeof p == "number" && (o.target === "draft-04" || o.target === "openapi-3.0" ? (r.maximum = p, r.exclusiveMaximum = !0) : r.exclusiveMaximum = p), typeof i == "number" && (r.maximum = i, typeof p == "number" && o.target !== "draft-04" && (p <= i ? delete r.maximum : delete r.exclusiveMaximum)), typeof u == "number" && (r.multipleOf = u);
}, Ar = (e, o, t, n) => {
  t.type = "boolean";
}, xr = (e, o, t, n) => {
  o.target === "openapi-3.0" ? (t.type = "string", t.nullable = !0, t.enum = [null]) : t.type = "null";
}, jr = (e, o, t, n) => {
  t.not = {};
}, Rr = (e, o, t, n) => {
}, Dr = (e, o, t, n) => {
  const r = e._zod.def, a = Ke(r.entries);
  a.every((i) => typeof i == "number") && (t.type = "number"), a.every((i) => typeof i == "string") && (t.type = "string"), t.enum = a;
}, Ur = (e, o, t, n) => {
  if (o.unrepresentable === "throw")
    throw new Error("Custom types cannot be represented in JSON Schema");
}, Fr = (e, o, t, n) => {
  if (o.unrepresentable === "throw")
    throw new Error("Transforms cannot be represented in JSON Schema");
}, Lr = (e, o, t, n) => {
  const r = t, a = e._zod.def, { minimum: i, maximum: l } = e._zod.bag;
  typeof i == "number" && (r.minItems = i), typeof l == "number" && (r.maxItems = l), r.type = "array", r.items = O(a.element, o, { ...n, path: [...n.path, "items"] });
}, Mr = (e, o, t, n) => {
  const r = t, a = e._zod.def;
  r.type = "object", r.properties = {};
  const i = a.shape;
  for (const p in i)
    r.properties[p] = O(i[p], o, {
      ...n,
      path: [...n.path, "properties", p]
    });
  const l = new Set(Object.keys(i)), u = new Set([...l].filter((p) => {
    const h = a.shape[p]._zod;
    return o.io === "input" ? h.optin === void 0 : h.optout === void 0;
  }));
  u.size > 0 && (r.required = Array.from(u)), a.catchall?._zod.def.type === "never" ? r.additionalProperties = !1 : a.catchall ? a.catchall && (r.additionalProperties = O(a.catchall, o, {
    ...n,
    path: [...n.path, "additionalProperties"]
  })) : o.io === "output" && (r.additionalProperties = !1);
}, Jr = (e, o, t, n) => {
  const r = e._zod.def, a = r.inclusive === !1, i = r.options.map((l, u) => O(l, o, {
    ...n,
    path: [...n.path, a ? "oneOf" : "anyOf", u]
  }));
  a ? t.oneOf = i : t.anyOf = i;
}, Vr = (e, o, t, n) => {
  const r = e._zod.def, a = O(r.left, o, {
    ...n,
    path: [...n.path, "allOf", 0]
  }), i = O(r.right, o, {
    ...n,
    path: [...n.path, "allOf", 1]
  }), l = (p) => "allOf" in p && Object.keys(p).length === 1, u = [
    ...l(a) ? a.allOf : [a],
    ...l(i) ? i.allOf : [i]
  ];
  t.allOf = u;
}, Br = (e, o, t, n) => {
  const r = t, a = e._zod.def;
  r.type = "array";
  const i = o.target === "draft-2020-12" ? "prefixItems" : "items", l = o.target === "draft-2020-12" || o.target === "openapi-3.0" ? "items" : "additionalItems", u = a.items.map((w, S) => O(w, o, {
    ...n,
    path: [...n.path, i, S]
  })), p = a.rest ? O(a.rest, o, {
    ...n,
    path: [...n.path, l, ...o.target === "openapi-3.0" ? [a.items.length] : []]
  }) : null;
  o.target === "draft-2020-12" ? (r.prefixItems = u, p && (r.items = p)) : o.target === "openapi-3.0" ? (r.items = {
    anyOf: u
  }, p && r.items.anyOf.push(p), r.minItems = u.length, p || (r.maxItems = u.length)) : (r.items = u, p && (r.additionalItems = p));
  const { minimum: h, maximum: _ } = e._zod.bag;
  typeof h == "number" && (r.minItems = h), typeof _ == "number" && (r.maxItems = _);
}, Wr = (e, o, t, n) => {
  const r = t, a = e._zod.def;
  r.type = "object";
  const i = a.keyType, u = i._zod.bag?.patterns;
  if (a.mode === "loose" && u && u.size > 0) {
    const h = O(a.valueType, o, {
      ...n,
      path: [...n.path, "patternProperties", "*"]
    });
    r.patternProperties = {};
    for (const _ of u)
      r.patternProperties[_.source] = h;
  } else
    (o.target === "draft-07" || o.target === "draft-2020-12") && (r.propertyNames = O(a.keyType, o, {
      ...n,
      path: [...n.path, "propertyNames"]
    })), r.additionalProperties = O(a.valueType, o, {
      ...n,
      path: [...n.path, "additionalProperties"]
    });
  const p = i._zod.values;
  if (p) {
    const h = [...p].filter((_) => typeof _ == "string" || typeof _ == "number");
    h.length > 0 && (r.required = h);
  }
}, Gr = (e, o, t, n) => {
  const r = e._zod.def, a = O(r.innerType, o, n), i = o.seen.get(e);
  o.target === "openapi-3.0" ? (i.ref = r.innerType, t.nullable = !0) : t.anyOf = [a, { type: "null" }];
}, qr = (e, o, t, n) => {
  const r = e._zod.def;
  O(r.innerType, o, n);
  const a = o.seen.get(e);
  a.ref = r.innerType;
}, Kr = (e, o, t, n) => {
  const r = e._zod.def;
  O(r.innerType, o, n);
  const a = o.seen.get(e);
  a.ref = r.innerType, t.default = JSON.parse(JSON.stringify(r.defaultValue));
}, Xr = (e, o, t, n) => {
  const r = e._zod.def;
  O(r.innerType, o, n);
  const a = o.seen.get(e);
  a.ref = r.innerType, o.io === "input" && (t._prefault = JSON.parse(JSON.stringify(r.defaultValue)));
}, Yr = (e, o, t, n) => {
  const r = e._zod.def;
  O(r.innerType, o, n);
  const a = o.seen.get(e);
  a.ref = r.innerType;
  let i;
  try {
    i = r.catchValue(void 0);
  } catch {
    throw new Error("Dynamic catch values are not supported in JSON Schema");
  }
  t.default = i;
}, Hr = (e, o, t, n) => {
  const r = e._zod.def, a = o.io === "input" ? r.in._zod.def.type === "transform" ? r.out : r.in : r.out;
  O(a, o, n);
  const i = o.seen.get(e);
  i.ref = a;
}, Qr = (e, o, t, n) => {
  const r = e._zod.def;
  O(r.innerType, o, n);
  const a = o.seen.get(e);
  a.ref = r.innerType, t.readOnly = !0;
}, bo = (e, o, t, n) => {
  const r = e._zod.def;
  O(r.innerType, o, n);
  const a = o.seen.get(e);
  a.ref = r.innerType;
}, ei = (e, o, t, n) => {
  const r = e._zod.innerType;
  O(r, o, n);
  const a = o.seen.get(e);
  a.ref = r;
}, oi = /* @__PURE__ */ d("ZodISODateTime", (e, o) => {
  Xt.init(e, o), I.init(e, o);
});
function ti(e) {
  return /* @__PURE__ */ ir(oi, e);
}
const ni = /* @__PURE__ */ d("ZodISODate", (e, o) => {
  Yt.init(e, o), I.init(e, o);
});
function ri(e) {
  return /* @__PURE__ */ ar(ni, e);
}
const ii = /* @__PURE__ */ d("ZodISOTime", (e, o) => {
  Ht.init(e, o), I.init(e, o);
});
function ai(e) {
  return /* @__PURE__ */ sr(ii, e);
}
const si = /* @__PURE__ */ d("ZodISODuration", (e, o) => {
  Qt.init(e, o), I.init(e, o);
});
function ci(e) {
  return /* @__PURE__ */ cr(si, e);
}
const yo = (e, o) => {
  Qe.init(e, o), e.name = "ZodError", Object.defineProperties(e, {
    format: {
      value: (t) => Mo(e, t)
      // enumerable: false,
    },
    flatten: {
      value: (t) => Lo(e, t)
      // enumerable: false,
    },
    addIssue: {
      value: (t) => {
        e.issues.push(t), e.message = JSON.stringify(e.issues, ge, 2);
      }
      // enumerable: false,
    },
    addIssues: {
      value: (t) => {
        e.issues.push(...t), e.message = JSON.stringify(e.issues, ge, 2);
      }
      // enumerable: false,
    },
    isEmpty: {
      get() {
        return e.issues.length === 0;
      }
      // enumerable: false,
    }
  });
}, li = d("ZodError", yo), j = d("ZodError", yo, {
  Parent: Error
}), ui = /* @__PURE__ */ we(j), pi = /* @__PURE__ */ Se(j), di = /* @__PURE__ */ pe(j), fi = /* @__PURE__ */ de(j), hi = /* @__PURE__ */ Bo(j), mi = /* @__PURE__ */ Wo(j), gi = /* @__PURE__ */ Go(j), _i = /* @__PURE__ */ qo(j), vi = /* @__PURE__ */ Ko(j), bi = /* @__PURE__ */ Xo(j), yi = /* @__PURE__ */ Yo(j), zi = /* @__PURE__ */ Ho(j), P = /* @__PURE__ */ d("ZodType", (e, o) => (Z.init(e, o), Object.assign(e["~standard"], {
  jsonSchema: {
    input: le(e, "input"),
    output: le(e, "output")
  }
}), e.toJSONSchema = Ir(e, {}), e.def = o, e.type = o.type, Object.defineProperty(e, "_def", { value: o }), e.check = (...t) => e.clone(M(o, {
  checks: [
    ...o.checks ?? [],
    ...t.map((n) => typeof n == "function" ? { _zod: { check: n, def: { check: "custom" }, onattach: [] } } : n)
  ]
}), {
  parent: !0
}), e.with = e.check, e.clone = (t, n) => J(e, t, n), e.brand = () => e, e.register = ((t, n) => (t.add(e, n), e)), e.parse = (t, n) => ui(e, t, n, { callee: e.parse }), e.safeParse = (t, n) => di(e, t, n), e.parseAsync = async (t, n) => pi(e, t, n, { callee: e.parseAsync }), e.safeParseAsync = async (t, n) => fi(e, t, n), e.spa = e.safeParseAsync, e.encode = (t, n) => hi(e, t, n), e.decode = (t, n) => mi(e, t, n), e.encodeAsync = async (t, n) => gi(e, t, n), e.decodeAsync = async (t, n) => _i(e, t, n), e.safeEncode = (t, n) => vi(e, t, n), e.safeDecode = (t, n) => bi(e, t, n), e.safeEncodeAsync = async (t, n) => yi(e, t, n), e.safeDecodeAsync = async (t, n) => zi(e, t, n), e.refine = (t, n) => e.check(_a(t, n)), e.superRefine = (t) => e.check(va(t)), e.overwrite = (t) => e.check(/* @__PURE__ */ X(t)), e.optional = () => Ve(e), e.exactOptional = () => ta(e), e.nullable = () => Be(e), e.nullish = () => Ve(Be(e)), e.nonoptional = (t) => ca(e, t), e.array = () => U(e), e.or = (t) => v([e, t]), e.and = (t) => Xi(e, t), e.transform = (t) => We(e, ea(t)), e.default = (t) => ia(e, t), e.prefault = (t) => sa(e, t), e.catch = (t) => ua(e, t), e.pipe = (t) => We(e, t), e.readonly = () => fa(e), e.describe = (t) => {
  const n = e.clone();
  return Y.add(n, { description: t }), n;
}, Object.defineProperty(e, "description", {
  get() {
    return Y.get(e)?.description;
  },
  configurable: !0
}), e.meta = (...t) => {
  if (t.length === 0)
    return Y.get(e);
  const n = e.clone();
  return Y.add(n, t[0]), n;
}, e.isOptional = () => e.safeParse(void 0).success, e.isNullable = () => e.safeParse(null).success, e.apply = (t) => t(e), e)), zo = /* @__PURE__ */ d("_ZodString", (e, o) => {
  $e.init(e, o), P.init(e, o), e._zod.processJSONSchema = (n, r, a) => Er(e, n, r);
  const t = e._zod.bag;
  e.format = t.format ?? null, e.minLength = t.minimum ?? null, e.maxLength = t.maximum ?? null, e.regex = (...n) => e.check(/* @__PURE__ */ mr(...n)), e.includes = (...n) => e.check(/* @__PURE__ */ vr(...n)), e.startsWith = (...n) => e.check(/* @__PURE__ */ br(...n)), e.endsWith = (...n) => e.check(/* @__PURE__ */ yr(...n)), e.min = (...n) => e.check(/* @__PURE__ */ ce(...n)), e.max = (...n) => e.check(/* @__PURE__ */ ho(...n)), e.length = (...n) => e.check(/* @__PURE__ */ mo(...n)), e.nonempty = (...n) => e.check(/* @__PURE__ */ ce(1, ...n)), e.lowercase = (n) => e.check(/* @__PURE__ */ gr(n)), e.uppercase = (n) => e.check(/* @__PURE__ */ _r(n)), e.trim = () => e.check(/* @__PURE__ */ kr()), e.normalize = (...n) => e.check(/* @__PURE__ */ zr(...n)), e.toLowerCase = () => e.check(/* @__PURE__ */ wr()), e.toUpperCase = () => e.check(/* @__PURE__ */ Sr()), e.slugify = () => e.check(/* @__PURE__ */ $r());
}), ki = /* @__PURE__ */ d("ZodString", (e, o) => {
  $e.init(e, o), zo.init(e, o), e.email = (t) => e.check(/* @__PURE__ */ Dn(wi, t)), e.url = (t) => e.check(/* @__PURE__ */ Jn(Si, t)), e.jwt = (t) => e.check(/* @__PURE__ */ rr(Ui, t)), e.emoji = (t) => e.check(/* @__PURE__ */ Vn($i, t)), e.guid = (t) => e.check(/* @__PURE__ */ De(Me, t)), e.uuid = (t) => e.check(/* @__PURE__ */ Un(re, t)), e.uuidv4 = (t) => e.check(/* @__PURE__ */ Fn(re, t)), e.uuidv6 = (t) => e.check(/* @__PURE__ */ Ln(re, t)), e.uuidv7 = (t) => e.check(/* @__PURE__ */ Mn(re, t)), e.nanoid = (t) => e.check(/* @__PURE__ */ Bn(Zi, t)), e.guid = (t) => e.check(/* @__PURE__ */ De(Me, t)), e.cuid = (t) => e.check(/* @__PURE__ */ Wn(Pi, t)), e.cuid2 = (t) => e.check(/* @__PURE__ */ Gn(Ti, t)), e.ulid = (t) => e.check(/* @__PURE__ */ qn(Oi, t)), e.base64 = (t) => e.check(/* @__PURE__ */ or(ji, t)), e.base64url = (t) => e.check(/* @__PURE__ */ tr(Ri, t)), e.xid = (t) => e.check(/* @__PURE__ */ Kn(Ii, t)), e.ksuid = (t) => e.check(/* @__PURE__ */ Xn(Ni, t)), e.ipv4 = (t) => e.check(/* @__PURE__ */ Yn(Ei, t)), e.ipv6 = (t) => e.check(/* @__PURE__ */ Hn(Ci, t)), e.cidrv4 = (t) => e.check(/* @__PURE__ */ Qn(Ai, t)), e.cidrv6 = (t) => e.check(/* @__PURE__ */ er(xi, t)), e.e164 = (t) => e.check(/* @__PURE__ */ nr(Di, t)), e.datetime = (t) => e.check(ti(t)), e.date = (t) => e.check(ri(t)), e.time = (t) => e.check(ai(t)), e.duration = (t) => e.check(ci(t));
});
function c(e) {
  return /* @__PURE__ */ Rn(ki, e);
}
const I = /* @__PURE__ */ d("ZodStringFormat", (e, o) => {
  T.init(e, o), zo.init(e, o);
}), wi = /* @__PURE__ */ d("ZodEmail", (e, o) => {
  Lt.init(e, o), I.init(e, o);
}), Me = /* @__PURE__ */ d("ZodGUID", (e, o) => {
  Ut.init(e, o), I.init(e, o);
}), re = /* @__PURE__ */ d("ZodUUID", (e, o) => {
  Ft.init(e, o), I.init(e, o);
}), Si = /* @__PURE__ */ d("ZodURL", (e, o) => {
  Mt.init(e, o), I.init(e, o);
}), $i = /* @__PURE__ */ d("ZodEmoji", (e, o) => {
  Jt.init(e, o), I.init(e, o);
}), Zi = /* @__PURE__ */ d("ZodNanoID", (e, o) => {
  Vt.init(e, o), I.init(e, o);
}), Pi = /* @__PURE__ */ d("ZodCUID", (e, o) => {
  Bt.init(e, o), I.init(e, o);
}), Ti = /* @__PURE__ */ d("ZodCUID2", (e, o) => {
  Wt.init(e, o), I.init(e, o);
}), Oi = /* @__PURE__ */ d("ZodULID", (e, o) => {
  Gt.init(e, o), I.init(e, o);
}), Ii = /* @__PURE__ */ d("ZodXID", (e, o) => {
  qt.init(e, o), I.init(e, o);
}), Ni = /* @__PURE__ */ d("ZodKSUID", (e, o) => {
  Kt.init(e, o), I.init(e, o);
}), Ei = /* @__PURE__ */ d("ZodIPv4", (e, o) => {
  en.init(e, o), I.init(e, o);
}), Ci = /* @__PURE__ */ d("ZodIPv6", (e, o) => {
  on.init(e, o), I.init(e, o);
}), Ai = /* @__PURE__ */ d("ZodCIDRv4", (e, o) => {
  tn.init(e, o), I.init(e, o);
}), xi = /* @__PURE__ */ d("ZodCIDRv6", (e, o) => {
  nn.init(e, o), I.init(e, o);
}), ji = /* @__PURE__ */ d("ZodBase64", (e, o) => {
  rn.init(e, o), I.init(e, o);
}), Ri = /* @__PURE__ */ d("ZodBase64URL", (e, o) => {
  sn.init(e, o), I.init(e, o);
}), Di = /* @__PURE__ */ d("ZodE164", (e, o) => {
  cn.init(e, o), I.init(e, o);
}), Ui = /* @__PURE__ */ d("ZodJWT", (e, o) => {
  un.init(e, o), I.init(e, o);
}), ko = /* @__PURE__ */ d("ZodNumber", (e, o) => {
  lo.init(e, o), P.init(e, o), e._zod.processJSONSchema = (n, r, a) => Cr(e, n, r), e.gt = (n, r) => e.check(/* @__PURE__ */ Fe(n, r)), e.gte = (n, r) => e.check(/* @__PURE__ */ me(n, r)), e.min = (n, r) => e.check(/* @__PURE__ */ me(n, r)), e.lt = (n, r) => e.check(/* @__PURE__ */ Ue(n, r)), e.lte = (n, r) => e.check(/* @__PURE__ */ he(n, r)), e.max = (n, r) => e.check(/* @__PURE__ */ he(n, r)), e.int = (n) => e.check(Je(n)), e.safe = (n) => e.check(Je(n)), e.positive = (n) => e.check(/* @__PURE__ */ Fe(0, n)), e.nonnegative = (n) => e.check(/* @__PURE__ */ me(0, n)), e.negative = (n) => e.check(/* @__PURE__ */ Ue(0, n)), e.nonpositive = (n) => e.check(/* @__PURE__ */ he(0, n)), e.multipleOf = (n, r) => e.check(/* @__PURE__ */ Le(n, r)), e.step = (n, r) => e.check(/* @__PURE__ */ Le(n, r)), e.finite = () => e;
  const t = e._zod.bag;
  e.minValue = Math.max(t.minimum ?? Number.NEGATIVE_INFINITY, t.exclusiveMinimum ?? Number.NEGATIVE_INFINITY) ?? null, e.maxValue = Math.min(t.maximum ?? Number.POSITIVE_INFINITY, t.exclusiveMaximum ?? Number.POSITIVE_INFINITY) ?? null, e.isInt = (t.format ?? "").includes("int") || Number.isSafeInteger(t.multipleOf ?? 0.5), e.isFinite = !0, e.format = t.format ?? null;
});
function g(e) {
  return /* @__PURE__ */ lr(ko, e);
}
const Fi = /* @__PURE__ */ d("ZodNumberFormat", (e, o) => {
  pn.init(e, o), ko.init(e, o);
});
function Je(e) {
  return /* @__PURE__ */ ur(Fi, e);
}
const Li = /* @__PURE__ */ d("ZodBoolean", (e, o) => {
  dn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Ar(e, t, n);
});
function f(e) {
  return /* @__PURE__ */ pr(Li, e);
}
const Mi = /* @__PURE__ */ d("ZodNull", (e, o) => {
  fn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => xr(e, t, n);
});
function b(e) {
  return /* @__PURE__ */ dr(Mi, e);
}
const Ji = /* @__PURE__ */ d("ZodUnknown", (e, o) => {
  hn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Rr();
});
function s() {
  return /* @__PURE__ */ fr(Ji);
}
const Vi = /* @__PURE__ */ d("ZodNever", (e, o) => {
  mn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => jr(e, t, n);
});
function Bi(e) {
  return /* @__PURE__ */ hr(Vi, e);
}
const Wi = /* @__PURE__ */ d("ZodArray", (e, o) => {
  gn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Lr(e, t, n, r), e.element = o.element, e.min = (t, n) => e.check(/* @__PURE__ */ ce(t, n)), e.nonempty = (t) => e.check(/* @__PURE__ */ ce(1, t)), e.max = (t, n) => e.check(/* @__PURE__ */ ho(t, n)), e.length = (t, n) => e.check(/* @__PURE__ */ mo(t, n)), e.unwrap = () => e.element;
});
function U(e, o) {
  return /* @__PURE__ */ Zr(Wi, e, o);
}
const Gi = /* @__PURE__ */ d("ZodObject", (e, o) => {
  vn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Mr(e, t, n, r), $(e, "shape", () => o.shape), e.keyof = () => y(Object.keys(e._zod.def.shape)), e.catchall = (t) => e.clone({ ...e._zod.def, catchall: t }), e.passthrough = () => e.clone({ ...e._zod.def, catchall: s() }), e.loose = () => e.clone({ ...e._zod.def, catchall: s() }), e.strict = () => e.clone({ ...e._zod.def, catchall: Bi() }), e.strip = () => e.clone({ ...e._zod.def, catchall: void 0 }), e.extend = (t) => jo(e, t), e.safeExtend = (t) => Ro(e, t), e.merge = (t) => Do(e, t), e.pick = (t) => Ao(e, t), e.omit = (t) => xo(e, t), e.partial = (...t) => Uo(wo, e, t[0]), e.required = (...t) => Fo(So, e, t[0]);
});
function z(e, o) {
  const t = {
    type: "object",
    shape: e ?? {},
    ...k(o)
  };
  return new Gi(t);
}
const qi = /* @__PURE__ */ d("ZodUnion", (e, o) => {
  bn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Jr(e, t, n, r), e.options = o.options;
});
function v(e, o) {
  return new qi({
    type: "union",
    options: e,
    ...k(o)
  });
}
const Ki = /* @__PURE__ */ d("ZodIntersection", (e, o) => {
  yn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Vr(e, t, n, r);
});
function Xi(e, o) {
  return new Ki({
    type: "intersection",
    left: e,
    right: o
  });
}
const Yi = /* @__PURE__ */ d("ZodTuple", (e, o) => {
  zn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Br(e, t, n, r), e.rest = (t) => e.clone({
    ...e._zod.def,
    rest: t
  });
});
function W(e, o, t) {
  const n = o instanceof Z, r = n ? t : o, a = n ? o : null;
  return new Yi({
    type: "tuple",
    items: e,
    rest: a,
    ...k(r)
  });
}
const Hi = /* @__PURE__ */ d("ZodRecord", (e, o) => {
  kn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Wr(e, t, n, r), e.keyType = o.keyType, e.valueType = o.valueType;
});
function H(e, o, t) {
  return new Hi({
    type: "record",
    keyType: e,
    valueType: o,
    ...k(t)
  });
}
const ve = /* @__PURE__ */ d("ZodEnum", (e, o) => {
  wn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (n, r, a) => Dr(e, n, r), e.enum = o.entries, e.options = Object.values(o.entries);
  const t = new Set(Object.keys(o.entries));
  e.extract = (n, r) => {
    const a = {};
    for (const i of n)
      if (t.has(i))
        a[i] = o.entries[i];
      else
        throw new Error(`Key ${i} not found in enum`);
    return new ve({
      ...o,
      checks: [],
      ...k(r),
      entries: a
    });
  }, e.exclude = (n, r) => {
    const a = { ...o.entries };
    for (const i of n)
      if (t.has(i))
        delete a[i];
      else
        throw new Error(`Key ${i} not found in enum`);
    return new ve({
      ...o,
      checks: [],
      ...k(r),
      entries: a
    });
  };
});
function y(e, o) {
  const t = Array.isArray(e) ? Object.fromEntries(e.map((n) => [n, n])) : e;
  return new ve({
    type: "enum",
    entries: t,
    ...k(o)
  });
}
const Qi = /* @__PURE__ */ d("ZodTransform", (e, o) => {
  Sn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Fr(e, t), e._zod.parse = (t, n) => {
    if (n.direction === "backward")
      throw new Ge(e.constructor.name);
    t.addIssue = (a) => {
      if (typeof a == "string")
        t.issues.push(Q(a, t.value, o));
      else {
        const i = a;
        i.fatal && (i.continue = !1), i.code ?? (i.code = "custom"), i.input ?? (i.input = t.value), i.inst ?? (i.inst = e), t.issues.push(Q(i));
      }
    };
    const r = o.transform(t.value, t);
    return r instanceof Promise ? r.then((a) => (t.value = a, t)) : (t.value = r, t);
  };
});
function ea(e) {
  return new Qi({
    type: "transform",
    transform: e
  });
}
const wo = /* @__PURE__ */ d("ZodOptional", (e, o) => {
  fo.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => bo(e, t, n, r), e.unwrap = () => e._zod.def.innerType;
});
function Ve(e) {
  return new wo({
    type: "optional",
    innerType: e
  });
}
const oa = /* @__PURE__ */ d("ZodExactOptional", (e, o) => {
  $n.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => bo(e, t, n, r), e.unwrap = () => e._zod.def.innerType;
});
function ta(e) {
  return new oa({
    type: "optional",
    innerType: e
  });
}
const na = /* @__PURE__ */ d("ZodNullable", (e, o) => {
  Zn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Gr(e, t, n, r), e.unwrap = () => e._zod.def.innerType;
});
function Be(e) {
  return new na({
    type: "nullable",
    innerType: e
  });
}
const ra = /* @__PURE__ */ d("ZodDefault", (e, o) => {
  Pn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Kr(e, t, n, r), e.unwrap = () => e._zod.def.innerType, e.removeDefault = e.unwrap;
});
function ia(e, o) {
  return new ra({
    type: "default",
    innerType: e,
    get defaultValue() {
      return typeof o == "function" ? o() : Ye(o);
    }
  });
}
const aa = /* @__PURE__ */ d("ZodPrefault", (e, o) => {
  Tn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Xr(e, t, n, r), e.unwrap = () => e._zod.def.innerType;
});
function sa(e, o) {
  return new aa({
    type: "prefault",
    innerType: e,
    get defaultValue() {
      return typeof o == "function" ? o() : Ye(o);
    }
  });
}
const So = /* @__PURE__ */ d("ZodNonOptional", (e, o) => {
  On.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => qr(e, t, n, r), e.unwrap = () => e._zod.def.innerType;
});
function ca(e, o) {
  return new So({
    type: "nonoptional",
    innerType: e,
    ...k(o)
  });
}
const la = /* @__PURE__ */ d("ZodCatch", (e, o) => {
  In.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Yr(e, t, n, r), e.unwrap = () => e._zod.def.innerType, e.removeCatch = e.unwrap;
});
function ua(e, o) {
  return new la({
    type: "catch",
    innerType: e,
    catchValue: typeof o == "function" ? o : () => o
  });
}
const pa = /* @__PURE__ */ d("ZodPipe", (e, o) => {
  Nn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Hr(e, t, n, r), e.in = o.in, e.out = o.out;
});
function We(e, o) {
  return new pa({
    type: "pipe",
    in: e,
    out: o
    // ...util.normalizeParams(params),
  });
}
const da = /* @__PURE__ */ d("ZodReadonly", (e, o) => {
  En.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Qr(e, t, n, r), e.unwrap = () => e._zod.def.innerType;
});
function fa(e) {
  return new da({
    type: "readonly",
    innerType: e
  });
}
const ha = /* @__PURE__ */ d("ZodLazy", (e, o) => {
  Cn.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => ei(e, t, n, r), e.unwrap = () => e._zod.def.getter();
});
function ma(e) {
  return new ha({
    type: "lazy",
    getter: e
  });
}
const ga = /* @__PURE__ */ d("ZodCustom", (e, o) => {
  An.init(e, o), P.init(e, o), e._zod.processJSONSchema = (t, n, r) => Ur(e, t);
});
function _a(e, o = {}) {
  return /* @__PURE__ */ Pr(ga, e, o);
}
function va(e) {
  return /* @__PURE__ */ Tr(e);
}
const m = v([
  c(),
  g(),
  f(),
  b(),
  z({ path: c() })
]), ie = ma(
  () => v([
    z({ and: U(ie) }),
    z({ or: U(ie) }),
    z({ not: ie }),
    z({ path: c() }),
    z({ eq: W([m, m]) }),
    z({ neq: W([m, m]) }),
    z({ gt: W([m, m]) }),
    z({ gte: W([m, m]) }),
    z({ lt: W([m, m]) }),
    z({ lte: W([m, m]) })
  ])
), ba = v([
  f(),
  z({ path: c() }),
  z({ auth: y(["signedIn", "signedOut"]) }),
  ie
]), ya = z({
  title: c(),
  message: c(),
  variant: y(["default", "danger"]).optional(),
  confirmLabel: c().optional(),
  cancelLabel: c().optional()
}), za = z({
  name: c(),
  params: H(c(), m).optional(),
  confirm: ya.optional(),
  onSuccess: z({ set: H(c(), m) }).optional(),
  onError: z({ set: H(c(), m) }).optional()
}), ka = z({
  items: U(s()),
  // Array of items to display in the dropdown
  value: U(s()).optional(),
  // The controlled input value
  open: f().optional(),
  // Whether the popup is open (controlled)
  children: v([c(), g(), f(), b(), m]).optional(),
  // Autocomplete content (input group, popup content)
  className: c().optional(),
  // Additional CSS classes
  label: v([c(), g(), f(), b(), m]).optional(),
  // Label content (enables Field wrapper)
  required: f().optional(),
  // Whether the field is required
  labelTooltip: v([c(), g(), f(), b(), m]).optional(),
  // Tooltip content to display next to the label
  description: v([c(), g(), f(), b(), m]).optional(),
  // Helper text displayed below the field
  error: s().optional()
  // Error message or validation error object
}), wa = z({
  variant: y(["primary", "secondary", "error", "warning", "success", "destructive", "info", "beta", "outline", "red", "green", "neutral", "orange", "purple", "teal", "teal-subtle", "blue"]).optional(),
  // Color variant of the badge. Recommended semantic variants: - `"primary"` — Primary badge - `"secondary"` — Secondary badge - `"error"` — Error badge - `"warning"` — Warning badge - `"success"` — Success badge - `"info"` — Info badge Additional token variants: - `"red"`, `"orange"`, `"green"`, `"teal"`, `"blue"`, `"purple"`, `"neutral"` - `"teal-subtle"`, `"neutral-subtle"` - `"inverted"` - `"outline"` — Bordered badge with transparent background - `"beta"` — Dashed-border badge for beta/experimental features
  className: c().optional(),
  // Additional CSS classes merged via `cn()`.
  children: v([c(), g(), f(), b(), m]).optional()
  // Content rendered inside the badge.
}), Sa = z({
  icon: v([c(), g(), f(), b(), m]).optional(),
  // Icon element rendered before the banner content (e.g. from `@phosphor-icons/react`).
  title: c().optional(),
  // Primary heading text for the banner. Use for i18n string injection.
  description: v([c(), g(), f(), b(), m]).optional(),
  // Secondary description text displayed below the title. Use for i18n string injection.
  action: v([c(), g(), f(), b(), m]).optional(),
  // Action slot rendered at the trailing end of the banner (e.g. a CTA button or link). Only used in structured mode (with `title` or `description`).
  text: c().optional(),
  children: v([c(), g(), f(), b(), m]).optional(),
  variant: y(["default", "alert", "error"]).optional(),
  // Visual style of the banner. - `"default"` — Informational blue banner for general messages - `"alert"` — Warning yellow banner for cautionary messages - `"error"` — Error red banner for critical issues
  className: c().optional()
  // Additional CSS classes merged via `cn()`.
}), $a = z({
  size: y(["sm", "base"]).optional(),
  // Size of the breadcrumbs. - `"sm"` — Compact breadcrumbs for dense UIs - `"base"` — Default breadcrumbs size
  children: v([c(), g(), f(), b(), m]).optional(),
  className: c().optional()
  // Additional CSS classes merged via `cn()`.
}), Za = z({
  shape: y(["base", "square", "circle"]).optional(),
  size: y(["xs", "sm", "base", "lg"]).optional(),
  variant: y(["primary", "secondary", "ghost", "destructive", "secondary-destructive", "outline"]).optional(),
  children: v([c(), g(), f(), b(), m]).optional(),
  className: c().optional(),
  icon: v([c(), g(), f(), b(), m]).optional(),
  // Icon from `@phosphor-icons/react` or a React element. Rendered before children.
  loading: f().optional(),
  // Shows a loading spinner and disables interaction.
  title: c().optional(),
  id: c().optional(),
  lang: c().optional(),
  disabled: f().optional(),
  name: c().optional(),
  type: y(["submit", "reset", "button"]).optional(),
  value: s().optional()
}), Pa = z({
  variant: y(["default", "error"]).optional(),
  // Visual variant: "default" or "error" for validation failures (visual only, no error text)
  label: v([c(), g(), f(), b(), m]).optional(),
  // Label content for the checkbox (enables built-in Field wrapper) - can be a string or any React node
  labelTooltip: v([c(), g(), f(), b(), m]).optional(),
  // Tooltip content to display next to the label via an info icon
  controlFirst: f().optional(),
  // When true (default), checkbox appears before label. When false, label appears before checkbox.
  checked: f().optional(),
  // Whether the checkbox is checked (controlled)
  indeterminate: f().optional(),
  // Whether the checkbox is in indeterminate state
  disabled: f().optional(),
  // Whether the checkbox is disabled
  name: c().optional(),
  // Name for form submission
  required: f().optional(),
  // Whether the field is required
  className: c().optional()
  // Additional class name
}), Ta = z({
  size: y(["sm", "base", "lg"]).optional(),
  // Size of the clipboard text field. - `"sm"` — Small clipboard text for compact UIs - `"base"` — Default clipboard text size - `"lg"` — Large clipboard text for prominent display
  text: c(),
  // The text to display and copy to clipboard.
  textToCopy: c().optional(),
  // If provided, this text will be copied to clipboard instead of the `text` prop.
  className: c().optional(),
  // Additional CSS classes merged via `cn()`.
  tooltip: s().optional(),
  // Tooltip config. Shows tooltip on hover, anchored toast on click.
  labels: s().optional()
  // Accessible labels for i18n.
}), Oa = z({
  children: v([c(), g(), f(), b(), m]).optional(),
  className: c().optional(),
  height: s().optional(),
  id: c().optional(),
  lang: c().optional(),
  media: c().optional(),
  method: c().optional(),
  name: c().optional(),
  target: c().optional(),
  type: c().optional(),
  width: s().optional(),
  accentHeight: s().optional(),
  accumulate: y(["none", "sum"]).optional(),
  additive: y(["replace", "sum"]).optional(),
  alignmentBaseline: y(["auto", "baseline", "before-edge", "text-before-edge", "middle", "central", "after-edge", "text-after-edge", "ideographic", "alphabetic", "hanging", "mathematical", "inherit"]).optional(),
  allowReorder: y(["no", "yes"]).optional(),
  alphabetic: s().optional(),
  amplitude: s().optional(),
  arabicForm: y(["initial", "medial", "terminal", "isolated"]).optional(),
  ascent: s().optional(),
  attributeName: c().optional(),
  attributeType: c().optional(),
  autoReverse: s().optional(),
  azimuth: s().optional(),
  baseFrequency: s().optional(),
  baselineShift: s().optional(),
  baseProfile: s().optional(),
  bbox: s().optional(),
  begin: s().optional(),
  bias: s().optional(),
  by: s().optional(),
  calcMode: s().optional(),
  capHeight: s().optional(),
  clip: s().optional(),
  clipPath: c().optional(),
  clipPathUnits: s().optional(),
  clipRule: s().optional(),
  colorInterpolation: s().optional(),
  colorInterpolationFilters: y(["auto", "sRGB", "linearRGB", "inherit"]).optional(),
  colorProfile: s().optional(),
  colorRendering: s().optional(),
  contentScriptType: s().optional(),
  contentStyleType: s().optional(),
  cursor: s().optional(),
  cx: s().optional(),
  cy: s().optional(),
  d: c().optional(),
  decelerate: s().optional(),
  descent: s().optional(),
  diffuseConstant: s().optional(),
  direction: s().optional(),
  display: s().optional(),
  divisor: s().optional(),
  dominantBaseline: y(["auto", "use-script", "no-change", "reset-size", "ideographic", "alphabetic", "hanging", "mathematical", "central", "middle", "text-after-edge", "text-before-edge", "inherit"]).optional(),
  dur: s().optional(),
  dx: s().optional(),
  dy: s().optional(),
  edgeMode: s().optional(),
  elevation: s().optional(),
  enableBackground: s().optional(),
  end: s().optional(),
  exponent: s().optional(),
  externalResourcesRequired: s().optional(),
  fill: c().optional(),
  fillOpacity: s().optional(),
  fillRule: y(["nonzero", "evenodd", "inherit"]).optional(),
  filter: c().optional(),
  filterRes: s().optional(),
  filterUnits: s().optional(),
  floodColor: s().optional(),
  floodOpacity: s().optional(),
  focusable: s().optional(),
  fontFamily: c().optional(),
  fontSize: s().optional(),
  fontSizeAdjust: s().optional(),
  fontStretch: s().optional(),
  fontStyle: s().optional(),
  fontVariant: s().optional(),
  fontWeight: s().optional(),
  format: s().optional(),
  fr: s().optional(),
  from: s().optional(),
  fx: s().optional(),
  fy: s().optional(),
  g1: s().optional(),
  g2: s().optional(),
  glyphName: s().optional(),
  glyphOrientationHorizontal: s().optional(),
  glyphOrientationVertical: s().optional(),
  glyphRef: s().optional(),
  gradientTransform: c().optional(),
  gradientUnits: c().optional(),
  hanging: s().optional(),
  horizAdvX: s().optional(),
  horizOriginX: s().optional(),
  href: c().optional(),
  ideographic: s().optional(),
  imageRendering: s().optional(),
  in2: s().optional(),
  in: c().optional(),
  intercept: s().optional(),
  k1: s().optional(),
  k2: s().optional(),
  k3: s().optional(),
  k4: s().optional(),
  k: s().optional(),
  kernelMatrix: s().optional(),
  kernelUnitLength: s().optional(),
  kerning: s().optional(),
  keyPoints: s().optional(),
  keySplines: s().optional(),
  keyTimes: s().optional(),
  lengthAdjust: s().optional(),
  letterSpacing: s().optional(),
  lightingColor: s().optional(),
  limitingConeAngle: s().optional(),
  local: s().optional(),
  markerEnd: c().optional(),
  markerHeight: s().optional(),
  markerMid: c().optional(),
  markerStart: c().optional(),
  markerUnits: s().optional(),
  markerWidth: s().optional(),
  mask: c().optional(),
  maskContentUnits: s().optional(),
  maskUnits: s().optional(),
  mathematical: s().optional(),
  mode: s().optional(),
  numOctaves: s().optional(),
  offset: s().optional(),
  opacity: s().optional(),
  operator: s().optional(),
  order: s().optional(),
  orient: s().optional(),
  orientation: s().optional(),
  origin: s().optional(),
  overflow: s().optional(),
  overlinePosition: s().optional(),
  overlineThickness: s().optional(),
  paintOrder: s().optional(),
  panose1: s().optional(),
  path: c().optional(),
  pathLength: s().optional(),
  patternContentUnits: c().optional(),
  patternTransform: s().optional(),
  patternUnits: c().optional(),
  pointerEvents: s().optional(),
  points: c().optional(),
  pointsAtX: s().optional(),
  pointsAtY: s().optional(),
  pointsAtZ: s().optional(),
  preserveAlpha: s().optional(),
  preserveAspectRatio: c().optional(),
  primitiveUnits: s().optional(),
  r: s().optional(),
  radius: s().optional(),
  refX: s().optional(),
  refY: s().optional(),
  renderingIntent: s().optional(),
  repeatCount: s().optional(),
  repeatDur: s().optional(),
  requiredExtensions: s().optional(),
  requiredFeatures: s().optional(),
  restart: s().optional(),
  result: c().optional(),
  rotate: s().optional(),
  rx: s().optional(),
  ry: s().optional(),
  scale: s().optional(),
  seed: s().optional(),
  shapeRendering: s().optional(),
  slope: s().optional(),
  spacing: s().optional(),
  specularConstant: s().optional(),
  specularExponent: s().optional(),
  speed: s().optional(),
  spreadMethod: c().optional(),
  startOffset: s().optional(),
  stdDeviation: s().optional(),
  stemh: s().optional(),
  stemv: s().optional(),
  stitchTiles: s().optional(),
  stopColor: c().optional(),
  stopOpacity: s().optional(),
  strikethroughPosition: s().optional(),
  strikethroughThickness: s().optional(),
  string: s().optional(),
  stroke: c().optional(),
  strokeDasharray: s().optional(),
  strokeDashoffset: s().optional(),
  strokeLinecap: y(["butt", "round", "square", "inherit"]).optional(),
  strokeLinejoin: y(["miter", "round", "bevel", "inherit"]).optional(),
  strokeMiterlimit: s().optional(),
  strokeOpacity: s().optional(),
  strokeWidth: s().optional(),
  surfaceScale: s().optional(),
  systemLanguage: s().optional(),
  tableValues: s().optional(),
  targetX: s().optional(),
  targetY: s().optional(),
  textAnchor: y(["start", "middle", "end", "inherit"]).optional(),
  textDecoration: s().optional(),
  textLength: s().optional(),
  textRendering: s().optional(),
  to: s().optional(),
  transform: c().optional(),
  u1: s().optional(),
  u2: s().optional(),
  underlinePosition: s().optional(),
  underlineThickness: s().optional(),
  unicode: s().optional(),
  unicodeBidi: s().optional(),
  unicodeRange: s().optional(),
  unitsPerEm: s().optional(),
  vAlphabetic: s().optional(),
  values: c().optional(),
  vectorEffect: s().optional(),
  version: c().optional(),
  vertAdvY: s().optional(),
  vertOriginX: s().optional(),
  vertOriginY: s().optional(),
  vHanging: s().optional(),
  vIdeographic: s().optional(),
  viewBox: c().optional(),
  viewTarget: s().optional(),
  visibility: s().optional(),
  vMathematical: s().optional(),
  widths: s().optional(),
  wordSpacing: s().optional(),
  writingMode: s().optional(),
  x1: s().optional(),
  x2: s().optional(),
  x: s().optional(),
  xChannelSelector: c().optional(),
  xHeight: s().optional(),
  xlinkActuate: c().optional(),
  xlinkArcrole: c().optional(),
  xlinkHref: c().optional(),
  xlinkRole: c().optional(),
  xlinkShow: c().optional(),
  xlinkTitle: c().optional(),
  xlinkType: c().optional(),
  xmlBase: c().optional(),
  xmlLang: c().optional(),
  xmlns: c().optional(),
  xmlnsXlink: c().optional(),
  xmlSpace: c().optional(),
  y1: s().optional(),
  y2: s().optional(),
  y: s().optional(),
  yChannelSelector: c().optional(),
  z: s().optional(),
  zoomAndPan: c().optional(),
  variant: y(["glyph", "full"]).optional()
  // Logo variant - `glyph`: Cloud icon only - `full`: Cloud icon with "CLOUDFLARE" wordmark below
}), Ia = z({
  lang: y(["ts", "tsx", "jsonc", "bash", "css"]).optional(),
  // Language hint for the code content. - `"ts"` — TypeScript code - `"tsx"` — TypeScript JSX code - `"jsonc"` — JSON with comments - `"bash"` — Shell/Bash commands - `"css"` — CSS styles
  code: c(),
  // The code string to display.
  values: s().optional(),
  // Template values for `{{key}}` interpolation. Values with `highlight: true` are visually emphasized.
  className: c().optional()
  // Additional CSS classes merged via `cn()`.
}), Na = z({
  className: c().optional(),
  // Additional CSS classes
  onOpenChange: s().optional()
  // Callback when collapsed state changes
}), Ea = z({
  size: y(["xs", "sm", "base", "lg"]).optional(),
  // Size of the combobox trigger. Matches Input component sizes. - `"xs"` — Extra small for compact UIs (h-5 / 20px) - `"sm"` — Small for secondary fields (h-6.5 / 26px) - `"base"` — Default size (h-9 / 36px) - `"lg"` — Large for prominent fields (h-10 / 40px)
  inputSide: y(["right", "top"]).optional(),
  // Position of the text input relative to chips in multi-select mode. - `"right"` — Input inline to the right of chips - `"top"` — Input above chips
  items: U(s()),
  // Array of items to display in the dropdown
  value: U(s()).optional(),
  // Currently selected value(s)
  children: v([c(), g(), f(), b(), m]).optional(),
  // Combobox content (trigger, content, items)
  className: c().optional(),
  // Additional CSS classes
  label: v([c(), g(), f(), b(), m]).optional(),
  // Label content for the combobox (enables Field wrapper) - can be a string or any React node
  required: f().optional(),
  // Whether the combobox is required
  labelTooltip: v([c(), g(), f(), b(), m]).optional(),
  // Tooltip content to display next to the label via an info icon
  description: v([c(), g(), f(), b(), m]).optional(),
  // Helper text displayed below the combobox
  error: s().optional(),
  // Error message or validation error object
  onValueChange: s().optional(),
  // Callback when selection changes
  multiple: f().optional(),
  // Allow multiple selections
  isItemEqualToValue: s().optional()
  // Custom equality function for comparing items
}), Ca = z({
  open: f(),
  // Whether the dialog is open
  children: v([c(), g(), f(), b(), m]).optional(),
  // Child content - typically one or more Panel components
  container: s().optional()
  // Container element for the portal. Use this to render the command palette inside a Shadow DOM or custom container. Overrides `KumoPortalProvider` context.
}), Aa = z({
  className: c().optional(),
  // Additional CSS classes
  children: v([c(), g(), f(), b(), m]).optional()
  // Child elements
}), xa = z({
  size: y(["sm", "base", "lg"]).optional(),
  // Calendar size. - `"sm"` — Compact calendar for tight spaces - `"base"` — Default calendar size - `"lg"` — Large calendar for prominent date selection
  variant: y(["default", "subtle"]).optional(),
  // Visual variant. - `"default"` — Standard appearance with overlay background - `"subtle"` — Minimal background
  timezone: c().optional(),
  // Display timezone string shown in the footer.
  className: c().optional(),
  // Additional CSS classes merged via `cn()`.
  onStartDateChange: s().optional(),
  // Callback when start date changes
  onEndDateChange: s().optional()
  // Callback when end date changes
}), ja = z({
  className: c().optional(),
  // Additional CSS classes merged via `cn()`.
  children: v([c(), g(), f(), b(), m]).optional(),
  // Dialog content (typically Title, Description, Close, and action buttons).
  container: s().optional(),
  // Container element for the portal. Use this to render the dialog inside a Shadow DOM or custom container. Overrides `KumoPortalProvider` context.
  size: y(["base", "sm", "lg", "xl"]).optional()
  // Dialog width. - `"sm"` — Small (min 288px) for simple confirmations - `"base"` — Default (min 384px) - `"lg"` — Large (min 512px) for complex content - `"xl"` — Extra large (min 768px) for detailed views
}), Ra = z({
  variant: y(["default", "danger"]).optional()
  // Visual style of the dropdown item. - `"default"` — Standard item appearance - `"danger"` — Destructive action with red text
}), Da = z({
  size: y(["sm", "base", "lg"]).optional(),
  // Size of the empty state container. - `"sm"` — Compact empty state for smaller containers - `"base"` — Default empty state size - `"lg"` — Large empty state for prominent placement
  icon: v([c(), g(), f(), b(), m]).optional(),
  // Decorative icon displayed above the title (e.g. from `@phosphor-icons/react`).
  title: c(),
  // Primary heading text for the empty state.
  description: c().optional(),
  // Secondary description text displayed below the title.
  commandLine: c().optional(),
  // Shell command displayed in a copyable code block.
  contents: v([c(), g(), f(), b(), m]).optional(),
  // Additional content (buttons, links) rendered below the description.
  className: c().optional()
  // Additional CSS classes merged via `cn()`.
}), Ua = z({
  controlFirst: f().optional(),
  // When `true`, places the control before the label (for checkbox/switch layouts).
  children: v([c(), g(), f(), b(), m]).optional(),
  // The form control element(s) to wrap (Input, Select, Checkbox, etc.).
  label: v([c(), g(), f(), b(), m]).optional(),
  // The label content — can be a string or any React node.
  required: f().optional(),
  // When explicitly `false`, shows gray "(optional)" text after the label. When `true` or `undefined`, no indicator is shown.
  labelTooltip: v([c(), g(), f(), b(), m]).optional(),
  // Tooltip content displayed next to the label via an info icon.
  error: s().optional(),
  // Validation error with a message and a browser `ValidityState` match key.
  description: v([c(), g(), f(), b(), m]).optional()
  // Helper text displayed below the control (hidden when `error` is present).
}), Fa = z({
  children: v([c(), g(), f(), b(), m]).optional(),
  // Grid items to render.
  className: c().optional(),
  // Additional CSS classes merged via `cn()`.
  id: c().optional(),
  lang: c().optional(),
  title: c().optional(),
  mobileDivider: f().optional(),
  // Show dividers between grid items on mobile (only works with `"4up"` variant).
  gap: y(["none", "sm", "base", "lg"]).optional(),
  // Gap size between grid items. - `"none"` — No gap - `"sm"` — 12px gap - `"base"` — Responsive gap (8px → 24px → 32px) - `"lg"` — 32px gap
  variant: y(["2up", "side-by-side", "2-1", "1-2", "1-3up", "3up", "4up", "6up", "1-2-4up"]).optional()
  // Responsive column layout variant. - `"2up"` — 1 col → 2 cols at md - `"side-by-side"` — Always 2 cols - `"2-1"` — 66%/33% split at md - `"1-2"` — 33%/66% split at md - `"3up"` — 1 → 2 → 3 cols - `"4up"` — 1 → 2 → 3 → 4 cols - `"6up"` — 2 → 3 → 4 → 6 cols - `"1-2-4up"` — 1 → 2 → 4 cols
}), La = z({
  label: v([c(), g(), f(), b(), m]).optional(),
  // Label content for the input (enables Field wrapper) - can be a string or any React node
  labelTooltip: v([c(), g(), f(), b(), m]).optional(),
  // Tooltip content to display next to the label via an info icon
  description: v([c(), g(), f(), b(), m]).optional(),
  // Helper text displayed below the input
  error: s().optional(),
  // Error message or validation error object
  size: y(["xs", "sm", "base", "lg"]).optional(),
  // Input size. - `"xs"` — Extra small for compact UIs - `"sm"` — Small for secondary fields - `"base"` — Default size - `"lg"` — Large for prominent fields
  variant: y(["default", "error"]).optional()
  // Visual variant. - `"default"` — Standard input - `"error"` — Error state for validation failures
}), Ma = z({}), Ja = z({
  label: v([c(), g(), f(), b(), m]).optional(),
  // The label content — can be a string or any React node.
  description: v([c(), g(), f(), b(), m]).optional(),
  // Helper text displayed below the control (hidden when `error` is present).
  error: s().optional(),
  // Validation error with a message and a browser `ValidityState` match key.
  required: f().optional(),
  // When explicitly `false`, shows gray "(optional)" text after the label. When `true` or `undefined`, no indicator is shown.
  labelTooltip: v([c(), g(), f(), b(), m]).optional(),
  // Tooltip content displayed next to the label via an info icon.
  children: v([c(), g(), f(), b(), m]).optional(),
  className: c().optional(),
  id: c().optional(),
  lang: c().optional(),
  title: c().optional(),
  size: y(["xs", "sm", "base", "lg"]).optional(),
  disabled: f().optional()
}), Va = z({
  children: v([c(), g(), f(), b(), m]).optional(),
  // The label content — can be a string or any React node.
  showOptional: f().optional(),
  // When `true`, shows gray "(optional)" text after the label.
  tooltip: v([c(), g(), f(), b(), m]).optional(),
  // Tooltip content displayed next to the label via an info icon.
  className: c().optional(),
  // Additional CSS classes merged via `cn()`.
  htmlFor: c().optional(),
  // The id of the form element this label is associated with
  asContent: f().optional()
  // When true, only renders the inline content (indicators, tooltip) without the outer label element with font styling. Useful when composed inside another label element that already provides the text styling.
}), Ba = z({
  render: v([c(), g(), f(), b(), m]).optional(),
  // Allows you to replace the component's HTML element with a different tag, or compose it with another component. Accepts a `ReactElement` or a function that returns the element to render.
  children: v([c(), g(), f(), b(), m]).optional(),
  className: c().optional(),
  id: c().optional(),
  lang: c().optional(),
  title: c().optional()
}), Wa = z({
  variant: y(["inline", "current", "plain"]).optional(),
  // Visual style of the link. - `"inline"` — Inline text link that flows with content - `"current"` — Link that inherits color from parent text - `"plain"` — Link without underline decoration
  to: c().optional(),
  children: v([c(), g(), f(), b(), m]).optional(),
  className: c().optional(),
  id: c().optional(),
  lang: c().optional(),
  title: c().optional(),
  download: s().optional(),
  href: c().optional(),
  hrefLang: c().optional(),
  media: c().optional(),
  ping: c().optional(),
  target: s().optional(),
  type: c().optional(),
  referrerPolicy: y(["", "no-referrer", "no-referrer-when-downgrade", "origin", "origin-when-cross-origin", "same-origin", "strict-origin", "strict-origin-when-cross-origin", "unsafe-url"]).optional(),
  render: v([c(), g(), f(), b(), m]).optional()
  // Allows you to replace the component's HTML element with a different tag, or compose it with another component. Accepts a `ReactElement` or a function that returns the element to render.
}), Ga = z({
  className: c().optional(),
  // Additional CSS classes merged via `cn()`.
  size: y(["sm", "base", "lg"]).optional()
  // Size of the spinner. Use a preset name or a custom pixel number. - `"sm"` — 16px, for inline use - `"base"` — 24px, default size - `"lg"` — 32px, for prominent loading states
}), qa = z({
  className: c().optional(),
  // Additional CSS classes merged via `cn()`.
  isActive: s().optional(),
  // The currently active option value — matched against option index or `id`.
  options: U(s()),
  // Array of menu option configurations.
  optionIds: f().optional()
  // When true, each option's `id` field is used for matching instead of its array index.
}), Ka = z({
  customValue: c().optional(),
  // Custom formatted value text (e.g. "750 / 1,000") displayed instead of percentage.
  label: c(),
  // Label text displayed above the meter track.
  showValue: f().optional(),
  // Whether to display the percentage value next to the label.
  trackClassName: c().optional(),
  // Additional CSS classes for the track (background bar).
  indicatorClassName: c().optional(),
  // Additional CSS classes for the indicator (filled bar).
  value: g().optional(),
  // Current value of the meter
  max: g().optional(),
  // Maximum value of the meter (default: 100)
  min: g().optional()
  // Minimum value of the meter (default: 0)
}), Xa = z({
  setPage: s(),
  // Callback when page changes
  page: g().optional(),
  // Current page number (1-indexed).
  perPage: g().optional(),
  // Number of items displayed per page.
  totalCount: g().optional(),
  // Total number of items across all pages.
  className: c().optional(),
  // Additional CSS classes for the container
  labels: s().optional(),
  // Labels for internationalization of aria-labels. All labels have English defaults. For visible text like "Showing X of Y", use render props on sub-components: - `Pagination.Info` children for the info text - `Pagination.PageSize` label prop for the "Per page:" text
  children: v([c(), g(), f(), b(), m]).optional(),
  // Compound component children for custom layouts. Use Pagination.Info, Pagination.PageSize, Pagination.Controls, and Pagination.Separator.
  controls: y(["full", "simple"]).optional(),
  text: s().optional()
}), Ya = z({
  side: y(["top", "bottom", "left", "right"]).optional()
  // Which side of the trigger the popover appears on. - `"top"` — Above the trigger - `"bottom"` — Below the trigger - `"left"` — Left of the trigger - `"right"` — Right of the trigger
}), Ha = z({
  legend: c().optional(),
  // Legend text for the group (required for accessibility). For more control over legend styling, omit this prop and use `<Radio.Legend>` as a child instead.
  children: v([c(), g(), f(), b(), m]).optional(),
  // Child Radio.Item components (and optionally a Radio.Legend)
  orientation: y(["vertical", "horizontal"]).optional(),
  // Layout direction of the radio items
  appearance: y(["default", "card"]).optional(),
  // Visual appearance applied to all Radio.Item children. - `"default"` — Standard inline radio items - `"card"` — Choice card with border, padding, and highlighted selection state Individual items can override this with their own `appearance` prop.
  error: c().optional(),
  // Error message for the group
  description: v([c(), g(), f(), b(), m]).optional(),
  // Helper text for the group
  value: c().optional(),
  // Value of the radio that should be selected (controlled)
  disabled: f().optional(),
  // Whether all radios in the group are disabled
  controlPosition: y(["start", "end"]).optional(),
  // Position of radio control relative to label: "start" puts radio before label, "end" puts label before radio. Defaults to "start" for default appearance and "end" for card appearance.
  name: c().optional(),
  // Form submission name for the radio group
  className: c().optional()
  // Additional CSS classes
}), Qa = z({
  className: c().optional(),
  // Additional CSS classes merged via `cn()`.
  size: y(["xs", "sm", "base", "lg"]).optional(),
  // Size of the select trigger. Matches Input component sizes.
  label: v([c(), g(), f(), b(), m]).optional(),
  // Label content for the select. When provided, enables the Field wrapper with a visible label above the select. For accessibility without a visible label, use `aria-label` instead.
  hideLabel: f().optional(),
  placeholder: c().optional(),
  // Placeholder text shown when no value is selected.
  loading: f().optional(),
  // When `true`, shows a skeleton loader in place of the selected value.
  disabled: f().optional(),
  // Whether the select is disabled.
  required: f().optional(),
  // Whether the select is required. When `false`, shows "(optional)" text.
  labelTooltip: v([c(), g(), f(), b(), m]).optional(),
  // Tooltip content displayed next to the label via an info icon.
  value: s().optional(),
  // Currently selected value (controlled mode).
  children: v([c(), g(), f(), b(), m]).optional(),
  // `Select.Option` elements to render in the dropdown.
  description: v([c(), g(), f(), b(), m]).optional(),
  // Helper text displayed below the select.
  error: s().optional(),
  // Error message string or validation error object with `match` key.
  onValueChange: s().optional(),
  // Callback when selection changes
  defaultValue: s().optional(),
  // Initial value for uncontrolled mode
  renderValue: s().optional(),
  // A function that returns a ReactNode to format the selected value in the trigger. Required when using object values. Use `placeholder` for the empty state.
  items: s().optional(),
  // Data structure of items rendered in the popup. Accepts a plain object map (`{ key: "Label" }`) or an array of `{ label, value }` for object/complex values.
  isItemEqualToValue: s().optional()
  // Custom equality function for comparing items. Required when value is an object, since object identity (`===`) won't match across renders.
}), es = z({
  alt: c().optional(),
  autoComplete: s().optional(),
  checked: f().optional(),
  disabled: f().optional(),
  height: s().optional(),
  list: c().optional(),
  name: c().optional(),
  placeholder: c().optional(),
  readOnly: f().optional(),
  required: f().optional(),
  width: s().optional(),
  className: c().optional(),
  id: c().optional(),
  lang: c().optional(),
  title: c().optional(),
  children: v([c(), g(), f(), b(), m]).optional(),
  value: c().optional(),
  // Controlled value
  size: y(["xs", "sm", "base", "lg"]).optional(),
  // Size of the input. - `"xs"` — Extra small for compact UIs - `"sm"` — Small for secondary fields - `"base"` — Default input size - `"lg"` — Large for prominent fields
  variant: y(["default", "error"]).optional(),
  // Style variant of the input. - `"default"` — Default input appearance - `"error"` — Error state for validation failures
  label: v([c(), g(), f(), b(), m]).optional(),
  // Label content for the input (enables Field wrapper and sets masked state label) - can be a string or any React node
  labelTooltip: v([c(), g(), f(), b(), m]).optional(),
  // Tooltip content to display next to the label via an info icon
  description: v([c(), g(), f(), b(), m]).optional(),
  // Helper text displayed below the input
  error: s().optional()
  // Error message or validation error object
}), os = z({
  defaultOpen: f().optional(),
  // Initial open state when uncontrolled.
  open: f().optional(),
  // Controlled open state.
  variant: y(["sidebar", "floating", "inset"]).optional(),
  // Sidebar layout variant.
  side: y(["left", "right"]).optional(),
  // Which side the sidebar is on.
  collapsible: y(["icon", "offcanvas", "none"]).optional(),
  resizable: f().optional(),
  // Enable drag-to-resize on the sidebar edge.
  defaultWidth: g().optional(),
  // Initial width in pixels when resizable.
  minWidth: g().optional(),
  // Minimum width in pixels when resizing.
  maxWidth: g().optional(),
  // Maximum width in pixels when resizing.
  children: v([c(), g(), f(), b(), m]).optional(),
  // Content — typically `<Sidebar>` + main content.
  className: c().optional()
  // Additional CSS classes for the wrapper div.
}), ts = z({
  as: s().optional(),
  render: v([c(), g(), f(), b(), m]).optional(),
  // Allows you to replace the component's HTML element with a different tag, or compose it with another component. Accepts a `ReactElement` or a function that returns the element to render.
  children: v([c(), g(), f(), b(), m]).optional(),
  className: c().optional(),
  id: c().optional(),
  lang: c().optional(),
  title: c().optional()
}), ns = z({
  variant: y(["default", "neutral"]).optional(),
  // Visual variant: "default" (pill, brand color) or "neutral" (squircle, monochrome)
  label: v([c(), g(), f(), b(), m]).optional(),
  // Label content for the switch (Field wrapper is built-in) - can be a string or any React node. Optional when used standalone for visual-only purposes.
  labelTooltip: v([c(), g(), f(), b(), m]).optional(),
  // Tooltip content to display next to the label via an info icon
  required: f().optional(),
  // Whether the switch is required. When explicitly false, shows "(optional)" text after the label.
  controlFirst: f().optional(),
  // When true (default), switch appears before label. When false, label appears before switch.
  size: y(["sm", "base", "lg"]).optional(),
  checked: f().optional(),
  disabled: f().optional(),
  transitioning: f().optional(),
  name: c().optional(),
  type: y(["submit", "reset", "button"]).optional(),
  value: s().optional(),
  className: c().optional(),
  id: c().optional(),
  lang: c().optional(),
  title: c().optional(),
  onClick: s()
  // Callback when switch is clicked
}), rs = z({
  layout: y(["auto", "fixed"]).optional(),
  variant: y(["default", "selected"]).optional(),
  sticky: y(["left", "right"]).optional(),
  className: c().optional(),
  // Additional CSS classes
  children: v([c(), g(), f(), b(), m]).optional()
  // Child elements
}), is = z({
  children: v([c(), g(), f(), b(), m]).optional(),
  className: c().optional(),
  id: c().optional(),
  lang: c().optional(),
  title: c().optional()
}), as = z({
  tabs: U(s()).optional(),
  // Array of tab items to render.
  value: c().optional(),
  // Controlled value. When set, component becomes controlled.
  selectedValue: c().optional(),
  // Default selected value for uncontrolled mode. Ignored when `value` is set.
  activateOnFocus: f().optional(),
  // When `true`, tabs are activated immediately upon receiving focus via arrow keys. When `false` (default), tabs receive focus but require Enter/Space to activate.
  className: c().optional(),
  // Additional CSS classes for the root element.
  listClassName: c().optional(),
  // Additional CSS classes for the tab list element.
  indicatorClassName: c().optional(),
  // Additional CSS classes for the indicator element.
  variant: y(["segmented", "underline"]).optional(),
  // Tab style. - `"segmented"` — Pill-shaped indicator on a filled track - `"underline"` — Underline indicator below tab text
  onValueChange: s().optional()
  // Callback when active tab changes
}), ss = z({
  variant: y(["heading1", "heading2", "heading3", "body", "secondary", "success", "error", "mono", "mono-secondary"]).optional(),
  // Text style variant. Determines color, font, and weight. - `"heading1"` — Large page title (30px, semibold) - `"heading2"` — Section title (24px, semibold) - `"heading3"` — Subsection title (18px, semibold) - `"body"` — Default body text - `"secondary"` — Muted text for secondary information - `"success"` — Success state text - `"error"` — Error state text - `"mono"` — Monospace text for code - `"mono-secondary"` — Muted monospace text
  size: y(["xs", "sm", "base", "lg"]).optional(),
  // Text size (only applies to body/secondary/success/error variants). - `"xs"` — 12px - `"sm"` — 14px - `"base"` — 16px - `"lg"` — 18px
  bold: f().optional(),
  // Whether to use bold font weight (only applies to body variants).
  truncate: f().optional(),
  // Whether to truncate overflowing text with an ellipsis. Adds `truncate min-w-0` classes.
  as: y(["h1", "h2", "h3", "h4", "h5", "h6", "p", "span", "label", "dt", "dd", "li", "figcaption", "legend", "pre", "code", "em", "strong", "small", "abbr", "time"]).optional(),
  // The HTML element to render. Accepts headings (`"h1"`–`"h6"`), block text (`"p"`, `"pre"`), inline text (`"span"`, `"code"`, `"em"`, `"strong"`, `"small"`, `"abbr"`, `"time"`), form-related (`"label"`, `"legend"`), list/definition (`"dt"`, `"dd"`, `"li"`), and `"figcaption"`. - **Required** for heading variants (`"heading1"`, `"heading2"`,   `"heading3"`) — pick the element that reflects this text's place in   the document outline, or `"span"` for decorative heading-styled text   that is not a section heading. - **Optional** for body variants (defaults to `"p"`) and monospace   variants (defaults to `"span"`).
  children: v([c(), g(), f(), b(), m]).optional()
  // Text content.
}), cs = z({
  variant: y(["default", "success", "error", "warning", "info"]).optional(),
  className: c().optional(),
  // Additional CSS classes
  children: v([c(), g(), f(), b(), m]).optional()
  // Child elements
}), ls = z({
  side: y(["top", "bottom", "left", "right"]).optional(),
  className: c().optional(),
  // Additional CSS classes
  children: v([c(), g(), f(), b(), m]).optional(),
  // Child elements
  content: v([c(), g(), f(), b(), m])
  // Content to display in the tooltip
}), us = y([
  "Autocomplete",
  "Badge",
  "Banner",
  "Breadcrumbs",
  "Button",
  "Checkbox",
  "ClipboardText",
  "CloudflareLogo",
  "Code",
  "Collapsible",
  "Combobox",
  "CommandPalette",
  "DatePicker",
  "DateRangePicker",
  "Dialog",
  "DropdownMenu",
  "Empty",
  "Field",
  "Grid",
  "Input",
  "InputArea",
  "InputGroup",
  "Label",
  "LayerCard",
  "Link",
  "Loader",
  "MenuBar",
  "Meter",
  "Pagination",
  "Popover",
  "Radio",
  "Select",
  "SensitiveInput",
  "Sidebar",
  "Surface",
  "Switch",
  "Table",
  "TableOfContents",
  "Tabs",
  "Text",
  "Toasty",
  "Tooltip"
]), ps = {
  Autocomplete: ka,
  Badge: wa,
  Banner: Sa,
  Breadcrumbs: $a,
  Button: Za,
  Checkbox: Pa,
  ClipboardText: Ta,
  CloudflareLogo: Oa,
  Code: Ia,
  Collapsible: Na,
  Combobox: Ea,
  CommandPalette: Ca,
  DatePicker: Aa,
  DateRangePicker: xa,
  Dialog: ja,
  DropdownMenu: Ra,
  Empty: Da,
  Field: Ua,
  Grid: Fa,
  Input: La,
  InputArea: Ma,
  InputGroup: Ja,
  Label: Va,
  LayerCard: Ba,
  Link: Wa,
  Loader: Ga,
  MenuBar: qa,
  Meter: Ka,
  Pagination: Xa,
  Popover: Ya,
  Radio: Ha,
  Select: Qa,
  SensitiveInput: es,
  Sidebar: os,
  Surface: ts,
  Switch: ns,
  Table: rs,
  TableOfContents: is,
  Tabs: as,
  Text: ss,
  Toasty: cs,
  Tooltip: ls
}, ds = z({
  key: c(),
  type: us,
  props: H(c(), s()),
  children: U(c()).optional(),
  parentKey: c().nullable().optional(),
  visible: ba.optional(),
  action: za.optional()
}), fs = z({
  root: c(),
  elements: H(c(), ds)
});
function ms(e) {
  const o = ps[e.type];
  return o ? o.safeParse(e.props) : { success: !1, error: new li([{ code: "custom", message: `Unknown component type: ${e.type}`, path: ["type"] }]) };
}
function gs(e) {
  return fs.safeParse(e);
}
const _s = ["Autocomplete", "Badge", "Banner", "Breadcrumbs", "Button", "Checkbox", "ClipboardText", "CloudflareLogo", "Code", "Collapsible", "Combobox", "CommandPalette", "DatePicker", "DateRangePicker", "Dialog", "DropdownMenu", "Empty", "Field", "Grid", "Input", "InputArea", "InputGroup", "Label", "LayerCard", "Link", "Loader", "MenuBar", "Meter", "Pagination", "Popover", "Radio", "Select", "SensitiveInput", "Sidebar", "Surface", "Switch", "Table", "TableOfContents", "Tabs", "Text", "Toasty", "Tooltip"];
export {
  ya as ActionConfirmSchema,
  za as ActionSchema,
  ka as AutocompletePropsSchema,
  wa as BadgePropsSchema,
  Sa as BannerPropsSchema,
  $a as BreadcrumbsPropsSchema,
  Za as ButtonPropsSchema,
  Pa as CheckboxPropsSchema,
  Ta as ClipboardTextPropsSchema,
  Oa as CloudflareLogoPropsSchema,
  Ia as CodePropsSchema,
  Na as CollapsiblePropsSchema,
  Ea as ComboboxPropsSchema,
  Ca as CommandPalettePropsSchema,
  ps as ComponentPropsSchemas,
  Aa as DatePickerPropsSchema,
  xa as DateRangePickerPropsSchema,
  ja as DialogPropsSchema,
  Ra as DropdownMenuPropsSchema,
  m as DynamicValueSchema,
  Da as EmptyPropsSchema,
  Ua as FieldPropsSchema,
  Fa as GridPropsSchema,
  Ma as InputAreaPropsSchema,
  Ja as InputGroupPropsSchema,
  La as InputPropsSchema,
  _s as KUMO_COMPONENT_NAMES,
  us as KumoComponentTypeSchema,
  Va as LabelPropsSchema,
  Ba as LayerCardPropsSchema,
  Wa as LinkPropsSchema,
  Ga as LoaderPropsSchema,
  qa as MenuBarPropsSchema,
  Ka as MeterPropsSchema,
  Xa as PaginationPropsSchema,
  Ya as PopoverPropsSchema,
  Ha as RadioPropsSchema,
  Qa as SelectPropsSchema,
  es as SensitiveInputPropsSchema,
  os as SidebarPropsSchema,
  ts as SurfacePropsSchema,
  ns as SwitchPropsSchema,
  is as TableOfContentsPropsSchema,
  rs as TablePropsSchema,
  as as TabsPropsSchema,
  ss as TextPropsSchema,
  cs as ToastyPropsSchema,
  ls as TooltipPropsSchema,
  ds as UIElementBaseSchema,
  fs as UITreeSchema,
  ba as VisibilityConditionSchema,
  ms as validateElementProps,
  gs as validateUITree
};
